---
title: "DE_analysis"
author: "Ville Lehtonen"
date: "11/3/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Set working dir to the source file location
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
```

# Marker gene detection
Here, we perform marker gene detection to identify differentially expressed (DE) genes between samples in the datasets

## 1. Load packages and data
### 1.1 Load packages
```{r load-packages, message=FALSE}
library(SingleCellExperiment)
library(scran)
library(pheatmap)
library(Seurat)
library(plyr)
```

### 1.2 Load pre-calculated data
**Single Cell Experiments**
```{r load-objects}
# Load pre-processed sce objects
denoised.sce <- readRDS("./../saved/R_objects/denoised_sce_clust")
denoised.sce.deMicheli <- readRDS("./../saved/R_objects/denoised_sce_deMicheli")

# Generate the CC corrected data
denoised.sce.cc <- as(altExp(denoised.sce, "CC_corrected"), "SingleCellExperiment")
denoised.sce.cc$cluster.cc <- denoised.sce$cluster.cc
```

## 1. Primary data marker genes
### 1.1 Pairwise comparisons with t-test
**Markers by sample: Control**
```{r de-genes-primary-data-samples, eval=FALSE}
# Find markers by sample
markers.primary <- findMarkers(denoised.sce, 
                               groups = denoised.sce$sample,
                               lfc = log2(1.5))

# Choose markers for control sample
control.markers <- markers.primary[["control"]]

# Save marker information to .csv file
write.csv(control.markers, file = "./../saved/tables/control_marker_genes.csv")

# Subset top 6 genes (ranked by p -value) from each pairwise comparison involving control sample
best.set <- control.markers[control.markers$Top <= 6,]

# Extract the marker effect sizes
logFCs <- getMarkerEffects(best.set)

# Plot results in a heatmap
pheatmap(logFCs, breaks=seq(-5, 5, length.out=101), show_rownames = TRUE, show_colnames = TRUE, main = "Marker genes (t-test): Control vs. other samples")
```

**Markers by cluster**
```{r de-genes-primary-data-clusters, eval=FALSE}
# Find markers by sample
markers.primary.clust <- findMarkers(denoised.sce, 
                               groups = denoised.sce$cluster,
                               lfc = log2(1.5))

# Perform marker gene identification for all clusters
for(cluster in unique(denoised.sce$cluster)) {
  # Choose marker for cluster x
  markers <- markers.primary.clust[[cluster]]
  # Save marker information to .csv file
  fpath <- paste0("./../saved/tables/marker_genes_cluster_", cluster, ".csv")
  write.csv(markers, file = fpath)
  # Subset top 6 genes (ranked by p -value) from each pairwise comparison involving the cluster x
  best.set <- markers[markers$Top <= 6,]
  # Extract the marker effect sizes
  logFCs <- getMarkerEffects(best.set)
  # Plot results in a heatmap
  pheatmap(logFCs, breaks=seq(-5, 5, length.out=101), show_rownames = TRUE, show_colnames = TRUE, main = paste("Marker genes (t-test): cluster:", cluster))
}
```

**Markers by cell cycle corrected clusters**
```{r de-genes-primary-data-clusters-cc, eval=FALSE}
# Find markers by sample
markers.primary.clust.cc <- findMarkers(denoised.sce, 
                               groups = denoised.sce$cluster.cc,
                               lfc = log2(1.5))

# Perform marker gene identification for all clusters
for(cluster in unique(denoised.sce$cluster.cc)) {
  # Choose marker for cluster x
  markers <- markers.primary.clust.cc[[cluster]]
  # Save marker information to .csv file
  fpath <- paste0("./../saved/tables/marker_genes_cluster_", cluster, "_cc.csv")
  write.csv(markers, file = fpath)
  # Subset top 6 genes (ranked by p -value) from each pairwise comparison involving the cluster x
  best.set <- markers[markers$Top <= 6,]
  # Extract the marker effect sizes
  logFCs <- getMarkerEffects(best.set)
  # Plot results in a heatmap
  pheatmap(logFCs, breaks=seq(-5, 5, length.out=101), show_rownames = TRUE, show_colnames = TRUE, main = paste("Marker genes (t-test): cc corected cluster:", cluster))
}
```


### 1.2 Pairwise comparisons with Wilcoxon rank sum (Wilcoxon-Mann-Whitney, wmw)  test
**Markers by sample: Control**
```{r de-genes-primary-data-samples-wmw}
# Find markers by sample
markers.primary.wmw <- findMarkers(denoised.sce, 
                                   groups = denoised.sce$sample,
                                   lfc = log2(1.5),
                                   test = "wilcox")

# Choose markers for control sample
control.markers.wmw <- markers.primary.wmw[["control"]]

# Save marker information to .csv file
write.csv(control.markers.wmw, file = "./../saved/tables/control_marker_genes_wmw.csv")

# Subset top 6 genes (ranked by p -value) from each pairwise comparison involving control sample
best.set <- control.markers.wmw[control.markers.wmw$Top <= 6,]

# Extract the area under curves
AUCs <- getMarkerEffects(best.set, prefix="AUC")

# Plot results in a heatmap
pheatmap(AUCs, breaks=seq(0, 1, length.out=21), color=viridis::viridis(21), main = "Marker genes (wmw): Control vs. other samples")
```

**Markers by cluster**
```{r de-genes-primary-data-cluster-wmw}
# Find markers by sample
markers.primary.clust.wmw <- findMarkers(denoised.sce, 
                                   groups = denoised.sce$cluster,
                                   lfc = log2(1.5),
                                   test = "wilcox")

# Perform marker gene identification for all clusters
for(cluster in unique(denoised.sce$cluster)) {
  # Choose marker for cluster x
  markers.clust.wmw <- markers.primary.clust.wmw[[cluster]]
  # Save marker information to .csv file
  fpath <- paste0("./../saved/tables/marker_genes_cluster_", cluster, "_wmw.csv")
  write.csv(markers.clust.wmw, file = fpath)
  # Subset top 6 genes (ranked by p -value) from each pairwise comparison involving the cluster x
  best.set <- markers.clust.wmw[markers.clust.wmw$Top <= 6,]
  # Extract the area under curves
  AUCs <- getMarkerEffects(best.set, prefix="AUC")
  # Plot results in a heatmap
  pheatmap(AUCs, breaks=seq(0, 1, length.out=21), color=viridis::viridis(21), main = paste("Marker genes (wmw): cluster:", cluster))
}
```

**Markers by cell cycle corrected clusters**
```{r de-genes-primary-data-cluster-wmw-cc}
# Find markers by sample
markers.primary.clust.wmw.cc <- findMarkers(denoised.sce, 
                                   groups = denoised.sce$cluster.cc,
                                   lfc = log2(1.5),
                                   test = "wilcox")

# Perform marker gene identification for all clusters
for(cluster in unique(denoised.sce$cluster.cc)) {
  # Choose marker for cluster x
  markers.clust.wmw <- markers.primary.clust.wmw.cc[[cluster]]
  # Save marker information to .csv file
  fpath <- paste0("./../saved/tables/marker_genes_cluster_", cluster, "cc_wmw.csv")
  write.csv(markers.clust.wmw, file = fpath)
  # Subset top 6 genes (ranked by p -value) from each pairwise comparison involving the cluster x
  best.set <- markers.clust.wmw[markers.clust.wmw$Top <= 6,]
  # Extract the area under curves
  AUCs <- getMarkerEffects(best.set, prefix="AUC")
  # Plot results in a heatmap
  pheatmap(AUCs, breaks=seq(0, 1, length.out=21), color=viridis::viridis(21), main = paste("Marker genes (wmw): cc corrected cluster:", cluster))
}
```


## 2. Ridgeline plots to inspect how the distribution of marker genes differs in the samples of primary and reference datasets
```{r ridge-plots, message=FALSE, warning=FALSE}
genes.primary = c("Pax7", "Myf5", "Myod1", "Cdh15", "Tnnt3", "Hes1", "Hes6", "Notch1", "Id1", "Id2", "Id3", "Wnt7b")
genes.deMicheli = c("Pax7", "Myf5", "Myod1", "Myog", "Cdh15", "Tnnt3", "Hes1", "Heyl", "Hes6", "Notch1", "Notch2", "Notch3", "Id1", "Id2", "Id3", "Wnt2", "Wnt10b")

# De Micheli data
sce.deMicheli.f <- denoised.sce.deMicheli[, denoised.sce.deMicheli$cell_type %in% c("d0_MuSCs_and_progenitors", "d5_MuSCs_and_progenitors")]

sce.deMicheli.f$cell_type <- revalue(sce.deMicheli.f$cell_type, c("d0_MuSCs_and_progenitors" = "Day 0", "d5_MuSCs_and_progenitors" = "Day 5"))
seurat.deMicheli <- as.Seurat(sce.deMicheli.f)
Idents(seurat.deMicheli) <- "cell_type"
ridge.deMicheli <- RidgePlot(seurat.deMicheli, features = genes.deMicheli, ncol = 2)

# Primary data
seurat.primary <- as.Seurat(denoised.sce)
Idents(seurat.primary) <- "sample"
ridge.primary <- RidgePlot(seurat.primary, features = genes.primary, ncol = 2)
```

**Save ridge line plots as pdf**
```{r ridge-plots-pdf, eval=FALSE}
pdf("./../saved/figures/ridge_plot_primary.pdf", width = 10, height = 25)
ridge.primary
dev.off()

pdf("./../saved/figures/ridge_plot_deMicheli.pdf", width = 10, height = 25)
ridge.deMicheli
dev.off()
```


