---
title: "DE_analysis"
author: "Ville Lehtonen"
date: "11/3/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Set working dir to the source file location
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
```

# Marker gene detection
Here, we perform marker gene detection to identify differentially expressed (DE) genes between samples in the datasets

## 1. Load packages and data
### 1.1 Load packages
```{r load-packages, message=FALSE}
library(SingleCellExperiment)
library(scran)
library(pheatmap)
library(Seurat)
library(plyr)
```

### 1.2 Load pre-calculated data
**Single Cell Experiments**
```{r load-objects}
# Load pre-processed sce objects
denoised.sce <- readRDS("./../saved/denoised_sce_clust")
denoised.sce.deMicheli <- readRDS("./../saved/denoised_sce_deMicheli")

# Generate the CC corrected data
denoised.sce.cc <- as(altExp(denoised.sce, "CC_corrected"), "SingleCellExperiment")
```

## 1. Primary data marker genes
### 1.1 Uncorrected dataset
```{r de-genes-primary-data}
markers.primary <- findMarkers(denoised.sce, 
                               groups = denoised.sce$sample,
                               pval.type = "some",
                               direction = "up",
                               lfc = 1)

for(sample in unique(denoised.sce$sample)) {
  markers <- markers.primary[[sample]]
  markers.interesting <- markers[markers$p.value <= .05, ]
  # Print out the marker genes
  print(markers.interesting)
  # Get marker effects and plot on a heatmap
  logFCs <- getMarkerEffects(markers.interesting)
  pheatmap(logFCs, breaks=seq(-2.5, 2.5, length.out=101), main = paste(sample, " marker genes"), fontsize_row = 5)
}
```


### 1.2 Cell cycle corrected dataset
```{r de-genes-primary-data-cc-corrected}
markers.primary.cc <- findMarkers(denoised.sce.cc, 
                               groups = denoised.sce.cc$sample,
                               assay.type = "scaled_logcounts",
                               pval.type = "some",
                               direction = "up",
                               lfc = 1)

for(sample in unique(denoised.sce$sample)) {
  markers <- markers.primary.cc[[sample]]
  markers.interesting <- markers[markers$p.value < .05, ]
  # Print out the marker genes
  print(markers.interesting)
  # Get marker effects and plot on a heatmap
  logFCs <- getMarkerEffects(markers.interesting)
  pheatmap(logFCs, breaks=seq(-2.5, 2.5, length.out=101), main = paste(sample, " marker genes"), fontsize_row = 5)
}
```


## 2. Ridgeline plots to inspect how the distribution of marker genes differs in the samples of primary and reference datasets
```{r ridge-plots, message=FALSE, warning=FALSE}
genes.primary = c("Pax7", "Myf5", "Myod1", "Cdh15", "Tnnt3", "Hes1", "Hes6", "Notch1", "Id1", "Id2", "Id3", "Wnt7b")
genes.deMicheli = c("Pax7", "Myf5", "Myod1", "Myog", "Cdh15", "Tnnt3", "Hes1", "Heyl", "Hes6", "Notch1", "Notch2", "Notch3", "Id1", "Id2", "Id3", "Wnt2", "Wnt10b")

# De Micheli data
sce.deMicheli.f <- denoised.sce.deMicheli[, denoised.sce.deMicheli$cell_type %in% c("d0_MuSCs_and_progenitors", "d5_MuSCs_and_progenitors")]

sce.deMicheli.f$cell_type <- revalue(sce.deMicheli.f$cell_type, c("d0_MuSCs_and_progenitors" = "Day 0", "d5_MuSCs_and_progenitors" = "Day 5"))
seurat.deMicheli <- as.Seurat(sce.deMicheli.f)
Idents(seurat.deMicheli) <- "cell_type"
ridge.deMicheli <- RidgePlot(seurat.deMicheli, features = genes.deMicheli, ncol = 2)

# Primary data
seurat.primary <- as.Seurat(denoised.sce)
Idents(seurat.primary) <- "sample"
ridge.primary <- RidgePlot(seurat.primary, features = genes.primary, ncol = 2)
```

**Save ridge line plots as pdf**
```{r ridge-plots-pdf, eval=FALSE}
pdf("./../saved/ridge_plot_primary.pdf", width = 10, height = 25)
ridge.primary
dev.off()

pdf("./../saved/ridge_plot_deMicheli.pdf", width = 10, height = 25)
ridge.deMicheli
dev.off()
```


