---
title: "Downstream_analysis"
author: "Ville Lehtonen"
date: "9/7/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Set working dir to the source file location
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
```
# Trajectory inference
SCORPIUS (https://github.com/rcannood/SCORPIUS) will be used for performing the trajectory inference as it is designed for inferring linear developmental trajectories from scRNA-seq data


## 1. Load packages and data
### Load packages
```{r load-packages, message=FALSE}
library(SCORPIUS)
library(scater)
library(scran)
library(SingleCellExperiment)
library(dplyr)
library(SingleR)
library(igraph)
library(stats)
library(cluster)
library(pheatmap)
library(fossil)
library(Seurat)
library(cerebroApp)
library(reshape)
library(data.table)
library(scRNAseq)
library(org.Mm.eg.db)
library(tidyr)
library(BBmisc)
library(KEGGREST)
library(RColorBrewer)
library(ggpubr)
library(grid)
library(gridExtra)
library(plyr)
library(stringr)
library(ggrepel)
```


### Load saved objects
**Single Cell Experiments**
```{r load-objects}
# Load pre-processed sce objects
denoised.sce <- readRDS("./../saved/denoised_sce_clust")
denoised.sce.deMicheli <- readRDS("./../saved/denoised_sce_deMicheli")
denoised.sce.dellOrso <- readRDS("./../saved/denoised_sce_dellOrso")
```

**Cell cycle genes (From Seurat pipeline)**
```{r cc-genes-seurat}
# Load G2/M- and S- phase marker genes
g2m.genes.mouse <- readRDS("./../saved/cc_g2m_genes_mouse")
s.genes.mouse <- readRDS("./../saved/cc_gs_genes_mouse")

# Combine genes into a single set of cc genes
cc.genes.seurat <- c(g2m.genes.mouse, s.genes.mouse)
```


**Cell cycle genes (from OSCA pipeline)**
```{r cc-genes-osca-shortcut}
cc.genes.osca <- readRDS("./../saved/cc_genes_osca")

# Remove cell cycle genes from the datasets
denoised.sce.deMicheli.cc.genes <- denoised.sce.deMicheli[!rownames(denoised.sce.deMicheli) %in% cc.genes.osca, ]
denoised.sce.dellOrso.cc.genes <- denoised.sce.dellOrso[!rownames(denoised.sce.dellOrso) %in% cc.genes.osca, ]
denoised.sce.cc.genes <- denoised.sce[!rownames(denoised.sce) %in% cc.genes.osca, ]
```


```{r cc-genes-osca, eval=FALSE}
# Use Buettner et al. (2015) mouse data with known cell cycle phases as a reference
sce.ref <- BuettnerESCData()

# Find genes present in both our datasets and Buettner data and that are cell cycle related
cycle.anno <- unique(select(org.Mm.eg.db, keytype="GOALL", keys="GO:0007049", 
    columns="SYMBOL")[,"SYMBOL"])

cc.genes.osca <- Reduce(intersect,
                     list(as.character(rowData(sce.ref)$AssociatedGeneName),    
                          rownames(denoised.sce),
                          rownames(denoised.sce.deMicheli),
                          rownames(denoised.sce.deMicheli),
                          cycle.anno))

saveRDS(cc.genes.osca, "./../saved/cc_genes_osca")

# Identifying markers between cell cycle phases.
sce.ref <- logNormCounts(sce.ref)
rownames(sce.ref) <- as.character(rowData(sce.ref)$AssociatedGeneName)
phase.stats <- pairwiseWilcox(logcounts(sce.ref), sce.ref$phase, 
                              direction="up", subset.row=cc.genes.osca)
cycle.markers <- getTopMarkers(phase.stats[[1]], phase.stats[[2]])
```

```{r cc-genes-comparison}
# Inspect the cc related gene sets
length(cc.genes.seurat)
length(cc.genes.osca)

# Inspect the intersection of the two gene sets
intersect(cc.genes.osca, cc.genes.seurat)
```


## 2. Clustering
### 2.1 Primary data

**The clusters are calculated using Puhti supercomputer and results are loaded in the code chunk below to speed up the process**
```{r primary-data-clustering-shortcut}
# Load data with clusters
snng <- readRDS("./../saved/snng")
snng.cc <- readRDS("./../saved/snng_cc")
clust <- readRDS("./../saved/clust")
clust.cc <- readRDS("./../saved/clust_cc")

# Generate the CC corrected data
denoised.sce.cc <- as(altExp(denoised.sce, "CC_corrected"), "SingleCellExperiment")
denoised.sce.cc$cluster.cc <- factor(clust.cc)
reducedDim(denoised.sce.cc, "UMAP_cc") <- reducedDim(denoised.sce, "UMAP_cc")
reducedDim(denoised.sce.cc, "TSNE_cc") <- reducedDim(denoised.sce, "TSNE_cc")
reducedDim(denoised.sce.cc, "force_cc") <- reducedDim(denoised.sce, "force_cc")
```

**Perform clustering (in case no pre-calculated files are available):**
```{r primary-data-clustering, eval=FALSE}
# Load data
denoised.sce <- readRDS("./../saved/denoised_sce")

# Get # of reduced dimensions for uncorrected & cc-corrected datasets
dimensions <- ncol(reducedDim(denoised.sce, "PCA"))
dimensions.cc <- ncol(reducedDim(denoised.sce, "PCA_cc"))
dimensions
dimensions.cc

# Build SNN graph with k=180
snng <- buildSNNGraph(denoised.sce, k = 180, use.dimred = "PCA") 

denoised.sce.cc <- as(altExp(denoised.sce, "CC_corrected"), "SingleCellExperiment")
reducedDim(denoised.sce.cc, "PCA_cc") <- reducedDim(denoised.sce, "PCA_cc")
snng.cc <- buildSNNGraph(denoised.sce.cc, k = 180, use.dimred = "PCA_cc")

# Perform clustering and assign cluster labels
clust <- igraph::cluster_walktrap(snng)$membership
clust.cc <- igraph::cluster_walktrap(snng.cc)$membership
table(clust)
table(clust.cc)
denoised.sce$cluster <- factor(clust)
denoised.sce$cluster.cc <- factor(clust.cc)

# Run UMAP, t-SNE, and force-directed layout
## Uncorrected dataset
set.seed(123)
reducedDim(denoised.sce, "force") <- igraph::layout_with_fr(snng)
denoised.sce <- runUMAP(denoised.sce, dimred = "PCA", n_dimred = dimensions)
denoised.sce <- runTSNE(denoised.sce, dimred = "PCA", n_dimred = dimensions)
## Cc corrected data
reducedDim(denoised.sce, "force_cc") <- igraph::layout_with_fr(snng.cc)
denoised.sce <- runTSNE(denoised.sce, altexp = "CC_corrected", exprs_values = "scaled_logcounts", dimred = "PCA_cc", name = "TSNE_cc")
denoised.sce.cc <- runUMAP(denoised.sce.cc, exprs_values = "scaled_logcounts", dimred = "PCA_cc", name = "UMAP_cc", n_neighbors = 180)

# Set reduced dims to cc corrected experiment and vice versa
reducedDim(denoised.sce, "UMAP_cc") <- reducedDim(denoised.sce.cc, "UMAP_cc")
reducedDim(denoised.sce.cc, "TSNE_cc") <- reducedDim(denoised.sce, "TSNE_cc")
reducedDim(denoised.sce.cc, "force_cc") <- reducedDim(denoised.sce, "force_cc")

# Save sce file
#saveRDS(denoised.sce, file = "./../saved/denoised_sce_clust")
```


**Plot the data using UMAP, force-directed layout, and t-SNE**
```{r primary-data-clustering-plots}
# force directed layout plots
p.force <- plotReducedDim(denoised.sce, colour_by = "sample", dimred = "force") + ggtitle("Primary data: force-directed layout (k=180)") + theme(plot.title = element_text(size = 16, face = "bold"))
p.force.cc <- plotReducedDim(denoised.sce.cc, colour_by = "sample", dimred = "force_cc") + ggtitle("Primary data (cc corrected): force-directed layout (k=180)") + theme(plot.title = element_text(size = 16, face = "bold"))
gridExtra::grid.arrange(p.force, p.force.cc, ncol = 2)

# t-SNE plots
## Uncorrected
p.tsne <- plotReducedDim(denoised.sce, 
               colour_by = "sample",
               text_by = "cluster",
               dimred = "TSNE",
               point_alpha = 0.4,
               point_size = 1) +
  ggtitle("t-SNE projection to visualize variation \nin single cell transcriptomes", subtitle = "t-SNE plot of primary data: Number of nearest neighbours (k): 180,\ncells coloured by sample groups and clusters (1-5) shown in the plot") +
  theme(plot.title = element_text(size = 20, face = "bold"),
        plot.subtitle = element_text(size = 12),
        legend.text = element_text(size = 16),
        legend.title = element_text(size = 20, face = "bold"),
        axis.title = element_text(size = 20, face = "bold")) +
  scale_fill_discrete(type = c("#E41A1C", "#377EB8", "#4DAF4A", "#FFFF33", "#FF7F00"), guide = guide_legend(override.aes = list(size = 5, alpha = 1)))

## cc corrected
p.tsne.cc <- plotReducedDim(denoised.sce.cc, 
               colour_by = "sample",
               text_by = "cluster.cc",
               dimred = "TSNE_cc",
               point_alpha = 0.4,
               point_size = 1) +
  ggtitle("t-SNE projection to visualize variation \nin single cell transcriptomes after cell\ncycle correction", subtitle = "t-SNE plot of cell cycle corrected primary data: Number of\nnearest neighbours (k): 180, cells coloured by sample groups\nand clusters (1-5) shown in the plot") +
  theme(plot.title = element_text(size = 20, face = "bold"),
        plot.subtitle = element_text(size = 12),
        legend.text = element_text(size = 16),
        legend.title = element_text(size = 20, face = "bold"),
        axis.title = element_text(size = 20, face = "bold")) +
  scale_fill_discrete(type = c("#E41A1C", "#377EB8", "#4DAF4A", "#FFFF33", "#FF7F00"), guide = guide_legend(override.aes = list(size = 5, alpha = 1)))

p.tsne
p.tsne.cc

# UMAP plots by sample
## Uncorrected
p.umap <- plotReducedDim(denoised.sce, 
               colour_by = "sample",
               text_by = "cluster",
               dimred = "UMAP",
               point_alpha = 0.4,
               point_size = 1) +
  ggtitle("UMAP projection to visualize variation \nin single cell transcriptomes", subtitle = "UMAP plot of primary data: Number of nearest neighbours (k): 180,\ncells coloured by sample groups and clusters (1-5) shown in the plot") +
  theme(plot.title = element_text(size = 20, face = "bold"),
        plot.subtitle = element_text(size = 12),
        legend.text = element_text(size = 16),
        legend.title = element_text(size = 20, face = "bold"),
        axis.title = element_text(size = 20, face = "bold")) +
  scale_fill_discrete(type = c("#E41A1C", "#377EB8", "#4DAF4A", "#FFFF33", "#FF7F00"), guide = guide_legend(override.aes = list(size = 5, alpha = 1)))



## CC-corrected
p.umap.cc <- plotReducedDim(denoised.sce.cc, 
               colour_by = "sample",
               text_by = "cluster.cc",
               dimred = "UMAP_cc",
               point_alpha = 0.4,
               point_size = 1) +
  ggtitle("UMAP projection to visualize variation \nin single cell transcriptomes after cell\ncycle correction", subtitle = "UMAP plot of cell cycle corrected primary data: Number of\nnearest neighbours (k): 180, cells coloured by sample groups\nand clusters (1-5) shown in the plot") +
  theme(plot.title = element_text(size = 20, face = "bold"),
        plot.subtitle = element_text(size = 12),
        legend.text = element_text(size = 16),
        legend.title = element_text(size = 20, face = "bold"),
        axis.title = element_text(size = 20, face = "bold")) +
  scale_fill_discrete(type = c("#E41A1C", "#377EB8", "#4DAF4A", "#FFFF33", "#FF7F00"), guide = guide_legend(override.aes = list(size = 5, alpha = 1)))

p.umap
p.umap.cc

# Umap plots by cluster
plotReducedDim(denoised.sce, colour_by = "cluster", dimred = "UMAP") + ggtitle("Primary data: UMAP with clusters (k=180)") + theme(plot.title = element_text(size = 16, face = "bold"))
plotReducedDim(denoised.sce.cc, colour_by = "cluster.cc", dimred = "UMAP_cc") + ggtitle("Primary data (cc corrected): UMAP with clusters (k=180)") + theme(plot.title = element_text(size = 16, face = "bold"))

```
**Save UMAP and t-SNE plots in a pdf file**
```{r umap-tsne-into-pdf, eval=FALSE}
pdf("./../saved/primary_data_umap_and_tsne_plots.pdf", width = 7, height = 7)
p.tsne
p.tsne.cc
p.umap
p.umap.cc
dev.off()

png("./../saved/primary_data_umap_and_tsne.plots.png", width = 1600, height = 900)
gridExtra::grid.arrange(p.umap, p.umap.cc, p.tsne, p.tsne.cc, ncol = 2)
dev.off()

```



### 2.2 Clustering comparison
**Calculate graph modularity and ARI of clusterings**
```{r primary-data-clustering-comparison}
# Calculate modularities of the graphs
mod <- modularity(snng, clust)
mod.cc <- modularity(snng.cc, clust.cc)
mod
mod.cc

# Calculate adjusted Rand index
ari <- adj.rand.index(clust, clust.cc)
ari
```


**Calculate the modularity of each cluster and visualize the results**
```{r plot-modularities, eval=FALSE}
# Plot the modularities of clusters
ratio <- clusterModularity(snng, clust, as.ratio = TRUE)
ratio.cc <- clusterModularity(snng.cc, clust.cc, as.ratio = TRUE)

pheatmap(log2(ratio + 1), cluster_cols = FALSE, cluster_rows = FALSE, color=colorRampPalette(c("white", "blue"))(100), main = "Primary data, k = 180")
pheatmap(log2(ratio.cc + 1), cluster_cols = FALSE, cluster_rows = FALSE, color=colorRampPalette(c("white", "blue"))(100), main = "Primary data (cc corrected), k = 180")
```


**Silhouette analysis:**
```{r silhouette-analysis, eval=FALSE}
# Perform silhouette analysis
dist.data <- dist(reducedDim(denoised.sce, "PCA"))
dist.data.cc <- dist(reducedDim(denoised.sce, "PCA_cc"))
sil <- silhouette(clust, dist = dist.data)
sil.cc <- silhouette(clust.cc, dist = dist.data.cc)
plot(sil, main = "Primary data: silhouette plot")
plot(sil, main = "Primary data (cc corrected): silhouette plot")

# Silhouette with subset of cells
denoised.sce.f <- denoised.sce[, seq(1, 21808, by = 80)]
denoised.sce.f.cc <- denoised.sce[, seq(1, 21808, by = 80)]
clust.f <- clust[seq(1, 21808, by = 80)]
clust.f.cc <- clust.cc[seq(1, 21808, by = 80)]
dist.data.f <- dist(reducedDim(denoised.sce.f, "PCA"))
dist.data.f.cc <- dist(reducedDim(denoised.sce.f.cc, "PCA"))
sil.f <- silhouette(clust.f, dist = dist.data.f)
sil.f.cc <- silhouette(clust.f.cc, dist = dist.data.f.cc)

plot(sil.f, main = "Subset of primary data: silhouette plot")
plot(sil.f.cc, main = "Subset of primary data (cc corrected): silhouette plot")

# Plot a heatmap of the differences in clustering between original data and cc-corrected data
tab <- table(clust, clust.cc)
pheatmap(log10(tab+10), main = "Original data vs. CC corrected (k=180)", color=viridis::viridis(100))
```


### 2.3 Primary data: Subclustering and marker gene detection
#### 2.3.1 Differentially expressed genes between clusters in the Cap50 sample

**Uncorrected data**
```{r DE-analysis-of-cap50-sample, eval=FALSE}
# Subset the cap50 sample
cap50 <- denoised.sce[, denoised.sce$sample == "cap50"]

# Plot UMAP plots to inspect clusters
p1 <- plotReducedDim(denoised.sce, colour_by = "cluster", dimred = "UMAP") + ggtitle("Cap 50 sample - sample")
p2 <- plotReducedDim(cap50, colour_by = "cluster", dimred = "UMAP") + ggtitle("Cap 50 sample - cluster")

# Plot only clusters 2 and 4
cap50.f <- cap50[, cap50$cluster == c(2,4)]
p3 <- plotReducedDim(cap50.f, colour_by = "cluster", dimred = "UMAP") + ggtitle("Cap 50 sample - only clusters 2 and 4")
gridExtra::grid.arrange(p1,p2,p3, ncol = 3)


# Find DE genes between samples 2 and 4
cap50.f$cluster <- factor(cap50.f$cluster)
markers <- findMarkers(cap50.f, groups = cap50.f$cluster, lfc = 1, pval.type = "all", direction = "up")
clust.2 <- markers[["2"]]
clust.2.interesting <- clust.2[clust.2$p.value <= 0.05, ]

# Print the list of genes
genes.df <- as.data.frame(clust.2.interesting)
genes.df

# Write genes into a csv file
#write.csv(genes.df, file = "./../saved/cap50_genes.csv")
```

**CC-corrected data**
```{rDE-analysis-of-cap50-sample-cc, eval=FALSE}
# Create cc corrected sce
denoised.sce.cc <- as(altExp(denoised.sce, "CC_corrected"), "SingleCellExperiment")
denoised.sce.cc$cluster.cc <- factor(denoised.sce$cluster.cc)
reducedDim(denoised.sce.cc, "UMAP_cc") <- reducedDim(denoised.sce, "UMAP_cc")

# Subset the cap50 sample
cap50.cc <- denoised.sce.cc[, denoised.sce.cc$sample == "cap50"]

# Plot UMAP plots to inspect clusters
p1 <- plotReducedDim(denoised.sce.cc, colour_by = "cluster.cc", dimred = "UMAP_cc") + ggtitle("Primary data (cc corrected) - cluster")
p2 <- plotReducedDim(cap50.cc, colour_by = "cluster.cc", dimred = "UMAP_cc") + ggtitle("Cap 50 sample (cc corrected) - cluster")

# Plot only clusters 3 and 4
cap50.cc.f <- cap50.cc[, cap50$cluster.cc == c(3,4)]
p3 <- plotReducedDim(cap50.cc.f, colour_by = "cluster.cc", dimred = "UMAP_cc") + ggtitle("Cap 50 sample (cc corrected) - only clusters 3 and 4")
gridExtra::grid.arrange(p1,p2,p3, ncol = 3)


# Find DE genes between clusters 3 and 4
cap50.cc.f$cluster.cc <- factor(cap50.cc.f$cluster.cc)
markers.cc <- findMarkers(cap50.cc.f, groups = cap50.cc.f$cluster.cc, lfc =  1, pval.type = "all", assay.type = "scaled_logcounts")

clust.cc.4 <- markers.cc[["4"]]
clust.cc.4.interesting <- clust.cc.4[clust.cc.4$p.value <= 0.05, ]

# Print the list of genes
genes.cc.df <- as.data.frame(clust.cc.4.interesting)
genes.cc.df

# Write genes into a csv file
#write.csv(genes.cc.df, file = "./../saved/cap50_cc_genes.csv")
```


#### 2.3.2 Subclustering

**Inspect subclusters within the clusters of the primary data**
```{r primary-data-subclustering, eval=FALSE}
# Inspect clusters in each sample
table(denoised.sce$sample, denoised.sce$cluster)
table(denoised.sce.cc$sample, denoised.sce.cc$cluster)

# Perform subclustering
subclust <- quickSubCluster(denoised.sce, groups = denoised.sce$sample, normalize = FALSE)
subclust.cc <- quickSubCluster(denoised.sce.cc, groups = denoised.sce.cc$sample, normalize = FALSE)

# Plot sublusters
p.sub.control <- plotReducedDim(subclust[["control"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data, Control sample - UMAP plot: subclusters")
p.sub.control.cc <- plotReducedDim(subclust.cc[["control"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data (cc corrected), Control sample - UMAP plot: subclusters")

p.sub.cap50 <- plotReducedDim(subclust[["cap50"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data, Cap50 sample - UMAP plot: subclusters")
p.sub.cap50.cc <- plotReducedDim(subclust.cc[["cap50"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data (cc corrected), Cap50 sample - UMAP plot: subclusters")

p.sub.cap50r4h <- plotReducedDim(subclust[["cap50_r4h"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data, R4H sample - UMAP plot: subclusters")
p.sub.cap50r4h.cc <- plotReducedDim(subclust.cc[["cap50_r4h"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data (cc corrected), R4H sample - UMAP plot: subclusters")

p.sub.cap50r8h <- plotReducedDim(subclust[["cap50_r8h"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data, R8H sample - UMAP plot: subclusters")
p.sub.cap50r8h.cc <- plotReducedDim(subclust.cc[["cap50_r8h"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data (cc corrected), R8H sample - UMAP plot: subclusters")

p.sub.cap50r16h <- plotReducedDim(subclust[["cap50_r16h"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data, R16H sample - UMAP plot: subclusters")
p.sub.cap50r16h.cc <- plotReducedDim(subclust.cc[["cap50_r16h"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data (cc corrected), R16H sample - UMAP plot: subclusters")

gridExtra::grid.arrange(p.sub.control, p.sub.control.cc, ncol = 2)
gridExtra::grid.arrange(p.sub.cap50, p.sub.cap50.cc, ncol = 2)
gridExtra::grid.arrange(p.sub.cap50r4h, p.sub.cap50r4h.cc, ncol = 2)
gridExtra::grid.arrange(p.sub.cap50r8h, p.sub.cap50r8h.cc, ncol = 2)
gridExtra::grid.arrange(p.sub.cap50r16h, p.sub.cap50r16h.cc, ncol = 2)

```

#### 2.3.3 Marker gene detection
**This code chunk is not evaluated (as it would produce quite a few plots) but it can be run if one wants to see the marker genes for each sub-cluster** 
```{r marker-gene-detection, eval=FALSE}
# Find marker genes for Control sample
markers.control <- findMarkers(subclust[["control"]], groups = subclust[["control"]]$subcluster, pval.type = "any", lfc = 1)
markers.control.cc <- findMarkers(subclust.cc[["control"]], groups = subclust.cc[["control"]]$subcluster, pval.type = "any", lfc = 1)

Plot marker genes
for(i in 1:length(markers.control)) {
  print(i)
  chosen <- markers.control[[i]][c("Tnnt1", "Tnnt2", "Tnnt3", "Myog", "Tpm1", "Tpm2"), ]
  logFCs <- getMarkerEffects(chosen)
  pheatmap(logFCs, breaks=seq(-5, 5, length.out=101), main = paste("Control Subcluster:", i))
}

for(i in 1:length(markers.control.cc)) {
  print(i)
  chosen <- markers.control.cc[[i]][c("Tnnt1", "Tnnt2", "Tnnt3", "Myog", "Tpm1", "Tpm2"), ]
  logFCs <- getMarkerEffects(chosen)
  pheatmap(logFCs, breaks=seq(-5, 5, length.out=101), main = paste("Control Subcluster (cc):", i))
}

```


### 2.4 Create Cerebro files (.crb)
**This code chunk is used for producing cerebro files and it is by default not evaluated**
```{r cerebro-files, eval=FALSE}
sce.seurat <- as.Seurat(denoised.sce, counts = "counts", data = "logcounts")
sce.cc.seurat <- as.Seurat(denoised.sce.cc, counts = "scaled_logcounts", data = "scaled_logcounts")

exportFromSeurat(object = sce.seurat,
                 file = "./../saved/primary_data.crb", 
                 experiment_name = "Primary data (no correction)",
                 organism = "mm",
                 column_sample = "sample",
                 column_cluster = "cluster",
                 column_nUMI = "sum", 
                 column_cell_cycle_seurat = "Phase",
                 column_nGene = "detected", 
                 add_all_meta_data = TRUE)


exportFromSeurat(object = sce.cc.seurat,
                 file = "./../saved/primary_data_cc_corrected.crb", 
                 experiment_name = "Primary data (cell cycle corrected)",
                 organism = "mm",
                 column_sample = "sample",
                 column_cluster = "cluster.cc",
                 column_nUMI = "sum", 
                 column_cell_cycle_seurat = "Phase",
                 column_nGene = "detected", 
                 add_all_meta_data = TRUE)
```


## 3. Trajectory inference
### 3.1 De Micheli et al. (2020) reference data
```{r deMicheli-traj-inference}
# Create expression and sample objects
expression.deMicheli <- as.matrix(t(logcounts(denoised.sce.deMicheli)))
group_name.deMicheli <- as.factor(denoised.sce.deMicheli$cell_type)
cc.phases.deMicheli <- as.factor(denoised.sce.deMicheli$Phase)

# Trajectory building
set.seed(123)
space.deMicheli <- reduce_dimensionality(expression.deMicheli, dist = "spearman", ndim = 2)
set.seed(123)
traj.deMicheli <- infer_trajectory(space.deMicheli)
traj.deMicheli <- reverse_trajectory(traj.deMicheli)

# Plot trajectory by cell types
draw_trajectory_plot(space = space.deMicheli, 
                     progression_group = group_name.deMicheli,
                     path = traj.deMicheli$path,
                     point_size = 1,
                     point_alpha = 0.5,
                     contour = TRUE) +
  ggtitle("Trajectory inference", subtitle = "De Micheli et al. dataset: All genes (ndim = 2) - Cell types") +
  theme(plot.title = element_text(size = 20, face = "bold"),
        plot.subtitle = element_text(size = 14),
        axis.title = element_text(size = 16))


# Plot trajectory by cc phases
draw_trajectory_plot(space = space.deMicheli, 
                     progression_group = cc.phases.deMicheli,
                     path = traj.deMicheli$path,
                     contour = TRUE) + ggtitle("De Micheli et al. dataset: All genes (ndim = 2) - Cell cycle phases")

# Abundance of some marker genes
pax7 <- expression.deMicheli[, "Pax7"]
cdk1 <- expression.deMicheli[, "Cdk1"]
myod <- expression.deMicheli[, "Myod1"]
myog <- expression.deMicheli[, "Myog"]
btg2 <- expression.deMicheli[, "Btg2"]
cdc20 <- expression.deMicheli[, "Cdc20"]
cdkn1c <- expression.deMicheli[, "Cdkn1c"]

pax7.plot <- draw_trajectory_plot(space = space.deMicheli, 
                                  progression_group = pax7,
                                  path = traj.deMicheli$path,
                                  contour = FALSE) + 
  ggtitle("De Micheli et al. dataset: Pax7 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

cdk1.plot <- draw_trajectory_plot(space = space.deMicheli, 
                                  progression_group = cdk1,
                                  path = traj.deMicheli$path,
                                  contour = FALSE) +
  ggtitle("De Micheli et al. dataset: Cdk1 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))


myog.plot <- draw_trajectory_plot(space = space.deMicheli, 
                                  progression_group = myog,
                                  path = traj.deMicheli$path,
                                  contour = FALSE) + 
  ggtitle("De Micheli et al. dataset: Myog gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

myod.plot <- draw_trajectory_plot(space = space.deMicheli, 
                                  progression_group = myod,
                                  path = traj.deMicheli$path,
                                  contour = FALSE) + 
  ggtitle("De Micheli et al. dataset: Myod gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))


btg2.plot <- draw_trajectory_plot(space = space.deMicheli, 
                                  progression_group = btg2,
                                  path = traj.deMicheli$path,
                                  contour = FALSE) +
  ggtitle("De Micheli et al. dataset: Btg2 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))


cdc20.plot <- draw_trajectory_plot(space = space.deMicheli, 
                                  progression_group = cdc20,
                                  path = traj.deMicheli$path,
                                  contour = FALSE) + 
  ggtitle("De Micheli et al. dataset: Cdc20 gene") + 
  theme(plot.title = element_text(size = 10, face = "bold"))


cdkn1c.plot <- draw_trajectory_plot(space = space.deMicheli, 
                                  progression_group = cdkn1c,
                                  path = traj.deMicheli$path,
                                  contour = FALSE) + 
  ggtitle("De Micheli et al. dataset: Cdkn1c gene") + 
  theme(plot.title = element_text(size = 10, face = "bold"))


gridExtra::grid.arrange(pax7.plot, cdk1.plot, myog.plot, myod.plot, btg2.plot, cdc20.plot, cdkn1c.plot, ncol = 2)

```

**Trajectory inference without cell cycle genes (Seurat cc genes)**
```{r deMicheli-traj-inference-cc-genes}
# Create expression and sample objects
expression.deMicheli.cc.genes <- as.matrix(t(logcounts(denoised.sce.deMicheli.cc.genes)))
group_name.deMicheli.cc.genes <- as.factor(denoised.sce.deMicheli.cc.genes$cell_type)

# Trajectory building
set.seed(123)
space.deMicheli.cc.genes <- reduce_dimensionality(expression.deMicheli.cc.genes, dist = "spearman", ndim = 2)
set.seed(123)
traj.deMicheli.cc.genes <- infer_trajectory(space.deMicheli.cc.genes)

# Plot trajectory by cell types
draw_trajectory_plot(space = space.deMicheli.cc.genes, 
                     progression_group = group_name.deMicheli.cc.genes,
                     path = traj.deMicheli.cc.genes$path,
                     contour = TRUE) + ggtitle("De Micheli et al. dataset: All genes except cell cycle marker genes (from Seurat workflow), ndim = 2")
```

**Trajectory inference without cell cycle genes (OSCA cc genes)**
```{r deMicheli-traj-inference-cc-genes}
# Remove cell cycle genes from the data
denoised.sce.deMicheli.cc.genes <- denoised.sce.deMicheli[!rownames(denoised.sce.deMicheli) %in% cc.genes.osca, ]

# Create expression and sample objects
expression.deMicheli.cc.genes <- as.matrix(t(logcounts(denoised.sce.deMicheli.cc.genes)))
group_name.deMicheli.cc.genes <- as.factor(denoised.sce.deMicheli.cc.genes$cell_type)

# Trajectory building
set.seed(123)
space.deMicheli.cc.genes <- reduce_dimensionality(expression.deMicheli.cc.genes, dist = "spearman", ndim = 2)
set.seed(123)
traj.deMicheli.cc.genes <- infer_trajectory(space.deMicheli.cc.genes)

# Plot trajectory by cell types
draw_trajectory_plot(space = space.deMicheli.cc.genes, 
                     progression_group = group_name.deMicheli.cc.genes,
                     path = traj.deMicheli.cc.genes$path,
                     contour = TRUE) + ggtitle("De Micheli et al. dataset: All genes except cell cycle marker genes (from OSCA workflow), ndim = 2")
```


**Trajectory inference for cell cycle corrected expression values:**
```{r deMicheli-traj-inference-cc}
# Create a sce object from the CC corrected alternative experiment
denoised.sce.deMicheli.cc <- as(altExp(denoised.sce.deMicheli, "CC_corrected"), "SingleCellExperiment")

# Create expression and sample objects
expression.deMicheli.cc <- as.matrix(t(assay(denoised.sce.deMicheli.cc, "scaled_logcounts")))
group_name.deMicheli.cc <- as.factor(denoised.sce.deMicheli.cc$cell_type)
cc.phases.deMicheli <- as.factor(denoised.sce.deMicheli.cc$Phase)

# Trajectory building
set.seed(123)
space.deMicheli.cc <- reduce_dimensionality(expression.deMicheli.cc, dist = "spearman", ndim = 2)
set.seed(123)
traj.deMicheli.cc <- infer_trajectory(space.deMicheli.cc)

draw_trajectory_plot(space = space.deMicheli.cc, 
                     progression_group = group_name.deMicheli.cc,
                     path = traj.deMicheli.cc$path,
                     contour = TRUE) + ggtitle("De Micheli et al. dataset (cc corrected): HVGs (ndim = 2) - Cell types")

draw_trajectory_plot(space = space.deMicheli.cc, 
                     progression_group = cc.phases.deMicheli,
                     path = traj.deMicheli.cc$path,
                     contour = TRUE) + ggtitle("De Micheli et al. dataset (cc corrected): HVGs (ndim = 2) - Cell cycle phases")
```



### 3.2 Dell'Orso et al. (2019) reference data
```{r dellOrso-traj-inference}
# Create expression and sample objects
expression.dellOrso <- as.matrix(t(logcounts(denoised.sce.dellOrso)))
group_name.dellOrso <- as.factor(denoised.sce.dellOrso$cell_type)
cc.phases.dellOrso <- as.factor(denoised.sce.dellOrso$Phase)

# Trajectory inference
set.seed(123)
space.dellOrso <- reduce_dimensionality(expression.dellOrso, dist = "spearman", ndim = 2)
set.seed(123)
traj.dellOrso <- infer_trajectory(space.dellOrso)

# Reverse trajectory
traj.dellOrso <- reverse_trajectory(traj.dellOrso)

# plot by cell type
draw_trajectory_plot(space = space.dellOrso, 
                     progression_group = group_name.dellOrso,
                     path = traj.dellOrso$path,
                     point_size = 1,
                     point_alpha = 0.5,
                     contour = TRUE) +
  ggtitle("Trajectory inference", subtitle = "Dell'Orso et al. dataset: All genes, ndim = 2 - Cell types") +
  theme(plot.title = element_text(size = 20, face = "bold"),
        plot.subtitle = element_text(size = 10),
        axis.title = element_text(size = 16))


# Plot by cell cycle phase
draw_trajectory_plot(space = space.dellOrso, 
                     progression_group = cc.phases.dellOrso,
                     path = traj.dellOrso$path,
                     contour = TRUE) + ggtitle("Dell'Orso et al. dataset: All genes, ndim = 2 - Cell cycle phases")


# Abundance of some marker genes
pax7 <- expression.dellOrso[, "Pax7"]
hes1 <- expression.dellOrso[, "Hes1"]
myf5 <- expression.dellOrso[, "Myf5"]
myod <- expression.dellOrso[, "Myod1"]
ccne1 <- expression.dellOrso[, "Ccne1"]
eifs <- expression.dellOrso[, "Eif1"]
ccnd1 <- expression.dellOrso[, "Ccnd1"]
ccnd2 <- expression.dellOrso[, "Ccnd2"]
myog <- expression.dellOrso[, "Myog"]
tnnt2 <- expression.dellOrso[, "Tnnt2"]
cdkn1a <- expression.dellOrso[, "Cdkn1a"]
cdkn1b <- expression.dellOrso[, "Cdkn1b"]
gas1 <- expression.dellOrso[, "Gas1"]

pax7.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = pax7,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Pax7 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

hes1.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = hes1,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Hes1 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

gas1.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = gas1,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Gas1 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

myf5.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = myf5,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Myf5 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))


myod.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = myod,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Myod gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

ccne1.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = ccne1,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Ccne1 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

eifs.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = eifs,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: eIFs gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

ccnd1.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = ccnd1,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Ccnd1 gene") +
  theme(plot.title = element_text(size = 12, face = "bold"))

ccnd2.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = ccnd2,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Ccnd2 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

myog.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = myog,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Myog gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

tnnt2.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = tnnt2,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Tnnt2 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

cdkn1a.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = cdkn1a,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Cdkn1a gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

cdkn1b.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = cdkn1b,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Cdkn1b gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))


gridExtra::grid.arrange(pax7.plot, hes1.plot,
                        myf5.plot, myod.plot,
                        ccne1.plot, eifs.plot,
                        ccnd1.plot, ccnd2.plot,
                        myog.plot, tnnt2.plot,
                        cdkn1a.plot, cdkn1b.plot)

```


**Trajectory inference without the Primary Myoblast cells**
```{r dellOrso-traj-inference-without-primary-mb}
# Remove PMs from the dellOrso data
denoised.sce.dellOrso.f <- denoised.sce.dellOrso[, denoised.sce.dellOrso$cell_type != "Primary_MB"]

# Create expression and sample objects
expression.dellOrso.f <- as.matrix(t(logcounts(denoised.sce.dellOrso.f)))
group_name.dellOrso.f <- as.factor(denoised.sce.dellOrso.f$cell_type)

# Trajectory inference
set.seed(123)
space.dellOrso.f <- reduce_dimensionality(expression.dellOrso.f, dist = "spearman", ndim = 2)
set.seed(123)
traj.dellOrso.f <- infer_trajectory(space.dellOrso.f)

# plot by cell type
draw_trajectory_plot(space = space.dellOrso.f, 
                     progression_group = group_name.dellOrso.f,
                     path = traj.dellOrso.f$path,
                     contour = TRUE) + ggtitle("Dell'Orso et al. dataset: Only homeostatic and injured 60h MuSCs, All genes, ndim = 2")

```



**Trajectory inference without cell cycle genes**
```{r dellOrso-traj-inference-cc-genes}
# Create expression and sample objects
expression.dellOrso.cc.genes <- as.matrix(t(logcounts(denoised.sce.dellOrso.cc.genes)))
group_name.dellOrso.cc.genes <- as.factor(denoised.sce.dellOrso.cc.genes$cell_type)

# Trajectory inference
set.seed(123)
space.dellOrso.cc.genes <- reduce_dimensionality(expression.dellOrso.cc.genes, dist = "spearman", ndim = 2)
set.seed(123)
traj.dellOrso.cc.genes <- infer_trajectory(space.dellOrso.cc.genes)

# plot by cell type
draw_trajectory_plot(space = space.dellOrso.cc.genes, 
                     progression_group = group_name.dellOrso.cc.genes,
                     path = traj.dellOrso.cc.genes$path,
                     contour = TRUE) + ggtitle("Dell'Orso et al. dataset: All genes except cell cycle marker genes, ndim = 2 - Cell types")
```


**Trajectory inference for cell cycle corrected expression values:**
```{r dellOrso-traj-inference-cc}
# Create a sce object from the CC corrected alternative experiment
denoised.sce.dellOrso.cc <- as(altExp(denoised.sce.dellOrso, "CC_corrected"), "SingleCellExperiment")

# Create expression and sample objects
expression.dellOrso.cc <- as.matrix(t(assay(denoised.sce.dellOrso.cc, "scaled_logcounts")))
group_name.dellOrso.cc <- as.factor(denoised.sce.dellOrso.cc$cell_type)
cc.phases.dellOrso <- as.factor(denoised.sce.dellOrso.cc$Phase)

# Trajectory building
set.seed(123)
space.dellOrso.cc <- reduce_dimensionality(expression.dellOrso.cc, dist = "spearman", ndim = 2)
set.seed(123)
traj.dellOrso.cc <- infer_trajectory(space.dellOrso.cc)

draw_trajectory_plot(space = space.dellOrso.cc, 
                     progression_group = group_name.dellOrso.cc,
                     path = traj.dellOrso.cc$path,
                     contour = TRUE) + ggtitle("Dell'Orso et al. dataset (cc corrected): HVGs (ndim = 2) - Cell types")

draw_trajectory_plot(space = space.dellOrso.cc, 
                     progression_group = cc.phases.dellOrso,
                     path = traj.dellOrso.cc$path,
                     contour = TRUE) + ggtitle("Dell'Orso et al. dataset (cc corrected): HVGs (ndim = 2) - Cell cycle phases")
```



### 3.3 Primary data
**Trajectory inference with uncorrected values**
```{r primary-data-traj-inference}
# Create expression and sample objects
expression.primary <- as.matrix(t(logcounts(denoised.sce)))
group_name.primary <- as.factor(denoised.sce$sample)

# Trajectory inference
set.seed(123)
space.primary <- reduce_dimensionality(expression.primary, dist = "spearman", ndim = 2)
set.seed(123)
traj.primary <- infer_trajectory(space.primary)
traj.primary <- reverse_trajectory(traj.primary)

# Plot the trajectory
draw_trajectory_plot(space = space.primary, 
                     progression_group = group_name.primary,
                     path = traj.primary$path,
                     contour = TRUE) + ggtitle("Primary dataset: All genes (ndim = 2)")

# Plot expression of marker genes
cdh15 <- expression.primary[, "Cdh15"]
tnnt1 <- expression.primary[, "Tnnt1"]
tnnt3 <- expression.primary[, "Tnnt3"]

cdh15.plot <- draw_trajectory_plot(space = space.primary, 
                                  progression_group = cdh15,
                                  point_size = 0.5,
                                  point_alpha = 0.5,
                                  path = traj.primary$path,
                                  contour = FALSE) + 
  ggtitle("Primary data: Cdh15 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

tnnt1.plot <- draw_trajectory_plot(space = space.primary, 
                                  progression_group = tnnt1,
                                  point_size = 0.5,
                                  point_alpha = 0.5,
                                  path = traj.primary$path,
                                  contour = FALSE) + 
  ggtitle("Primary data: Tnnt1 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

tnnt3.plot <- draw_trajectory_plot(space = space.primary, 
                                  progression_group = tnnt3,
                                  point_size = 0.5,
                                  point_alpha = 0.5,
                                  path = traj.primary$path,
                                  contour = FALSE) + 
  ggtitle("Primary data: Tnnt3 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

gridExtra::grid.arrange(cdh15.plot, tnnt1.plot, tnnt3.plot, ncol = 3)


```


**Trajectory inference with cc-corrected expression values**
```{r primary-data-cc-traj-inference}
# Create a sce object from the CC corrected alternative experiment
denoised.sce.cc <- as(altExp(denoised.sce, "CC_corrected"), "SingleCellExperiment")

# Create expression and sample objects
expression.primary.cc <- as.matrix(t(assay(denoised.sce.cc, "scaled_logcounts")))
group_name.primary.cc <- as.factor(denoised.sce.cc$sample)

# Trajectory inference
set.seed(123)
space.primary.cc <- reduce_dimensionality(expression.primary.cc, dist = "spearman", ndim = 2)
set.seed(123)
traj.primary.cc <- infer_trajectory(space.primary.cc)

# Custom colors
colpalette.primary.cc <- colpal <- RColorBrewer::brewer.pal(5, "Set1")
names(colpalette.primary.cc) <- c("cap50", "cap50_r16h", "cap50_r4h", "cap50_r8h", "control")
colpalette.primary.cc["cap50_r8h"] <- "#FFFF33" # set purple -> yellow


# Plot the trajectory
draw_trajectory_plot(space = space.primary.cc, 
                     progression_group = group_name.primary.cc,
                     progression_group_palette = colpalette.primary.cc,
                     path = traj.primary.cc$path,
                     point_size = 1,
                     point_alpha = 0.4,
                     contour = TRUE) +
  ggtitle("Trajectory inference", subtitle = "Primary dataset (cc corrected): HVGs (ndim = 2)") +
  theme(plot.title = element_text(size = 20, face = "bold"),
        plot.subtitle = element_text(size = 16),
        legend.text = element_text(size = 16),
        legend.title = element_text(size = 20, face = "bold"),
        axis.title = element_text(size = 16))

# Marker genes
cdh15 <- expression.primary.cc[, "Cdh15"]
tnnt1 <- expression.primary.cc[, "Tnnt1"]
tnnt2 <- expression.primary.cc[, "Tnnt2"]
tnnt3 <- expression.primary.cc[, "Tnnt3"]
pax7 <- expression.primary.cc[, "Pax7"]
hes1 <- expression.primary.cc[, "Hes1"]
myf5 <- expression.primary.cc[, "Myf5"]
myod <- expression.primary.cc[, "Myod1"]
ccne1 <- expression.primary.cc[, "Ccne1"]
eifs <- expression.primary.cc[, "Eif1"]
ccnd1 <- expression.primary.cc[, "Ccnd1"]
ccnd2 <- expression.primary.cc[, "Ccnd2"]
myog <- expression.primary.cc[, "Myog"]
cdkn1a <- expression.primary.cc[, "Cdkn1a"]
gas1 <- expression.primary.cc[, "Gas1"]

prog.groups <- list(cdh15 = cdh15,
                    tnnt1 = tnnt1,
                    tnnt2 = tnnt2,
                    tnnt3 = tnnt3,
                    pax7 = pax7,
                    hes1 = hes1,
                    myf5 = myf5,
                    myod = myod,
                    ccne1 = ccne1,
                    eifs = eifs,
                    ccnd1 = ccnd1,
                    ccnd2 = ccnd2,
                    myog = myog,
                    cdkn1a = cdkn1a,
                    gas1 = gas1)

marker.gene.plots <- list()
for(gene in names(prog.groups)) {
  p <- draw_trajectory_plot(space = space.primary.cc, 
                            progression_group = prog.groups[[gene]],
                            progression_group_palette = colpalette.primary.cc.markers,
                            point_size = 0.5,
                            point_alpha = 0.5,
                            path = traj.primary.cc$path,
                            contour = FALSE) + 
  ggtitle(paste(gene, "gene")) +
  theme(plot.title = element_text(size = 14, face = "bold"), axis.title = element_text(size = 10))
  marker.gene.plots[[gene]] <- p
}

# Plot the marker genes
gridExtra::grid.arrange(marker.gene.plots$cdh15, marker.gene.plots$tnnt1, marker.gene.plots$tnnt2, marker.gene.plots$tnnt3, marker.gene.plots$pax7, 
                        marker.gene.plots$hes1, marker.gene.plots$myf5, marker.gene.plots$myod, marker.gene.plots$ccne1, marker.gene.plots$eifs,
                        marker.gene.plots$ccnd1, marker.gene.plots$ccnd2, marker.gene.plots$myog, marker.gene.plots$cdkn1a, marker.gene.plots$gas1,
                        ncol = 5)

``` 


**Trajectory inference with cc genes removed**
```{r primary-data-cc-genes-traj-inference}
# Create expression and sample objects
expression.primary.cc.genes <- as.matrix(t(logcounts(denoised.sce.cc.genes)))
group_name.primary.cc.genes <- as.factor(denoised.sce.cc.genes$sample)

# Trajectory inference
set.seed(123)
space.primary.cc.genes <- reduce_dimensionality(expression.primary.cc.genes, dist = "spearman", ndim = 2)
set.seed(123)
traj.primary.cc.genes <- infer_trajectory(space.primary.cc.genes)
#saveRDS(space.primary.cc.genes, file = "./../saved/space_primary_cc_genes")


# Plot the trajectory
draw_trajectory_plot(space = space.primary.cc.genes, 
                     progression_group = group_name.primary.cc.genes,
                     path = traj.primary.cc.genes$path,
                     contour = TRUE) + ggtitle("Primary dataset: All genes except cell cycle marker genes (ndim = 2)")
```


**Trajectory inference with (cc corrected) with only control, cap50 and cap50_r16h samples**
```{r primary-data-cc-traj-inference-subset-of-samples}
# Create a sce object from the CC corrected alternative experiment
denoised.sce.cc <- as(altExp(denoised.sce, "CC_corrected"), "SingleCellExperiment")
denoised.sce.cc.f <- denoised.sce.cc[, denoised.sce.cc$sample %in% c("control", "cap50", "cap50_r16h")]

# Create expression and sample objects
expression.primary.cc.f <- as.matrix(t(assay(denoised.sce.cc.f, "scaled_logcounts")))
group_name.primary.cc.f <- as.factor(denoised.sce.cc.f$sample)

# Trajectory inference
set.seed(123)
space.primary.cc.f <- reduce_dimensionality(expression.primary.cc.f, dist = "spearman", ndim = 2)
set.seed(123)
traj.primary.cc.f <- infer_trajectory(space.primary.cc.f)

# Plot the trajectory
draw_trajectory_plot(space = space.primary.cc.f, 
                     progression_group = group_name.primary.cc.f,
                     path = traj.primary.cc.f$path,
                     contour = TRUE) + ggtitle("Filtered primary dataset (cc corrected): HVGs (ndim = 2)")


# Plot expression of marker genes
cdh15 <- expression.primary.cc.f[, "Cdh15"]
tnnt1 <- expression.primary.cc.f[, "Tnnt1"]
tnnt3 <- expression.primary.cc.f[, "Tnnt3"]

cdh15.plot <- draw_trajectory_plot(space = space.primary.cc.f, 
                                  progression_group = cdh15,
                                  path = traj.primary.cc.f$path,
                                  contour = FALSE) + 
  ggtitle("Filtered primary dataset (cc corrected): Cdh15 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

tnnt1.plot <- draw_trajectory_plot(space = space.primary.cc.f, 
                                  progression_group = tnnt1,
                                  path = traj.primary.cc.f$path,
                                  contour = FALSE) + 
  ggtitle("Filtered primary dataset (cc corrected): Tnnt1 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

tnnt3.plot <- draw_trajectory_plot(space = space.primary.cc.f, 
                                  progression_group = tnnt3,
                                  path = traj.primary.cc.f$path,
                                  contour = FALSE) + 
  ggtitle("Filtered primary dataset (cc corrected): Tnnt3 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

gridExtra::grid.arrange(cdh15.plot, tnnt1.plot, tnnt3.plot, ncol = 2)


```


**Trajectory inference with (cc genes removed) with only control, cap50 and cap50_r16h samples**
```{r primary-data-cc-genes-traj-inference-subset-of-samples}
# Create expression and sample objects
denoised.sce.cc.genes.f <- denoised.sce.cc.genes[, denoised.sce.cc.genes$sample %in% c("control", "cap50", "cap50_r16h")]
expression.primary.cc.genes.f <- as.matrix(t(logcounts(denoised.sce.cc.genes.f)))
group_name.primary.cc.genes.f <- as.factor(denoised.sce.cc.genes.f$sample)

# Trajectory inference
set.seed(123)
space.primary.cc.genes.f <- reduce_dimensionality(expression.primary.cc.genes.f, dist = "spearman", ndim = 2)
set.seed(123)
traj.primary.cc.genes.f <- infer_trajectory(space.primary.cc.genes.f)
#saveRDS(space.primary.cc.genes.f, file = "./../saved/space_primary_cc_genes.f")

# Plot the trajectory
draw_trajectory_plot(space = space.primary.cc.genes.f, 
                     progression_group = group_name.primary.cc.genes.f,
                     path = traj.primary.cc.genes.f$path,
                     contour = TRUE) + ggtitle("Filtered primary dataset: All genes except cell cycle marker genes (ndim = 2)")


# Plot expression of marker genes
cdh15 <- expression.primary.cc.genes.f[, "Cdh15"]
tnnt1 <- expression.primary.cc.genes.f[, "Tnnt1"]
tnnt3 <- expression.primary.cc.genes.f[, "Tnnt3"]

cdh15.plot <- draw_trajectory_plot(space = space.primary.cc.genes.f, 
                                  progression_group = cdh15,
                                  path = traj.primary.cc.genes.f$path,
                                  contour = FALSE) + 
  ggtitle("Filtered primary dataset (cc genes excluded): Cdh15 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

tnnt1.plot <- draw_trajectory_plot(space = space.primary.cc.genes.f, 
                                  progression_group = tnnt1,
                                  path = traj.primary.cc.genes.f$path,
                                  contour = FALSE) + 
  ggtitle("Filtered primary dataset (cc genes excluded): Tnnt1 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

tnnt3.plot <- draw_trajectory_plot(space = space.primary.cc.genes.f, 
                                  progression_group = tnnt3,
                                  path = traj.primary.cc.genes.f$path,
                                  contour = FALSE) + 
  ggtitle("Filtered primary dataset (cc genes excluded): Tnnt3 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

gridExtra::grid.arrange(cdh15.plot, tnnt1.plot, tnnt3.plot, ncol = 2)

```



## 4. Identifying candidate marker genes
**Shortcut for fetching the gimps & modules generated by Puhti supercomputer**
```{r shortcut-gimp-modules}
gimp.deMicheli <- readRDS("./../saved/gimp_deMicheli")
gimp.dellOrso <- readRDS("./../saved/gimp_dellOrso")
modules.deMicheli <- readRDS("./../saved/modules_deMicheli")
modules.dellOrso <- readRDS("./../saved/modules_dellOrso")

# Generate objects for de Micheli dataset
gimp.deMicheli$qvalue <- p.adjust(p = gimp.deMicheli$pvalue, method = "BH")
gene_sel.deMicheli <- gimp.deMicheli$gene[gimp.deMicheli$qvalue < .05]
expr_sel.deMicheli <- scale_quantile(expression.deMicheli[, gene_sel.deMicheli])

# Generate objects for dellOrso dataset
gimp.dellOrso$qvalue <- p.adjust(gimp.dellOrso$pvalue, "BH", length(gimp.dellOrso$pvalue))
gene_sel.dellOrso <- gimp.dellOrso$gene[gimp.dellOrso$qvalue < .05]
expr_sel.dellOrso <- scale_quantile(expression.dellOrso[, gene_sel.dellOrso])
```

**Generate csv files from the important genes**
```{r modules-as-csv}
write.csv(modules.dellOrso, file = "./../saved/modules_dellOrso_excel.csv")
write.csv(modules.deMicheli, file = "./../saved/modules_deMicheli_excel.csv")

```

### 4.1 De Micheli et al. (2020) reference data
```{r marker-genes-deMicheli, eval=FALSE}
# Identify candidate marker genes
gimp.deMicheli <- gene_importances(x = expression.deMicheli,
                                   time = traj.deMicheli$time,
                                   num_permutations = 10,
                                   num_threads = 8)

gimp.deMicheli$qvalue <- p.adjust(p = gimp.deMicheli$pvalue, method = "BH")
gene_sel.deMicheli <- gimp.deMicheli$gene[gimp.deMicheli$qvalue < .05]
expr_sel.deMicheli <- scale_quantile(expression.deMicheli[, gene_sel.deMicheli])

# Group genes to modules
modules.deMicheli <- extract_modules(expr_sel.deMicheli,
                                     time = traj.deMicheli$time,
                                     verbose = FALSE)

#saveRDS(modules.deMicheli, file = "./../saved/modules_deMicheli")
#saveRDS(gimp.deMicheli, file = "./../saved/gimp_deMicheli")

```


**Visualization of gene importances and modules**
```{r DeMicheli-gimp-modules-visualization, eval=FALSE}
# Visualize expression of selected genes
pdf("./../saved/gimp_heatmap_deMicheli.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli, 
                        time = traj.deMicheli$time,
                        progression_group = group_name.deMicheli,
                        show_labels_row = TRUE,
                        fontsize_row = 2)
dev.off()

# Visualize modules
pdf("./../saved/modules_heatmap_deMicheli.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli, 
                        time = traj.deMicheli$time,
                        progression_group = group_name.deMicheli,
                        modules = modules.deMicheli,
                        show_labels_row = TRUE,
                        fontsize_row = 3)
dev.off()
```


### 4.2 Dell'Orso et al. (2019) reference data
```{r marker-genes-dellOrso, eval=FALSE}
# Identify candidate marker genes
gimp.dellOrso <- gene_importances(x = expression.dellOrso,
                                  time = traj.dellOrso$time,
                                  num_permutations = 10,
                                  num_threads = 8)

gimp.dellOrso$qvalue <- p.adjust(gimp.dellOrso$pvalue, "BH", length(gimp.dellOrso$pvalue))
gene_sel.dellOrso <- gimp.dellOrso$gene[gimp.dellOrso$qvalue < .05]
expr_sel.dellOrso <- scale_quantile(expression.dellOrso[, gene_sel.dellOrso])

# Group genes to modules
modules.dellOrso <- extract_modules(expr_sel.dellOrso,
                                     time = traj.dellOrso$time,
                                     verbose = FALSE)

#saveRDS(modules.dellOrso, file = "./../saved/modules_dellOrso")
#saveRDS(gimp.dellOrso, file = "./../saved/gimp_dellOrso")
```


**Visualization of gene importances and modules**
```{r DellOrso-gimp-modules-visualization, eval=FALSE}
# Visualize expression of selected genes
pdf("./../saved/gimp_heatmap_dellOrso.pdf")
draw_trajectory_heatmap(expr_sel.dellOrso, 
                        time = traj.dellOrso$time,
                        progression_group = group_name.dellOrso,
                        show_labels_row = TRUE,
                        fontsize_row = 1)
dev.off()

# Visualize modules
pdf("./../saved/modules_heatmap_dellOrso.pdf")
draw_trajectory_heatmap(expr_sel.dellOrso, 
                        time = traj.dellOrso$time,
                        progression_group = group_name.dellOrso,
                        modules = modules.dellOrso,
                        show_labels_row = TRUE,
                        fontsize_row = 1)
dev.off()
```


#### 4.2.1 - PMs removed
```{r shortcut-marker-genes-dellOrso-without-PMs}
# Load gimp and modules objects
gimp.dellOrso.f <- readRDS("./../saved/gimp_dellOrso_f")
gimp.dellOrso.f$qvalue <- p.adjust(gimp.dellOrso.f$pvalue, "BH", length(gimp.dellOrso.f$pvalue))
gene_sel.dellOrso.f <- gimp.dellOrso.f$gene[gimp.dellOrso.f$qvalue < .05]
expr_sel.dellOrso.f <- scale_quantile(expression.dellOrso.f[, gene_sel.dellOrso.f])

modules.dellOrso.f <- readRDS("./../saved/modules_dellOrso_f")

```

```{r marker-genes-dellOrso-without-PMs, eval=FALSE}
# Identify candidate marker genes
gimp.dellOrso.f <- gene_importances(x = expression.dellOrso.f,
                                    time = traj.dellOrso.f$time,
                                    num_permutations = 10,
                                    num_threads = 8)

gimp.dellOrso.f$qvalue <- p.adjust(gimp.dellOrso.f$pvalue, "BH", length(gimp.dellOrso.f$pvalue))
gene_sel.dellOrso.f <- gimp.dellOrso.f$gene[gimp.dellOrso.f$qvalue < .05]
expr_sel.dellOrso.f <- scale_quantile(expression.dellOrso.f[, gene_sel.dellOrso.f])

# Group genes to modules
modules.dellOrso.f <- extract_modules(expr_sel.dellOrso.f,
                                      time = traj.dellOrso.f$time,
                                      verbose = FALSE)

#saveRDS(modules.dellOrso.f, file = "./../saved/modules_dellOrso_f")
#saveRDS(gimp.dellOrso.f, file = "./../saved/gimp_dellOrso_f")

```

**Visualization of gene importances and modules**
```{r DellOrso-without-PMs-gimp-modules-visualization, eval=FALSE}
# Visualize expression of selected genes
pdf("./../saved/gimp_heatmap_dellOrso_f.pdf")
draw_trajectory_heatmap(expr_sel.dellOrso.f, 
                        time = traj.dellOrso.f$time,
                        progression_group = group_name.dellOrso.f,
                        show_labels_row = TRUE,
                        fontsize_row = 1)
dev.off()

# Visualize modules
pdf("./../saved/modules_heatmap_dellOrso_f.pdf")
draw_trajectory_heatmap(expr_sel.dellOrso.f, 
                        time = traj.dellOrso.f$time,
                        progression_group = group_name.dellOrso.f,
                        modules = modules.dellOrso.f,
                        show_labels_row = TRUE,
                        fontsize_row = 1)
dev.off()
```


### 4.2 Primary data

```{r shortcut-primary-data-gimp-modules}
# Load gimp and modules objects
gimp <- readRDS("./../saved/gimp")
gimp$qvalue <- p.adjust(p = gimp$pvalue, method = "BH")
gene_sel <- gimp$gene[gimp$qvalue < .05]
expr_sel <- scale_quantile(expression.primary[, gene_sel])

modules <- readRDS("./../saved/modules")
```


```{r primary-data-gimp-modules, eval=FALSE}
# Identify candidate marker genes
gimp <- gene_importances(x = expression.primary,
                         time = traj.primary$time,
                         num_permutations = 10,
                         num_threads = 8)

gimp$qvalue <- p.adjust(p = gimp$pvalue, method = "BH")
gene_sel <- gimp$gene[gimp$qvalue < .05]
expr_sel <- scale_quantile(expression.primary[, gene_sel])

# Group genes to modules
modules <- extract_modules(expr_sel,
                           time = traj.primary$time,
                           verbose = FALSE)

saveRDS(modules, file = "./../saved/modules")
saveRDS(gimp, file = "./../saved/gimp")
```

**Visualization of gene importances and modules**
```{r primary-data-gimp-modules-visualization, eval=FALSE}
# Visualize expression of selected genes
pdf("./../saved/gimp_heatmap_primary_data.pdf")
draw_trajectory_heatmap(expr_sel, 
                        time = traj.primary$time,
                        progression_group = group_name.primary,
                        show_labels_row = TRUE,
                        fontsize_row = 1)
dev.off()

# Visualize modules
pdf("./../saved/modules_heatmap_primary_data.pdf")
draw_trajectory_heatmap(expr_sel, 
                        time = traj.primary$time,
                        progression_group = group_name.primary,
                        modules = modules,
                        show_labels_row = TRUE,
                        fontsize_row = 1)
dev.off()
```



## 5. Mapping of primary data to reference datasets
### 5.1 Assign reference data cell types to primary data and inspect results

**Predict cell types from reference datasets (shortcut)**
```{r predict-cell-types-shortcut}
# Load pre-saved prediction objects
pred.deMicheli <- readRDS("./../saved/pred_deMicheli")
pred.dellOrso <- readRDS("./../saved/pred_dellOrso")

# Assign cell types to primary data
denoised.sce$cell_type_deMicheli <- pred.deMicheli$pruned.labels
denoised.sce$cell_type_dellOrso <- pred.dellOrso$pruned.labels

```

**Predict cell types from reference datasets**
```{r predict-cell-types, eval=FALSE}
## De Micheli
pred.deMicheli <- SingleR(test = denoised.sce,
                          assay.type.test = "logcounts",
                          ref = denoised.sce.deMicheli,
                          labels = denoised.sce.deMicheli$cell_type)

denoised.sce$cell_type_deMicheli <- pred.deMicheli$pruned.labels
#saveRDS(pred.deMicheli, file = "./../saved/pred_deMicheli")

## Dell'Orso
pred.dellOrso <- SingleR(test = denoised.sce,
                         assay.type.test = "logcounts",
                         ref = denoised.sce.dellOrso,
                         labels = denoised.sce.dellOrso$cell_type)

denoised.sce$cell_type_dellOrso <- pred.dellOrso$pruned.labels
#saveRDS(pred.dellOrso, file = "./../saved/pred_dellOrso")

```

**Inspect the cell mappings**
```{r cell-mapping-inspection}
# Inspection of predictions
table(pred.deMicheli$pruned.labels)
table(pred.dellOrso$labels)
tab.deMicheli <- t(table(pred.deMicheli$pruned.labels, denoised.sce$sample)[, c("control", "cap50", "cap50_r4h","cap50_r8h","cap50_r16h")])
tab.dellOrso <- t(table(pred.dellOrso$pruned.labels, denoised.sce$sample)[, c("control", "cap50", "cap50_r4h","cap50_r8h","cap50_r16h")])
tab.deMicheli
tab.dellOrso

## 'melt' tables
df.deMicheli <- melt(t(tab.deMicheli))
colnames(df.deMicheli) <- c("cell_type", "sample", "count")
df.dellOrso <- melt(t(tab.dellOrso))
colnames(df.dellOrso) <- c("cell_type", "sample", "count")

# Add missing cell types
df.deMicheli.missing.rows <- data.frame("cell_type" = c(rep("d0_MuSCs_and_progenitors", 5), rep("d0_mature_skeletal_muscle", 5), rep("d7_mature_skeletal_muscle", 5)),
                                        "sample" = rep(c("control", "cap50", "cap50_r4h", "cap50_r8h", "cap50_r16h"), 3),
                                        "count" = rep(0, 15))
df.dellOrso.missing.rows <- data.frame("cell_type" = rep("homeostatic_MuSCs", 5),
                                       "sample" = c("control", "cap50", "cap50_r4h", "cap50_r8h", "cap50_r16h"),
                                       "count" = rep(0, 5))

df.deMicheli <- rbind(df.deMicheli, df.deMicheli.missing.rows)
df.dellOrso <- rbind(df.dellOrso, df.dellOrso.missing.rows)

# De Micheli balloon plot
balloon.deMicheli <- ggballoonplot(df.deMicheli, x = "sample", y = "cell_type", size = "count",
              fill = "count", ggtheme = theme_bw()) +
  scale_fill_viridis_c(option = "C") +
  ggtitle("De Micheli et al. reference dataset")

# Dell Orso balloon plot
balloon.dellOrso <- ggballoonplot(df.dellOrso, x = "sample", y = "cell_type", size = "count",
              fill = "count", ggtheme = theme_bw()) +
  scale_fill_viridis_c(option = "C") +
  ggtitle("Dell'Orso et al. reference dataset")

gridExtra::grid.arrange(balloon.deMicheli, balloon.dellOrso, ncol = 2,      top = textGrob("Mapping of primary data cells to reference datasest shown in a balloon plot",gp=gpar(fontsize=20)))
```


**DeMicheli diagnostic plots**
```{r deMicheli-annotation-diagnostic-plots, eval=FALSE}
# Visualize the score matrix
pdf("./../saved/deMicheli_score_heatmap.pdf", width = 30, height = 50)
plotScoreHeatmap(pred.deMicheli, 
                 annotation_col = as.data.frame(colData(denoised.sce)[, "sample", drop = FALSE]), width = 10)
dev.off()

to.remove <- pruneScores(pred.deMicheli, min.diff.next=0.1)
table(Label=pred.deMicheli$labels, Removed=to.remove)


# Delta distribution
pdf("./../saved/deMicheli_score_distribution.pdf", width = 20, height = 15)
plotScoreDistribution(results = pred.deMicheli, show = "delta.med")
dev.off()

# Get marker genes
all.markers.deMicheli <- metadata(pred.deMicheli)$de.genes

# Write markers for each sample into a csv file
all.markers.deMicheli.df <- data.table("d0_mature_skeletal_muscle" = unique(unlist(all.markers.deMicheli$d0_mature_skeletal_muscle)),
                                       "d0_MuSCs_and_progenitors" = unique(unlist(all.markers.deMicheli$d0_MuSCs_and_progenitors)),
                                       "d5_mature_skeletal_muscle" = unique(unlist(all.markers.deMicheli$d5_mature_skeletal_muscle)),
                                       "d5_MuSCs_and_progenitors" = unique(unlist(all.markers.deMicheli$d5_MuSCs_and_progenitors)),
                                       "d7_mature_skeletal_muscle" = unique(unlist(all.markers.deMicheli$d7_mature_skeletal_muscle)),
                                       "d7_MuSCs_and_progenitors" = unique(unlist(all.markers.deMicheli$d7_MuSCs_and_progenitors)))
write.csv(all.markers.deMicheli.df, file ="./../saved/all_markers_deMicheli.csv")

# Examine the expression of the marker genes for each label in the test dataset
pdf("./../saved/heatmap_deMicheli_d0_muscle.pdf")
plotHeatmap(denoised.sce,
            order_columns_by="cell_type_deMicheli",
            features=unique(unlist(all.markers.deMicheli$d0_mature_skeletal_muscle)),
            fontsize_row = 2)
dev.off()
pdf("./../saved/heatmap_deMicheli_d0_musc.pdf")
plotHeatmap(denoised.sce,
            order_columns_by="cell_type_deMicheli",
            features=unique(unlist(all.markers.deMicheli$d0_MuSCs_and_progenitors)),
            fontsize_row = 2)
dev.off()
pdf("./../saved/heatmap_deMicheli_d5_muscle.pdf")
plotHeatmap(denoised.sce,
            order_columns_by="cell_type_deMicheli",
            features=unique(unlist(all.markers.deMicheli$d5_mature_skeletal_muscle)),
            fontsize_row = 2)
dev.off()
pdf("./../saved/heatmap_deMicheli_d5_musc.pdf")
plotHeatmap(denoised.sce,
            order_columns_by="cell_type_deMicheli",
            features=unique(unlist(all.markers.deMicheli$d5_MuSCs_and_progenitors)),
            fontsize_row = 2)
dev.off()
pdf("./../saved/heatmap_deMicheli_d7_muscle.pdf")
plotHeatmap(denoised.sce,
            order_columns_by="cell_type_deMicheli",
            features=unique(unlist(all.markers.deMicheli$d7_mature_skeletal_muscle)),
            fontsize_row = 2)
dev.off()
pdf("./../saved/heatmap_deMicheli_d7_musc.pdf")
plotHeatmap(denoised.sce,
            order_columns_by="cell_type_deMicheli",
            features=unique(unlist(all.markers.deMicheli$d7_MuSCs_and_progenitors)),
            fontsize_row = 2)
dev.off()
```

**Dell'Orso diagnostic plots**
```{r dellOrso-annotation-diagnostic-plots, eval=FALSE}
# Visualize the score matrix
pdf("./../saved/dellOrso_score_heatmap.pdf")
plotScoreHeatmap(pred.dellOrso, 
                 annotation_col = as.data.frame(colData(denoised.sce)[, "sample", drop = FALSE]))
dev.off()

# Delta distribution
pdf("./../saved/dellOrso_score_distribution.pdf")
plotScoreDistribution(pred.dellOrso, show = "delta.med")
dev.off()

# Get marker genes
all.markers.dellOrso <- metadata(pred.dellOrso)$de.genes

# Write markers for each sample into a csv file
all.markers.dellOrso.df <- data.table("Homeostatic_MuSCs" = unique(unlist(all.markers.dellOrso$homeostatic_MuSCs)),
                                      "inj_60h_MuSCs" = unique(unlist(all.markers.dellOrso$inj_60h_MuSCs)),
                                      "Primary_MB" = unique(unlist(all.markers.dellOrso$Primary_MB)))

write.csv(all.markers.dellOrso.df, file = "./../saved/all_markers_dellOrso.csv")

# Examine the expression of the marker genes for each label in the test dataset
pdf("./../saved/heatmap_dellOrso_homeostatic_muscs.pdf")
plotHeatmap(denoised.sce, order_columns_by="cell_type_dellOrso",
    features=unique(unlist(all.markers.dellOrso.df$homeostatic_MuSCs)),
    fontsize_row = 2)
dev.off()

pdf("./../saved/heatmap_dellOrso_inj_60h_muscs.pdf")
plotHeatmap(denoised.sce, order_columns_by="cell_type_dellOrso",
    features=unique(unlist(all.markers.dellOrso.df$inj_60h_MuSCs)),
    fontsize_row = 2)
dev.off()

pdf("./../saved/heatmap_dellOrso_primary_mb.pdf")
plotHeatmap(denoised.sce, order_columns_by="cell_type_dellOrso",
    features=unique(unlist(all.markers.dellOrso.df$Primary_MB)),
    fontsize_row = 2)
dev.off()
```


### 5.2 Cell mapping of reference datasets with cc-genes removed

**Shortcut for predicting the cell types from cc-gene excluded data**
```{r shortcut-predict-cell-types-cc-genes}
pred.deMicheli.cc.genes <- readRDS("./../saved/pred_deMicheli_cc_genes")
pred.dellOrso.cc.genes <- readRDS("./../saved/pred_dellOrso_cc_genes")
```


**Predict the cell types from data where cc genes have been removed**
```{r predict-cell-types-cc-genes, eval=FALSE}
# De Micheli
pred.deMicheli.cc.genes <- SingleR(test = denoised.sce.cc.genes,
                                   assay.type.test = "logcounts",
                                   ref = denoised.sce.deMicheli.cc.genes,
                                   labels = denoised.sce.deMicheli.cc.genes$cell_type)

#saveRDS(pred.deMicheli.cc.genes, file = "./../saved/pred_deMicheli_cc_genes")

## Dell'Orso
pred.dellOrso.cc.genes <- SingleR(test = denoised.sce.cc.genes,
                                  assay.type.test = "logcounts",
                                  ref = denoised.sce.dellOrso.cc.genes,
                                  labels = denoised.sce.dellOrso.cc.genes$cell_type)

#saveRDS(pred.dellOrso.cc.genes, file = "./../saved/pred_dellOrso_cc_genes")

```

**Inspect the cell mappings**
```{r inspect-cell-types-cc-genes}
table(pred.deMicheli.cc.genes$pruned.labels)
table(pred.dellOrso.cc.genes$labels)

tab.deMicheli.cc.genes <- t(table(pred.deMicheli.cc.genes$pruned.labels, denoised.sce.cc.genes$sample)[, c("control", "cap50", "cap50_r4h","cap50_r8h","cap50_r16h")])

tab.dellOrso.cc.genes <- t(table(pred.dellOrso.cc.genes$pruned.labels, denoised.sce.cc.genes$sample)[, c("control", "cap50", "cap50_r4h","cap50_r8h","cap50_r16h")])

tab.deMicheli.cc.genes
tab.dellOrso.cc.genes
```



#### 5.3 Assign cell types from reference to another reference dataset
**Shortcut for loading the prediction objects**
```{r cell-type-reference-datasets-shortcut}
# Load the pre-saved reference data cell annotation predictions
pred.reference <- readRDS("./../saved/pred_reference")
pred.reference2 <- readRDS("./../saved/pred_reference2")
```


```{r cell-type-reference-datasets, eval=FALSE}
# Dell'Orso as reference
pred.reference <- SingleR(test = denoised.sce.deMicheli,
                          assay.type.test = "logcounts",
                          ref = denoised.sce.dellOrso,
                          labels = denoised.sce.dellOrso$cell_type)
#saveRDS(pred.reference, file = "./../saved/pred_reference")

# De Micheli as reference
pred.reference2 <- SingleR(test = denoised.sce.dellOrso,
                          assay.type.test = "logcounts",
                          ref = denoised.sce.deMicheli,
                          labels = denoised.sce.deMicheli$cell_type)
#saveRDS(pred.reference2, file = "./../saved/pred_reference2")

```


```{r cell-type-reference-datasets-inspection}
# Dell'Orso as reference:
table(pred.reference$pruned.labels)
t(table(pred.reference$pruned.labels, denoised.sce.deMicheli$cell_type))

# DeMicheli as reference
table(pred.reference2$pruned.labels)
table(pred.reference2$pruned.labels, denoised.sce.dellOrso$cell_type)
```


## 6. Refined cell mapping based on smaller segments of the trajectory

### 6.1 Trajectory building and mapping of cells
**Using all cells from De Micheli dataset**
```{r deMicheli-traj-inference-segments}
# Trajectory building
expression.deMicheli <- as.matrix(t(logcounts(denoised.sce.deMicheli)))
set.seed(123)
space.deMicheli <- reduce_dimensionality(expression.deMicheli, dist = "spearman", ndim = 2)
set.seed(123)
traj.deMicheli <- infer_trajectory(space.deMicheli)

# Sort cells based on their progression
ordered <- sort(traj.deMicheli$time, decreasing = FALSE)

# Split cells into 10 segments based on their pseudo time value
n = 10
ordered_chunks = chunk(ordered, n.chunks = 10, shuffle = FALSE)
segment.1 <- names(ordered_chunks[[1]])
segment.2 <- names(ordered_chunks[[2]])
segment.3 <- names(ordered_chunks[[3]])
segment.4 <- names(ordered_chunks[[4]])
segment.5 <- names(ordered_chunks[[5]])
segment.6 <- names(ordered_chunks[[6]])
segment.7 <- names(ordered_chunks[[7]])
segment.8 <- names(ordered_chunks[[8]])
segment.9 <- names(ordered_chunks[[9]])
segment.10 <- names(ordered_chunks[[10]])

# Label cells by their segment
denoised.sce.deMicheli.s1 <- colnames(denoised.sce.deMicheli) %in% segment.1
denoised.sce.deMicheli.s2 <- colnames(denoised.sce.deMicheli) %in% segment.2
denoised.sce.deMicheli.s3 <- colnames(denoised.sce.deMicheli) %in% segment.3
denoised.sce.deMicheli.s4 <- colnames(denoised.sce.deMicheli) %in% segment.4
denoised.sce.deMicheli.s5 <- colnames(denoised.sce.deMicheli) %in% segment.5
denoised.sce.deMicheli.s6 <- colnames(denoised.sce.deMicheli) %in% segment.6
denoised.sce.deMicheli.s7 <- colnames(denoised.sce.deMicheli) %in% segment.7
denoised.sce.deMicheli.s8 <- colnames(denoised.sce.deMicheli) %in% segment.8
denoised.sce.deMicheli.s9 <- colnames(denoised.sce.deMicheli) %in% segment.9
denoised.sce.deMicheli.s10 <- colnames(denoised.sce.deMicheli) %in% segment.10

denoised.sce.deMicheli$segment[denoised.sce.deMicheli.s1] <- "segment_1"
denoised.sce.deMicheli$segment[denoised.sce.deMicheli.s2] <- "segment_2"
denoised.sce.deMicheli$segment[denoised.sce.deMicheli.s3] <- "segment_3"
denoised.sce.deMicheli$segment[denoised.sce.deMicheli.s4] <- "segment_4"
denoised.sce.deMicheli$segment[denoised.sce.deMicheli.s5] <- "segment_5"
denoised.sce.deMicheli$segment[denoised.sce.deMicheli.s6] <- "segment_6"
denoised.sce.deMicheli$segment[denoised.sce.deMicheli.s7] <- "segment_7"
denoised.sce.deMicheli$segment[denoised.sce.deMicheli.s8] <- "segment_8"
denoised.sce.deMicheli$segment[denoised.sce.deMicheli.s9] <- "segment_9"
denoised.sce.deMicheli$segment[denoised.sce.deMicheli.s10] <- "segment_10"
segment.deMicheli <- as.factor(denoised.sce.deMicheli$segment)

# Draw trajectory plot
draw_trajectory_plot(space = space.deMicheli,
                     progression_group = segment.deMicheli,
                     path = traj.deMicheli$path,
                     contour = TRUE) + ggtitle("De Micheli et al. dataset: All genes (ndim = 2) - Segment")

# Predict segments of primary data cells
pred.segments <- SingleR(test = denoised.sce,
                         assay.type.test = "logcounts",
                         ref = denoised.sce.deMicheli,
                         labels = denoised.sce.deMicheli$segment)

#saveRDS(pred.segments, file = "./../saved/pred_segments")
table(denoised.sce$sample, pred.segments$pruned.labels)
```

**Using only D5 MuSCs & progenitors sample**
```{r deMicheli-traj-inference-segments-d5}
# Sort cells by their pseudo time value
ordered <- sort(traj.deMicheli$time, decreasing = FALSE)

# Subset only D5 samples
d5.muscs <- names(ordered) %in% colnames(denoised.sce.deMicheli[, denoised.sce.deMicheli$cell_type == "d5_MuSCs_and_progenitors"])
ordered.d5 <- ordered[d5.muscs]

# Split cells into 10 segments based on their pseudo time value
n = 10
ordered_chunks.d5 = chunk(ordered.d5, n.chunks = 10, shuffle = FALSE)
segment.1 <- names(ordered_chunks.d5[[1]])
segment.2 <- names(ordered_chunks.d5[[2]])
segment.3 <- names(ordered_chunks.d5[[3]])
segment.4 <- names(ordered_chunks.d5[[4]])
segment.5 <- names(ordered_chunks.d5[[5]])
segment.6 <- names(ordered_chunks.d5[[6]])
segment.7 <- names(ordered_chunks.d5[[7]])
segment.8 <- names(ordered_chunks.d5[[8]])
segment.9 <- names(ordered_chunks.d5[[9]])
segment.10 <- names(ordered_chunks.d5[[10]])

# Inspect min / max values of each segment
df.values <- data.frame(segment = c(1:10),
                        min_pseudo_time_value = c(min(ordered_chunks.d5[[1]]),
                                min(ordered_chunks.d5[[2]]),
                                min(ordered_chunks.d5[[3]]),
                                min(ordered_chunks.d5[[4]]),
                                min(ordered_chunks.d5[[5]]),
                                min(ordered_chunks.d5[[6]]),
                                min(ordered_chunks.d5[[7]]),
                                min(ordered_chunks.d5[[8]]),
                                min(ordered_chunks.d5[[9]]),
                                min(ordered_chunks.d5[[10]])),
                        max_pseudo_time_value = c(max(ordered_chunks.d5[[1]]),
                                max(ordered_chunks.d5[[2]]),
                                max(ordered_chunks.d5[[3]]),
                                max(ordered_chunks.d5[[4]]),
                                max(ordered_chunks.d5[[5]]),
                                max(ordered_chunks.d5[[6]]),
                                max(ordered_chunks.d5[[7]]),
                                max(ordered_chunks.d5[[8]]),
                                max(ordered_chunks.d5[[9]]),
                                max(ordered_chunks.d5[[10]])))

                        

# Label cells by their segment
denoised.sce.deMicheli.s1 <- colnames(denoised.sce.deMicheli) %in% segment.1
denoised.sce.deMicheli.s2 <- colnames(denoised.sce.deMicheli) %in% segment.2
denoised.sce.deMicheli.s3 <- colnames(denoised.sce.deMicheli) %in% segment.3
denoised.sce.deMicheli.s4 <- colnames(denoised.sce.deMicheli) %in% segment.4
denoised.sce.deMicheli.s5 <- colnames(denoised.sce.deMicheli) %in% segment.5
denoised.sce.deMicheli.s6 <- colnames(denoised.sce.deMicheli) %in% segment.6
denoised.sce.deMicheli.s7 <- colnames(denoised.sce.deMicheli) %in% segment.7
denoised.sce.deMicheli.s8 <- colnames(denoised.sce.deMicheli) %in% segment.8
denoised.sce.deMicheli.s9 <- colnames(denoised.sce.deMicheli) %in% segment.9
denoised.sce.deMicheli.s10 <- colnames(denoised.sce.deMicheli) %in% segment.10

denoised.sce.deMicheli$segment.d5[denoised.sce.deMicheli.s1] <- "segment_1"
denoised.sce.deMicheli$segment.d5[denoised.sce.deMicheli.s2] <- "segment_2"
denoised.sce.deMicheli$segment.d5[denoised.sce.deMicheli.s3] <- "segment_3"
denoised.sce.deMicheli$segment.d5[denoised.sce.deMicheli.s4] <- "segment_4"
denoised.sce.deMicheli$segment.d5[denoised.sce.deMicheli.s5] <- "segment_5"
denoised.sce.deMicheli$segment.d5[denoised.sce.deMicheli.s6] <- "segment_6"
denoised.sce.deMicheli$segment.d5[denoised.sce.deMicheli.s7] <- "segment_7"
denoised.sce.deMicheli$segment.d5[denoised.sce.deMicheli.s8] <- "segment_8"
denoised.sce.deMicheli$segment.d5[denoised.sce.deMicheli.s9] <- "segment_9"
denoised.sce.deMicheli$segment.d5[denoised.sce.deMicheli.s10] <- "segment_10"

denoised.sce.deMicheli$segment.d5 <- replace_na(denoised.sce.deMicheli$segment.d5, "non_d5_musc_sample")

table(denoised.sce.deMicheli$segment.d5)
segment.deMicheli.d5 <- as.factor(denoised.sce.deMicheli$segment.d5)

# Draw trajectory plot
draw_trajectory_plot(space = space.deMicheli,
                     progression_group = segment.deMicheli.d5,
                     point_size = 1,
                     path = traj.deMicheli$path,
                     contour = FALSE) + ggtitle("De Micheli et al. dataset (Day 5 MuSCs & progenitors): All genes (ndim = 2) - Segment")

# Predict segments of primary data cells
pred.segments.d5 <- SingleR(test = denoised.sce,
                            assay.type.test = "logcounts",
                            ref = denoised.sce.deMicheli,
                            labels = denoised.sce.deMicheli$segment.d5)

#saveRDS(pred.segments.d5, file = "./../saved/pred_segments_d5")
table(denoised.sce$sample, pred.segments.d5$pruned.labels)
```


**Visualization of only segment 10 on trajectory**
```{r seg10-visualization-on-trajectory}
primary.data.cells <- denoised.sce.deMicheli$segment.d5 == "segment_10"
other.cells <- !primary.data.cells

denoised.sce.deMicheli$primary.cells[primary.data.cells] <- "primary_data_cells"
denoised.sce.deMicheli$primary.cells[other.cells] <- "other_cell_types"

colpalette <- c('primary_data_cells' ="#009900",
                'other_cell_types' = "#404040")

draw_trajectory_plot(space = space.deMicheli,
                     progression_group = as.factor(denoised.sce.deMicheli$primary.cells),
                     progression_group_palette = colpalette,
                     point_size = 2,
                     point_alpha = 0.5,
                     path = traj.deMicheli$path,
                     contour = FALSE) +
  ggtitle("Trajectory inference", subtitle = "De Micheli et al. trajectory: Primary data cells shown on trajectory") +
    theme(plot.title = element_text(size = 20, face = "bold"),
        plot.subtitle = element_text(size = 10),
        legend.text = element_text(size = 16),
        legend.title = element_text(size = 20, face = "bold"),
        axis.title = element_text(size = 16))
```




### 6.2 Day 5 MuSCS & Progenitors: Gene importances grouped into modules
```{r deMicheli-d5-muscs-gimp-modules-heatmap, eval=FALSE}
# Identify candidate marker genes from the Day 5 sample
# Visualize modules
group_name.deMicheli.d5 <- as.factor(denoised.sce.deMicheli$segment.d5)
  
pdf("./../saved/modules_heatmap_deMicheli_d5_segments.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli, 
                        time = traj.deMicheli$time,
                        progression_group = group_name.deMicheli.d5,
                        modules = modules.deMicheli,
                        show_labels_row = TRUE,
                        fontsize_row = 1)
dev.off()

```


## 7. Differential expressed genes

### 7.1 DE genes between samples in primary dataset
**uncorrected expression values**
```{r de-genes-primary-data}
markers.primary <- findMarkers(denoised.sce, groups = denoised.sce$sample, pval.type = "some", direction = "up", lfc = 1)
for(sample in unique(denoised.sce$sample)) {
  markers <- markers.primary[[sample]]
  markers.interesting <- markers[markers$p.value <= .05, ]
  # Print out the marker genes
  print(markers.interesting)
  # Get marker effects and plot on a heatmap
  logFCs <- getMarkerEffects(markers.interesting)
  pheatmap(logFCs, breaks=seq(-2.5, 2.5, length.out=101), main = paste(sample, " marker genes"), fontsize_row = 5)
}

```


**cell cycle corrected values**
```{r de-genes-primary-data-cc-corrected}
# All samples
markers.primary.cc <- findMarkers(denoised.sce.cc, groups = denoised.sce.cc$sample, assay.type = "scaled_logcounts", lfc = 1, direction = "up", pval.type = "some")

heatmap.plots = list()

for(sample in unique(denoised.sce.cc$sample)) {
  markers <- markers.primary.cc[[sample]]
  markers.interesting <- markers[markers$p.value < .05, ]
  # Print out the marker genes
  print(markers.interesting)
  # Get marker effects and plot on a heatmap
  logFCs <- getMarkerEffects(markers.interesting)
  p <- pheatmap(logFCs, breaks=seq(-2.5, 2.5, length.out=101), angle_col = 90, main = paste(sample, "\n(cell cycle corrected)\nmarker genes"), fontsize_row = 5, fontsize = 12)
  heatmap.plots[[sample]] <- p[[4]]
}

gridExtra::grid.arrange(heatmap.plots$control, heatmap.plots$cap50, ncol = 2)
gridExtra::grid.arrange(heatmap.plots$control, heatmap.plots$cap50, heatmap.plots$cap50_r4h, heatmap.plots$cap50_r8h, heatmap.plots$cap50_r16h, ncol = 3)
```


### 7.2 De Micheli et al.: DE genes between D5 MuSC segments
```{r de-genes-between-segments}
# Subset the day 5 MuSCs and progenitors sample
deMicheli.d5.muscs <- denoised.sce.deMicheli[, denoised.sce.deMicheli$cell_type == "d5_MuSCs_and_progenitors"]
table(deMicheli.d5.muscs$segment.d5)

# Find marker genes for each of the 10 segments
markers.d5 <- findMarkers(deMicheli.d5.muscs, groups = deMicheli.d5.muscs$segment.d5, direction = "up", pval.type = "any", lfc = 1)

# Add the DE genes into a list of marker genes for each segment
de_genes <- list()
for(segment in names(markers.d5)) {
  interesting <- markers.d5[[segment]]
  best.set <- interesting[interesting$Top <= 6, ]
  de_genes[[segment]] <- best.set
}

# Inspect the genes
de_genes

# Plot heatmaps for each segment
for(segment in names(de_genes)) {
  logFCs <- getMarkerEffects(de_genes[[segment]])
  pheatmap(logFCs, breaks=seq(-5, 5, length.out=101), main = paste("Upregulated genes in: ", segment, " (Top <=6)"))
}

# Inspect the segment 10 in more detail
seg10 <- markers.d5[["segment_10"]]
best.seg10 <- seg10[seg10$p.value < 0.05, ]
logFCs.seg10 <- getMarkerEffects(best.seg10)
pheatmap(logFCs.seg10, breaks=seq(-5, 5, length.out=101), main = "Upregulated genes in segment 10 (p.value <0.05)", fontsize_row = 7)

rownames(best.seg10)
write.csv(best.seg10, file = "./../saved/segment_10_genes.csv")
```


## 8. Differential metabolic pathways

### 8.1 Mapping of KEGG metabolic modules and mouse genes
**Map genes to metabolic modules**
  - Entry = set of genes
  - Each entry maps to multiple different genes
  - Each entry maps to multiple different modules
  - Each entry maps to multiple different pathways
  => Genes and modules have many-to-many relationship
  => Modules and pathways have many-to-many relationship
  => Genes and pathways have many-to-many relationship

```{r kegg-modules}
# Fetch all the modules (and KEGG entries) for mouse
mouse.org.id <- "T01002"
mouse.modules <- keggLink(target = "module", source = mouse.org.id)

# Build a dataframe to store module, gene, and pathway information for each KEGG entry
df.kegg.entries <- data.frame(entry = names(mouse.modules),
                              module = NA,
                              genes = NA,
                              pathway = NA,
                              module_info = NA,
                              pathway_info = NA)

# remove duplicate rows from data
df.kegg.entries <- distinct(df.kegg.entries)

# For each KEGG entry, fetch the information with keggGet()
for (i in 1:nrow(df.kegg.entries)) {
  print(i)
  entry <- df.kegg.entries[i,]$entry
  query <- keggGet(entry)
  genes <- query[[1]]$NAME
  module <- paste(names(query[[1]]$MODULE), collapse = ", ")
  pathway <- paste(names(query[[1]]$PATHWAY), collapse = ", ")
  module_info <- paste(unlist(query[[1]]$MODULE), collapse = ", ")
  pathway_info <- paste(unlist(query[[1]]$PATHWAY), collapse = ", ")
  df.kegg.entries[i,]$genes <- genes
  df.kegg.entries[i,]$module <- module
  df.kegg.entries[i,]$pathway <- pathway
  df.kegg.entries[i,]$module_info <- module_info
  df.kegg.entries[i,]$pathway_info <- pathway_info
}

# Save objects
saveRDS(df.kegg.entries, file = "./../saved/df_kegg_entries")
write.csv(df.kegg.entries, file = "./../saved/df_kegg_entries.csv")
```


### 8.2 Primary data
**Create a new sce with only metabolic genes**
```{r add-KEGG-information-to-primary-data, eval=FALSE}
# Load mouse modules if it doesn't exist
if(!exists("df.kegg.entries")) {
  df.kegg.entries <- readRDS("./../saved/df_kegg_entries")
}

denoised.sce.m <- denoised.sce
rowData(denoised.sce.m)$module <- ""
rowData(denoised.sce.m)$pathway <- ""
rowData(denoised.sce.m)$module_info <- ""
rowData(denoised.sce.m)$pathway_info <- ""

for(i in 1:nrow(df.kegg.entries)) {
  print(i)
  # Split the gene names into a list
  genes <- unlist(lapply(strsplit(df.kegg.entries[i,]$genes, split = ",")[[1]], trimws))
  useful_genes <- genes[genes %in% rownames(denoised.sce.m)]
  for(gene in useful_genes) {
    rowData(denoised.sce.m)[gene, ]$module <- df.kegg.entries[i,]$module
    rowData(denoised.sce.m)[gene, ]$pathway <- df.kegg.entries[i,]$pathway
    rowData(denoised.sce.m)[gene, ]$module_info <- df.kegg.entries[i,]$module_info
    rowData(denoised.sce.m)[gene, ]$pathway_info <- df.kegg.entries[i,]$pathway_info
  }
}

genes.to.use <- rownames(denoised.sce.m)[rowData(denoised.sce.m)$module != ""]
denoised.sce.m <- denoised.sce.m[rownames(denoised.sce.m) %in% genes.to.use, ]

# Save sce object containing only metabolic genes and module/pathway information stored on rowData
saveRDS(denoised.sce.m, "./../saved/denoised_sce_m")
```


**Shortcut: gimp and modules of primary data with metabolic genes**
```{r shortcut-gimp-modules-primary-metabolic-genes}
# Primary data with only metabolic genes
if(!exists("denoised.sce.m")){denoised.sce.m <- readRDS("./../saved/denoised_sce_m")}
gimp.primary.m <- readRDS("./../saved/gimp_primary_m")
modules.primary.m <- readRDS("./../saved/modules_primary_m")
if(!exists("expression.primary.m")) {expression.primary.m <- readRDS("./../saved/expression_primary_m")}
if(!exists("traj.primary")) {traj.primary <- readRDS("./../saved/traj_primary")}
if(!exists("group_name.primary.m")) {group_name.primary.m <- as.factor(denoised.sce.m$sample)}


gimp.primary.m$qvalue <- p.adjust(p = gimp.primary.m$pvalue, method = "BH")
gene_sel.primary.m <- gimp.primary.m$gene[gimp.primary.m$qvalue < .05]
expr_sel.primary.m <- scale_quantile(expression.primary.m[, gene_sel.primary.m])
```


**Gene importances and modules with metabolic genes**
```{r gimp-modules-primary-metabolic-genes, eval=FALSE}
# Read trajectory from a file (if not defined)
if(!exists("traj.primary")) {
  traj.primary <- readRDS("./../saved/traj_primary")
}

# Create expression and sample objects
expression.primary.m <- as.matrix(t(logcounts(denoised.sce.m)))
group_name.primary.m <- as.factor(denoised.sce.m$sample)

# Gene importances
gimp.primary.m <- gene_importances(x = expression.primary.m,
                                           time = traj.primary$time,
                                           num_permutations = 10,
                                           num_threads = 8)

gimp.primary.m$qvalue <- p.adjust(p = gimp.primary.m$pvalue, method = "BH")
gene_sel.primary.m <- gimp.primary.m$gene[gimp.primary.m$qvalue < .05]
expr_sel.primary.m <- scale_quantile(expression.primary.m[, gene_sel.primary.m])

#saveRDS(gimp.primary.m, file = "./../saved/gimp_primary_m")

# Extract modules
modules.primary.m <- extract_modules(expr_sel.primary.m,
                                     time = traj.primary$time,
                                     verbose = FALSE)
#saveRDS(modules.primary.m, "./../saved/modules_primary_m")
```


**Visualization of the gimp & modules plots**
```{r primary-data-gimp-modules-metabolic-genes, eval=FALSE}
# Get rowdata (incl. KEGG modules for each gene) for the metabolic genes
rowdata.primary <- rowData(denoised.sce.m)
genes.f <- rownames(rowdata.primary) %in% colnames(expr_sel.primary.m)
rowdata.primary.f <- rowdata.primary[genes.f, ]

# Create an annotation row df
ann_row <- data.frame(metabolic_module = factor(rowdata.primary.f$module))
rownames(ann_row) <- rownames(rowdata.primary.f)

# Order genes to same order as in modules object
ann_row <- ann_row[modules.primary.m$feature, ,drop=FALSE]

# Add metabolic module information to the modules object
modules.primary.m$metabolic_module <- ann_row$metabolic_module

# Arrange modules object by module and by metabolic_module
modules.primary.m <- arrange(modules.primary.m,
                             as.character(modules.primary.m$module),
                             as.character(modules.primary.m$metabolic_module))

# save modules object
saveRDS(modules.primary.m, file = "./../saved/modules_primary_m_kegg_modules")

# Create annotation row color palette (43 colors)
ann_colors <- colorRampPalette(grDevices::rainbow(length(unique(ann_row$metabolic_module))))
ann_colors <- ann_colors(length(unique(ann_row$metabolic_module)))
names(ann_colors) <- unique(ann_row$metabolic_module)
ann_colors <- list(metabolic_module = ann_colors)

# Gimp
pdf("./../saved/gimp_heatmap_primary_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.primary.m, 
                        time = traj.primary$time,
                        progression_group = group_name.primary.m,
                        show_labels_row = TRUE,
                        fontsize_row = 5)
dev.off()


# Modules
pdf("./../saved/modules_heatmap_primary_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.primary.m, 
                        time = traj.primary$time,
                        progression_group = group_name.primary.m,
                        modules = modules.primary.m,
                        show_labels_row = TRUE,
                        annotation_row = ann_row,
                        #annotation_colors = ann_colors,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Save modules object as csv
write.csv(as.data.frame(modules.primary.m), file = "./../saved/modules_primary_m.csv")
```


### 8.3 De Micheli dataset
**Create a new sce with only metabolic genes**
```{r add-KEGG-information-to-deMicheli-data, eval=FALSE}
# Load mouse modules if it doesn't exist
if(!exists("df.kegg.entries")) {
  df.kegg.entries <- readRDS("./../saved/df_kegg_entries")
}

denoised.sce.deMicheli.m <- denoised.sce.deMicheli
rowData(denoised.sce.deMicheli.m)$module <- ""
rowData(denoised.sce.deMicheli.m)$pathway <- ""
rowData(denoised.sce.deMicheli.m)$module_info <- ""
rowData(denoised.sce.deMicheli.m)$pathway_info <- ""

for(i in 1:nrow(df.kegg.entries)) {
  print(i)
  # Split the gene names into a list
  genes <- unlist(lapply(strsplit(df.kegg.entries[i,]$genes, split = ",")[[1]], trimws))
  useful_genes <- genes[genes %in% rownames(denoised.sce.deMicheli.m)]
  for(gene in useful_genes) {
    rowData(denoised.sce.deMicheli.m)[gene, ]$module <- df.kegg.entries[i,]$module
    rowData(denoised.sce.deMicheli.m)[gene, ]$pathway <- df.kegg.entries[i,]$pathway
    rowData(denoised.sce.deMicheli.m)[gene, ]$module_info <- df.kegg.entries[i,]$module_info
    rowData(denoised.sce.deMicheli.m)[gene, ]$pathway_info <- df.kegg.entries[i,]$pathway_info
  }
}

genes.to.use <- rownames(denoised.sce.deMicheli.m)[rowData(denoised.sce.deMicheli.m)$module != ""]
denoised.sce.deMicheli.m <- denoised.sce.deMicheli.m[rownames(denoised.sce.deMicheli.m) %in% genes.to.use, ]

# Save sce object containing only metabolic genes and module/pathway information stored on rowData
saveRDS(denoised.sce.deMicheli.m, "./../saved/denoised_sce_deMicheli_m")
```


**Shortcut: gimp and modules of deMicheli data with metabolic genes**
```{r shortcut-gimp-modules-deMicheli-metabolic-genes}
# Read deMicheli data with metabolic genes (if not defined)
if(!exists("denoised.sce.deMicheli.m")){denoised.sce.deMicheli.m <- readRDS("./../saved/denoised_sce_deMicheli_m")}

# Read gimp & modules from a file
gimp.deMicheli.m <- readRDS("./../saved/gimp_deMicheli_m")
modules.deMicheli.m <- readRDS("./../saved/modules_deMicheli_m")

# Create expression and sample objects
if(!exists("expression.deMicheli.m")) {expression.deMicheli.m <- as.matrix(t(logcounts(denoised.sce.deMicheli.m)))}
if(!exists("group_name.deMicheli.m")) {group_name.deMicheli.m <- as.factor(denoised.sce.deMicheli.m$cell_type)}

# Read trajectory from a file (if not defined)
if(!exists("traj.deMicheli")) {traj.deMicheli <- readRDS("./../saved/traj_deMicheli")}

gimp.deMicheli.m$qvalue <- p.adjust(p = gimp.deMicheli.m$pvalue, method = "BH")
gene_sel.deMicheli.m <- gimp.deMicheli.m$gene[gimp.deMicheli.m$qvalue < .05]
expr_sel.deMicheli.m <- scale_quantile(expression.deMicheli.m[, gene_sel.deMicheli.m])
```


**Gene importances and modules with metabolic genes**
```{r gimp-modules-deMicheli-metabolic-genes, eval=FALSE}
# Read trajectory from a file (if not defined)
if(!exists("traj.deMicheli")) {
  traj.deMicheli <- readRDS("./../saved/traj_deMicheli")
}

# Create expression and sample objects
expression.deMicheli.m <- as.matrix(t(logcounts(denoised.sce.deMicheli.m)))
group_name.deMicheli.m <- as.factor(denoised.sce.deMicheli.m$cell_type)

# Gene importances
gimp.deMicheli.m <- gene_importances(x = expression.deMicheli.m,
                                     time = traj.deMicheli$time,
                                     num_permutations = 10,
                                     num_threads = 8)

gimp.deMicheli.m$qvalue <- p.adjust(p = gimp.deMicheli.m$pvalue, method = "BH")
gene_sel.deMicheli.m <- gimp.deMicheli.m$gene[gimp.deMicheli.m$qvalue < .05]
expr_sel.deMicheli.m <- scale_quantile(expression.deMicheli.m[, gene_sel.deMicheli.m])

#saveRDS(gimp.deMicheli.m, file = "./../saved/gimp_deMicheli_m")

# Extract modules
modules.deMicheli.m <- extract_modules(expr_sel.deMicheli.m,
                                       time = traj.deMicheli$time,
                                       verbose = FALSE)

#saveRDS(modules.deMicheli.m, "./../saved/modules_deMicheli_m")
```


**Visualization of the gimp & modules plots**
```{r deMicheli-gimp-modules-metabolic-genes, eval=FALSE}
# Get rowdata (incl. KEGG modules for each gene) for the metabolic genes
rowdata.deMicheli <- rowData(denoised.sce.deMicheli.m)
genes.deMicheli.f <- rownames(rowdata.deMicheli) %in% colnames(expr_sel.deMicheli.m)
rowdata.deMicheli.f <- rowdata.deMicheli[genes.deMicheli.f, ]

# Create an annotation row df
ann_row_deMicheli <- data.frame(metabolic_module = factor(rowdata.deMicheli.f$module))
rownames(ann_row_deMicheli) <- rownames(rowdata.deMicheli.f)

# Order genes to same order as in modules object
ann_row_deMicheli <- ann_row_deMicheli[modules.deMicheli.m$feature, ,drop=FALSE]

# Add metabolic module information to the modules object
modules.deMicheli.m$metabolic_module <- ann_row_deMicheli$metabolic_module

# Arrange modules object by module and by metabolic_module
modules.deMicheli.m <- arrange(modules.deMicheli.m,
                             as.character(modules.deMicheli.m$module),
                             as.character(modules.deMicheli.m$metabolic_module))

# save modules object
saveRDS(modules.deMicheli.m, file = "./../saved/modules_deMicheli_m_kegg_modules")

# Gimp
pdf("./../saved/gimp_heatmap_deMicheli_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli.m, 
                        time = traj.deMicheli$time,
                        progression_group = group_name.deMicheli.m,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Modules
pdf("./../saved/modules_heatmap_deMicheli_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli.m, 
                        time = traj.deMicheli$time,
                        progression_group = group_name.deMicheli.m,
                        modules = modules.deMicheli.m,
                        annotation_row = ann_row_deMicheli,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Save modules object as csv
write.csv(as.data.frame(modules.deMicheli.m), file = "./../saved/modules_deMicheli_m.csv")

```



### 8.4 Dell'Orso dataset
**Create a new sce with only metabolic genes**
```{r add-KEGG-information-to-dellOrso-data, eval=FALSE}
# Load mouse modules if it doesn't exist
if(!exists("df.kegg.entries")) {
  df.kegg.entries <- readRDS("./../saved/df_kegg_entries")
}

denoised.sce.dellOrso.m <- denoised.sce.dellOrso
rowData(denoised.sce.dellOrso.m)$module <- ""
rowData(denoised.sce.dellOrso.m)$pathway <- ""
rowData(denoised.sce.dellOrso.m)$module_info <- ""
rowData(denoised.sce.dellOrso.m)$pathway_info <- ""

for(i in 1:nrow(df.kegg.entries)) {
  print(i)
  # Split the gene names into a list
  genes <- unlist(lapply(strsplit(df.kegg.entries[i,]$genes, split = ",")[[1]], trimws))
  useful_genes <- genes[genes %in% rownames(denoised.sce.dellOrso.m)]
  for(gene in useful_genes) {
    rowData(denoised.sce.dellOrso.m)[gene, ]$module <- df.kegg.entries[i,]$module
    rowData(denoised.sce.dellOrso.m)[gene, ]$pathway <- df.kegg.entries[i,]$pathway
    rowData(denoised.sce.dellOrso.m)[gene, ]$module_info <- df.kegg.entries[i,]$module_info
    rowData(denoised.sce.dellOrso.m)[gene, ]$pathway_info <- df.kegg.entries[i,]$pathway_info
  }
}

genes.to.use <- rownames(denoised.sce.dellOrso.m)[rowData(denoised.sce.dellOrso.m)$module != ""]
denoised.sce.dellOrso.m <- denoised.sce.dellOrso.m[rownames(denoised.sce.dellOrso.m) %in% genes.to.use, ]

# Save sce object containing only metabolic genes and module/pathway information stored on rowData
saveRDS(denoised.sce.dellOrso.m, "./../saved/denoised_sce_dellOrso_m")
```



**Shortcut: gimp and modules of dellOrso data with metabolic genes**
```{r shortcut-gimp-modules-dellOrso-metabolic-genes}
# Read deMicheli data with metabolic genes (if not defined)
if(!exists("denoised.sce.dellOrso.m")){denoised.sce.dellOrso.m <- readRDS("./../saved/denoised_sce_dellOrso_m")}

# Read gimp & modules from a file
gimp.dellOrso.m <- readRDS("./../saved/gimp_dellOrso_m")
modules.dellOrso.m <- readRDS("./../saved/modules_dellOrso_m")

# Create expression and sample objects
if(!exists("expression.dellOrso.m")) {expression.dellOrso.m <- as.matrix(t(logcounts(denoised.sce.dellOrso.m)))}
if(!exists("group_name.dellOrso.m")) {group_name.dellOrso.m <- as.factor(denoised.sce.dellOrso.m$cell_type)}

# Read trajectory from a file (if not defined)
if(!exists("traj.deMicheli")) {traj.deMicheli <- readRDS("./../saved/traj_deMicheli")}

gimp.dellOrso.m$qvalue <- p.adjust(p = gimp.dellOrso.m$pvalue, method = "BH")
gene_sel.dellOrso.m <- gimp.dellOrso.m$gene[gimp.dellOrso.m$qvalue < .05]
expr_sel.dellOrso.m <- scale_quantile(expression.dellOrso.m[, gene_sel.dellOrso.m])
```


**Gene importances and modules with metabolic genes**
```{r gimp-modules-dellOrso-metabolic-genes, eval=FALSE}
# Read trajectory from a file (if not defined)
if(!exists("traj.dellOrso")) {
  traj.dellOrso <- readRDS("./../saved/traj_dellOrso")
}

# Create expression and sample objects
expression.dellOrso.m <- as.matrix(t(logcounts(denoised.sce.dellOrso.m)))
group_name.dellOrso.m <- as.factor(denoised.sce.dellOrso.m$cell_type)

# Gene importances
gimp.dellOrso.m <- gene_importances(x = expression.dellOrso.m,
                                    time = traj.dellOrso$time,
                                    num_permutations = 10,
                                    num_threads = 8)

gimp.dellOrso.m$qvalue <- p.adjust(p = gimp.dellOrso.m$pvalue, method = "BH")
gene_sel.dellOrso.m <- gimp.dellOrso.m$gene[gimp.dellOrso.m$qvalue < .05]
expr_sel.dellOrso.m <- scale_quantile(expression.dellOrso.m[, gene_sel.dellOrso.m])

#saveRDS(gimp.dellOrso.m, file = "./../saved/gimp_dellOrso_m")

# Extract modules
modules.dellOrso.m <- extract_modules(expr_sel.dellOrso.m,
                                      time = traj.dellOrso$time,
                                      verbose = FALSE)

#saveRDS(modules.dellOrso.m, "./../saved/modules_dellOrso_m")
```



**Visualization of the gimp & modules plots**
```{r dellOrso-gimp-modules-metabolic-genes, eval=FALSE}
# Get rowdata (incl. KEGG modules for each gene) for the metabolic genes
rowdata.dellOrso <- rowData(denoised.sce.dellOrso.m)
genes.dellOrso.f <- rownames(rowdata.dellOrso) %in% colnames(expr_sel.dellOrso.m)
rowdata.dellOrso.f <- rowdata.dellOrso[genes.dellOrso.f, ]

# Create an annotation row df
ann_row_dellOrso <- data.frame(metabolic_module = factor(rowdata.dellOrso.f$module))
rownames(ann_row_dellOrso) <- rownames(rowdata.dellOrso.f)

# Order genes to same order as in modules object
ann_row_dellOrso <- ann_row_dellOrso[modules.dellOrso.m$feature, ,drop=FALSE]

# Add metabolic module information to the modules object
modules.dellOrso.m$metabolic_module <- ann_row_dellOrso$metabolic_module

# Arrange modules object by module and by metabolic_module
modules.dellOrso.m <- arrange(modules.dellOrso.m,
                              as.character(modules.dellOrso.m$module),
                              as.character(modules.dellOrso.m$metabolic_module))

# save modules object
saveRDS(modules.dellOrso.m, file = "./../saved/modules_dellOrso_m_kegg_modules")

# Gimp
pdf("./../saved/gimp_heatmap_dellOrso_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.dellOrso.m, 
                        time = traj.dellOrso$time,
                        progression_group = group_name.dellOrso.m,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Modules
pdf("./../saved/modules_heatmap_dellOrso_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.dellOrso.m, 
                        time = reverse_trajectory(traj.dellOrso)$time,
                        progression_group = group_name.dellOrso.m,
                        modules = modules.dellOrso.m,
                        annotation_row = ann_row_dellOrso,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Save modules object as csv
write.csv(as.data.frame(modules.dellOrso.m), file = "./../saved/modules_dellOrso_m.csv")

```


### 8.6 Inspection of metabolic modules
```{r inspection-of-metabolic-modules}
# Load modules objects with kegg modules
if(!exists("modules.primary.m")){modules.primary.m <- readRDS("./../saved/modules_primary_m_kegg_modules")}
if(!exists("modules.deMicheli.m")){modules.deMicheli.m <- readRDS("./../saved/modules_deMicheli_m_kegg_modules")}
if(!exists("modules.dellOrso.m")){modules.dellOrso.m <- readRDS("./../saved/modules_dellOrso_m_kegg_modules")}

# Create count tables
tab.modules.primary <- table(modules.primary.m$metabolic_module, modules.primary.m$module)
tab.modules.deMicheli <- table(modules.deMicheli.m$metabolic_module, modules.deMicheli.m$module)
tab.modules.dellOrso <- table(modules.dellOrso.m$metabolic_module, modules.dellOrso.m$module)

# Melt tables
df.modules.primary <- melt(tab.modules.primary)
colnames(df.modules.primary) <- c("KEGG_module", "Scorpius_module", "frequency")
df.modules.primary$KEGG_module <- unfactor(df.modules.primary$KEGG_module)

df.modules.deMicheli <- melt(tab.modules.deMicheli)
colnames(df.modules.deMicheli) <- c("KEGG_module", "Scorpius_module", "frequency")
df.modules.deMicheli$KEGG_module <- unfactor(df.modules.deMicheli$KEGG_module)

df.modules.dellOrso <- melt(tab.modules.dellOrso)
colnames(df.modules.dellOrso) <- c("KEGG_module", "Scorpius_module", "frequency")
df.modules.dellOrso$KEGG_module <- unfactor(df.modules.dellOrso$KEGG_module)

# Handle rows where there are multiple KEGG modules mapping to one scorpius module
df.modules.primary.separated <- distinct(as.data.frame(separate_rows(df.modules.primary, KEGG_module, sep = ", ")))
df.modules.deMicheli.separated <- distinct(as.data.frame(separate_rows(df.modules.deMicheli, KEGG_module, sep = ", ")))
df.modules.dellOrso.separated <- distinct(as.data.frame(separate_rows(df.modules.dellOrso, KEGG_module, sep = ", ")))

# Save all relevant KEGG modules in all datasets
all.kegg.modules <- df.modules.primary.separated$KEGG_module %>% 
  union(df.modules.deMicheli.separated$KEGG_module) %>% 
  union(df.modules.dellOrso.separated$KEGG_module) %>% 
  unique()

saveRDS(all.kegg.modules, file = "./../saved/all_kegg_modules")


# Make plots
p.primary <- ggplot(df.modules.primary.separated, aes(x = as.factor(Scorpius_module), y = frequency)) +
  geom_bar(aes(fill = KEGG_module), stat = "identity", position = position_dodge(1)) +
  ggtitle("Frequency of KEGG modules in different trajectory modules",
          subtitle = "Primary dataset") +
  theme(legend.position="right") +
  geom_label_repel(aes(label=ifelse(frequency>=3, KEGG_module, "")), position = position_dodge(width=0.9), vjust=-0.5)
  
p.deMicheli <- ggplot(df.modules.deMicheli.separated, aes(x = as.factor(Scorpius_module), y = frequency)) +
  geom_bar(aes(fill = KEGG_module), stat = "identity", position = position_dodge(1)) +
  ggtitle("Frequency of KEGG modules in different trajectory modules",
          subtitle = "De Micheli et al. reference dataset") +
  theme(legend.position="right") +
  geom_label_repel(aes(label=ifelse(frequency>=3, KEGG_module, "")), position = position_dodge(width=0.9), vjust=-0.5)
  
p.dellOrso <- ggplot(df.modules.dellOrso.separated, aes(x = as.factor(Scorpius_module), y = frequency)) +
  geom_bar(aes(fill = KEGG_module), stat = "identity", position = position_dodge(1)) +
  ggtitle("Frequency of KEGG modules in different trajectory modules",
          subtitle = "Dell'Orso et al. referene dataset") +
  theme(legend.position="right") +
  geom_label_repel(aes(label=ifelse(frequency>=3, KEGG_module, "")), position = position_dodge(width=0.9), vjust=-0.5)

```


**Save plots as pdf**
```{r kegg-module-plots-as-pdf, eval=FALSE}
pdf("./../saved/KEGG_modules_in_trajectories", width = 11, height = 5)
p.primary
p.deMicheli
p.dellOrso
dev.off()
```


## 8.7 Inspection of metabolic pathways
### 8.7.1 Create mapping objects between genes, Scorpius modules, and metabolic pathways
```{r mapping-genes-modules-pathways}
if(!exists("modules.primary.m")){modules.primary.m <- readRDS("./../saved/modules_primary_m_kegg_modules")}
if(!exists("modules.deMicheli.m")){modules.deMicheli.m <- readRDS("./../saved/modules_deMicheli_m_kegg_modules")}
if(!exists("modules.dellOrso.m")){modules.dellOrso.m <- readRDS("./../saved/modules_dellOrso_m_kegg_modules")}

###############
# primary data
###############
rowData.primary <- rowData(denoised.sce.m)
genes.primary <- rownames(rowdata.primary) %in% colnames(expr_sel.primary.m)
rowData.primary.f <- rowData.primary[genes.primary, ]

# Create mapping between genes and pathways
pathway.mapping.primary <- data.frame(pathway = factor(rowData.primary.f$pathway))
rownames(pathway.mapping.primary) <- rownames(rowData.primary.f)

# Order genes
pathway.mapping.primary <- pathway.mapping.primary[modules.primary.m$feature, , drop = FALSE]

# Add pathway information to modules object
modules.primary.m$pathway <- pathway.mapping.primary$pathway

# Arrange rows
modules.primary.m <- arrange(modules.primary.m,
                             as.character(modules.primary.m$module),
                             as.character(modules.primary.m$metabolic_module))

saveRDS(modules.primary.m, file = "./../saved/modules_primary_m_pathway")

#########################
# De Micheli et al. data
#########################
rowData.deMicheli <- rowData(denoised.sce.deMicheli.m)
genes.deMicheli <- rownames(rowData.deMicheli) %in% colnames(expr_sel.deMicheli.m)
rowData.deMicheli.f <- rowData.deMicheli[genes.deMicheli, ]

# Create mapping between genes and pathways
pathway.mapping.deMicheli <- data.frame(pathway = factor(rowData.deMicheli.f$pathway))
rownames(pathway.mapping.deMicheli) <- rownames(rowData.deMicheli.f)

# Order genes
pathway.mapping.deMicheli <- pathway.mapping.deMicheli[modules.deMicheli.m$feature, , drop = FALSE]

# Add pathway information to modules object
modules.deMicheli.m$pathway <- pathway.mapping.deMicheli$pathway

# Arrange rows
modules.deMicheli.m <- arrange(modules.deMicheli.m,
                             as.character(modules.deMicheli.m$module),
                             as.character(modules.deMicheli.m$metabolic_module))

saveRDS(modules.deMicheli.m, file = "./../saved/modules_deMicheli_m_pathway")


#########################
# Dell'Orso et al. data
#########################
rowData.dellOrso <- rowData(denoised.sce.dellOrso.m)
genes.dellOrso <- rownames(rowData.dellOrso) %in% colnames(expr_sel.deMicheli.m)
rowData.dellOrso.f <- rowData.dellOrso[genes.dellOrso, ]

# Create mapping between genes and pathways
pathway.mapping.dellOrso <- data.frame(pathway = factor(rowData.dellOrso.f$pathway))
rownames(pathway.mapping.dellOrso) <- rownames(rowData.dellOrso.f)

# Order genes
pathway.mapping.dellOrso <- pathway.mapping.dellOrso[modules.dellOrso.m$feature, , drop = FALSE]

# Add pathway information to modules object
modules.dellOrso.m$pathway <- pathway.mapping.dellOrso$pathway

# Arrange rows
modules.dellOrso.m <- arrange(modules.dellOrso.m,
                             as.character(modules.dellOrso.m$module),
                             as.character(modules.dellOrso.m$metabolic_module))

saveRDS(modules.dellOrso.m, file = "./../saved/modules_dellOrso_m_pathway")
```


### 8.7.2 KEGG pathways vs. Scorpius modules
```{r kegg-pathways-vs-scorpius-modules}
if(!exists("modules.primary.m")){modules.primary.m <- readRDS("./../saved/modules_primary_m_pathway")}
if(!exists("modules.deMicheli.m")){modules.deMicheli.m <- readRDS("./../saved/modules_deMicheli_m_pathway")}
if(!exists("modules.dellOrso.m")){modules.dellOrso.m <- readRDS("./../saved/modules_dellOrso_m_pathway")}

# Create count tables
tab.pathways.primary <- table(modules.primary.m$pathway, modules.primary.m$module)
tab.pathways.deMicheli <- table(modules.deMicheli.m$pathway, modules.deMicheli.m$module)
tab.pathways.dellOrso <- table(modules.dellOrso.m$pathway, modules.dellOrso.m$module)

# Melt tables
df.pathways.primary <- melt(tab.pathways.primary)
colnames(df.pathways.primary) <- c("KEGG_pathway", "Scorpius_module", "frequency")
df.pathways.primary$KEGG_pathway <- unfactor(df.pathways.primary$KEGG_pathway)

df.pathways.deMicheli <- melt(tab.pathways.deMicheli)
colnames(df.pathways.deMicheli) <- c("KEGG_pathway", "Scorpius_module", "frequency")
df.pathways.deMicheli$KEGG_pathway <- unfactor(df.pathways.deMicheli$KEGG_pathway)

df.pathways.dellOrso <- melt(tab.pathways.dellOrso)
colnames(df.pathways.dellOrso) <- c("KEGG_pathway", "Scorpius_module", "frequency")
df.pathways.dellOrso$KEGG_pathway <- unfactor(df.pathways.dellOrso$KEGG_pathway)

# Handle rows where there are multiple KEGG modules mapping to one scorpius module
df.pathways.primary.separated <- distinct(as.data.frame(separate_rows(df.pathways.primary, KEGG_pathway, sep = ", ")))
df.pathways.deMicheli.separated <- distinct(as.data.frame(separate_rows(df.pathways.deMicheli, KEGG_pathway, sep = ", ")))
df.pathways.dellOrso.separated <- distinct(as.data.frame(separate_rows(df.pathways.dellOrso, KEGG_pathway, sep = ", ")))

# Make plots
p.primary <- ggplot(df.pathways.primary.separated, aes(x = as.factor(Scorpius_module), y = frequency)) +
  geom_bar(aes(fill = KEGG_pathway), stat = "identity", position = position_dodge(1)) +
  ggtitle("Frequency of KEGG pathways in different trajectory modules",
          subtitle = "Primary dataset") +
  theme(legend.position="right") +
  geom_label_repel(aes(label=ifelse(frequency>=3, KEGG_pathway, "")), position = position_dodge(width=0.9), vjust=-0.5)
  
p.deMicheli <- ggplot(df.pathways.deMicheli.separated, aes(x = as.factor(Scorpius_module), y = frequency)) +
  geom_bar(aes(fill = KEGG_pathway), stat = "identity", position = position_dodge(1)) +
  ggtitle("Frequency of KEGG pathways in different trajectory modules",
          subtitle = "De Micheli et al. reference dataset") +
  theme(legend.position="right") +
  geom_label_repel(aes(label=ifelse(frequency>=10, KEGG_pathway, "")), position = position_dodge(width=0.9), vjust=-0.5)
  
p.dellOrso <- ggplot(df.pathways.dellOrso.separated, aes(x = as.factor(Scorpius_module), y = frequency)) +
  geom_bar(aes(fill = as.factor(KEGG_pathway)), stat = "identity", position = position_dodge(1)) +
  ggtitle("Frequency of KEGG pathways in different trajectory modules",
          subtitle = "Dell'Orso et al. referene dataset") +
  theme(legend.position="right") +
  geom_label_repel(aes(label=ifelse(frequency>=6, KEGG_pathway, "")), position = position_dodge(width=0.9), vjust=-0.5)

```

**Save plots as pdf**
```{r kegg-module-plots-as-pdf, eval=FALSE}
pdf("./../saved/KEGG_pathways_in_trajectories", width = 11, height = 5)
p.primary
p.deMicheli
p.dellOrso
dev.off()
```


## 10. Ridgeline plots to inspect how the distribution of marker genes differs in the samples of primary and reference datasets
```{r ridge-plots}
genes.primary = c("Pax7", "Myf5", "Myod1", "Cdh15", "Tnnt3", "Hes1", "Hes6", "Notch1", "Id1", "Id2", "Id3", "Wnt7b")
genes.deMicheli = c("Pax7", "Myf5", "Myod1", "Myog", "Cdh15", "Tnnt3", "Hes1", "Heyl", "Hes6", "Notch1", "Notch2", "Notch3", "Id1", "Id2", "Id3", "Wnt2", "Wnt10b")

# De Micheli data
sce.deMicheli.f <- denoised.sce.deMicheli[, denoised.sce.deMicheli$cell_type %in% c("d0_MuSCs_and_progenitors", "d5_MuSCs_and_progenitors")]

sce.deMicheli.f$cell_type <- revalue(sce.deMicheli.f$cell_type, c("d0_MuSCs_and_progenitors" = "Day 0", "d5_MuSCs_and_progenitors" = "Day 5"))
seurat.deMicheli <- as.Seurat(sce.deMicheli.f)
Idents(seurat.deMicheli) <- "cell_type"
ridge.deMicheli <- RidgePlot(seurat.deMicheli, features = genes.deMicheli, ncol = 2)
ridge.deMicheli

# Primary data
seurat.primary <- as.Seurat(denoised.sce)
Idents(seurat.primary) <- "sample"
ridge.primary <- RidgePlot(seurat.primary, features = genes.primary, ncol = 2)
ridge.primary

```


**Save ridge line plots as pdf**
```{r ridge-plots-pdf, eval=FALSE}
pdf("./../saved/ridge_plot_primary.pdf", width = 10, height = 25)
ridge.primary
dev.off()

pdf("./../saved/ridge_plot_deMicheli.pdf", width = 10, height = 25)
ridge.deMicheli
dev.off()
```


## 11. De Micheli trajectory analysis without mature skeletal muscle
### 11.1 Load and prepare objects
```{r traj-inference-deMicheli-without-muscle}
# Load objects (generated by Puhti)
traj.deMicheli.m.f <- readRDS("./../saved/traj_deMicheli_m_f")
gimp.deMicheli.m.f <- readRDS("./../saved/gimp_deMicheli_m_f")
modules.deMicheli.m.f <- readRDS("./../saved/modules_deMicheli_m_f")

# Read deMicheli data with metabolic genes and without mature skeletal muscle
if(!exists("denoised.sce.deMicheli.m")){denoised.sce.deMicheli.m <- readRDS("./../saved/denoised_sce_deMicheli_m")}
muscle_cells <- denoised.sce.deMicheli.m$cell_type == "d0_mature_skeletal_muscle" | denoised.sce.deMicheli.m$cell_type =="d5_mature_skeletal_muscle" | denoised.sce.deMicheli.m$cell_type == "d7_mature_skeletal_muscle"
denoised.sce.deMicheli.m.f <- denoised.sce.deMicheli.m[, !muscle_cells]

# Create expression and sample objects
expression.deMicheli.m.f <- as.matrix(t(logcounts(denoised.sce.deMicheli.m.f)))
set.seed(123)
space.deMicheli.m.f <- reduce_dimensionality(expression.deMicheli.m.f, dist = "spearman", ndim = 2)
group_name.deMicheli.m.f <- as.factor(denoised.sce.deMicheli.m.f$cell_type)

gimp.deMicheli.m.f$qvalue <- p.adjust(p = gimp.deMicheli.m.f$pvalue, method = "BH")
gene_sel.deMicheli.m.f <- gimp.deMicheli.m.f$gene[gimp.deMicheli.m.f$qvalue < .05]
expr_sel.deMicheli.m.f <- scale_quantile(expression.deMicheli.m.f[, gene_sel.deMicheli.m.f])

pdf("./../saved/trajectory_deMicheli_metabolic_genes_without_mature_muscle.pdf", width = 10)
draw_trajectory_plot(space = space.deMicheli.m.f, 
                     progression_group = group_name.deMicheli.m.f,
                     path = traj.deMicheli.m.f$path,
                     point_size = 1,
                     point_alpha = 0.5,
                     contour = TRUE) +
  ggtitle("Trajectory inference", subtitle = "De Micheli et al. dataset: Metabolic genes.\nWithout mature skeletal muscle cells (ndim = 2) - Cell types") +
  theme(plot.title = element_text(size = 20, face = "bold"),
        plot.subtitle = element_text(size = 14),
        axis.title = element_text(size = 16))
dev.off()
```

### 11.2 Trajectory heatmaps
```{r traj-heatmaps-deMicheli-without-muscle}
rowdata.deMicheli.m.f <- rowData(denoised.sce.deMicheli.m.f)
genes.deMicheli.f <- rownames(rowdata.deMicheli.m.f) %in% colnames(expr_sel.deMicheli.m.f)
rowdata.deMicheli.m.f <- rowdata.deMicheli.m.f[genes.deMicheli.f, ]

# Create an annotation row df
ann_row_deMicheli <- data.frame(metabolic_module = factor(rowdata.deMicheli.m.f$module))
rownames(ann_row_deMicheli) <- rownames(rowdata.deMicheli.m.f)

# Order genes to same order as in modules object
ann_row_deMicheli <- ann_row_deMicheli[modules.deMicheli.m.f$feature, ,drop=FALSE]

# Add metabolic module information to the modules object
modules.deMicheli.m.f$metabolic_module <- ann_row_deMicheli$metabolic_module

# Arrange modules object by module and by metabolic_module
modules.deMicheli.m.f <- arrange(modules.deMicheli.m.f,
                             as.character(modules.deMicheli.m.f$module),
                             as.character(modules.deMicheli.m.f$metabolic_module))

# Gimp
pdf("./../saved/gimp_heatmap_deMicheli_metabolic_genes_without_mature_muscle.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli.m.f, 
                        time = traj.deMicheli.m.f$time,
                        progression_group = group_name.deMicheli.m.f,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Modules
pdf("./../saved/modules_heatmap_deMicheli_metabolic_genes_without_mature_muscle.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli.m.f, 
                        time = reverse_trajectory(traj.deMicheli.m.f)$time,
                        progression_group = group_name.deMicheli.m.f,
                        modules = modules.deMicheli.m.f,
                        annotation_row = ann_row_deMicheli,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()
```
## 12- SingleR annotation using only metabolic genes
```{r singleR-metabolic-genes}
# Load prediction objects
pred.deMicheli.m <- readRDS("./../saved/pred_deMicheli_m")
pred.dellOrso.m <- readRDS("./../saved/pred_dellOrso_m")

tab.deMicheli.m <- t(table(pred.deMicheli.m$pruned.labels, denoised.sce.m$sample)[, c("control", "cap50", "cap50_r4h","cap50_r8h","cap50_r16h")])
tab.dellOrso.m <- t(table(pred.dellOrso.m$pruned.labels, denoised.sce.m$sample)[, c("control", "cap50", "cap50_r4h","cap50_r8h","cap50_r16h")])

# Inspect results
tab.deMicheli.m
tab.dellOrso.m
```
