---
title: "Downstream_analysis"
author: "Ville Lehtonen"
date: "9/7/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Set working dir to the source file location
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
```
# Trajectory inference
SCORPIUS (https://github.com/rcannood/SCORPIUS) will be used for performing the trajectory inference as it is designed for inferring linear developmental trajectories from scRNA-seq data


## 1. Load packages and data
### Load packages
```{r load-packages, message=FALSE}
library(SCORPIUS)
library(scater)
library(scran)
library(SingleCellExperiment)
library(dplyr)
library(SingleR)
library(igraph)
library(stats)
library(cluster)
library(pheatmap)
library(fossil)
library(Seurat)
library(cerebroApp)
library(reshape)
library(data.table)
```


### Load saved objects
**Single Cell Experiments**
```{r load-objects}
# Load pre-processed sce objects
denoised.sce.deMicheli <- readRDS("./../saved/denoised_sce_deMicheli")
denoised.sce.dellOrso <- readRDS("./../saved/denoised_sce_dellOrso")
```

**Cell cycle genes**
```{r}
# Load G2/M- and S- phase marker genes
g2m.genes.mouse <- readRDS("./../saved/cc_g2m_genes_mouse")
s.genes.mouse <- readRDS("./../saved/cc_gs_genes_mouse")

# Combine genes into a single set of cc genes
cc.genes.mouse <- c(g2m.genes.mouse, s.genes.mouse)
```


## 2. Clustering
### 2.1 Primary data

**The clusters are calculated using Puhti supercomputer and results are loaded in the code chunk below to speed up the process**
```{r primary-data-clustering-shortcut}
# Load data with clusters
denoised.sce <- readRDS("./../saved/denoised_sce_clust")
snng <- readRDS("./../saved/snng")
snng.cc <- readRDS("./../saved/snng_cc")
clust <- readRDS("./../saved/clust")
clust.cc <- readRDS("./../saved/clust_cc")

# Generate the CC corrected data
denoised.sce.cc <- as(altExp(denoised.sce, "CC_corrected"), "SingleCellExperiment")
denoised.sce.cc$cluster.cc <- factor(clust.cc)
reducedDim(denoised.sce.cc, "UMAP_cc") <- reducedDim(denoised.sce, "UMAP_cc")
reducedDim(denoised.sce.cc, "TSNE_cc") <- reducedDim(denoised.sce, "TSNE_cc")
reducedDim(denoised.sce.cc, "force_cc") <- reducedDim(denoised.sce, "force_cc")
```

**Perform clustering (in case no pre-calculated files are available):**
```{r primary-data-clustering, eval=FALSE}
# Load data
denoised.sce <- readRDS("./../saved/denoised_sce")

# Get # of reduced dimensions for uncorrected & cc-corrected datasets
dimensions <- ncol(reducedDim(denoised.sce, "PCA"))
dimensions.cc <- ncol(reducedDim(denoised.sce, "PCA_cc"))
dimensions
dimensions.cc

# Build SNN graph with k=180
snng <- buildSNNGraph(denoised.sce, k = 180, use.dimred = "PCA") 

denoised.sce.cc <- as(altExp(denoised.sce, "CC_corrected"), "SingleCellExperiment")
reducedDim(denoised.sce.cc, "PCA_cc") <- reducedDim(denoised.sce, "PCA_cc")
snng.cc <- buildSNNGraph(denoised.sce.cc, k = 180, use.dimred = "PCA_cc")

# Perform clustering and assign cluster labels
clust <- igraph::cluster_walktrap(snng)$membership
clust.cc <- igraph::cluster_walktrap(snng.cc)$membership
table(clust)
table(clust.cc)
denoised.sce$cluster <- factor(clust)
denoised.sce$cluster.cc <- factor(clust.cc)

# Run UMAP, t-SNE, and force-directed layout
## Uncorrected dataset
set.seed(123)
reducedDim(denoised.sce, "force") <- igraph::layout_with_fr(snng)
denoised.sce <- runUMAP(denoised.sce, dimred = "PCA", n_dimred = dimensions)
denoised.sce <- runTSNE(denoised.sce, dimred = "PCA", n_dimred = dimensions)
## Cc corrected data
reducedDim(denoised.sce, "force_cc") <- igraph::layout_with_fr(snng.cc)
denoised.sce <- runTSNE(denoised.sce, altexp = "CC_corrected", exprs_values = "scaled_logcounts", dimred = "PCA_cc", name = "TSNE_cc")
denoised.sce.cc <- runUMAP(denoised.sce.cc, exprs_values = "scaled_logcounts", dimred = "PCA_cc", name = "UMAP_cc", n_neighbors = 180)

# Set reduced dims to cc corrected experiment and vice versa
reducedDim(denoised.sce, "UMAP_cc") <- reducedDim(denoised.sce.cc, "UMAP_cc")
reducedDim(denoised.sce.cc, "TSNE_cc") <- reducedDim(denoised.sce, "TSNE_cc")
reducedDim(denoised.sce.cc, "force_cc") <- reducedDim(denoised.sce, "force_cc")

# Save sce file
#saveRDS(denoised.sce, file = "./../saved/denoised_sce_clust")
```


**Plot the data using UMAP, force-directed layout, and t-SNE**
```{r primary-data-clustering-plots}
# force directed layout plots
p.force <- plotReducedDim(denoised.sce, colour_by = "sample", dimred = "force") + ggtitle("Primary data: force-directed layout (k=180)") + theme(plot.title = element_text(size = 16, face = "bold"))
p.force.cc <- plotReducedDim(denoised.sce.cc, colour_by = "sample", dimred = "force_cc") + ggtitle("Primary data (cc corrected): force-directed layout (k=180)") + theme(plot.title = element_text(size = 16, face = "bold"))
gridExtra::grid.arrange(p.force, p.force.cc, ncol = 2)

# t-SNE plots
p.tsne <- plotReducedDim(denoised.sce, colour_by = "sample", dimred = "TSNE") + ggtitle("Primary data: t-SNE (k=180)") + theme(plot.title = element_text(size = 16, face = "bold"))
p.tsne.cc <- plotReducedDim(denoised.sce.cc, colour_by = "sample", dimred = "TSNE_cc") + ggtitle("Primary data (cc corrected): t-SNE (k=180)") + theme(plot.title = element_text(size = 16, face = "bold"))
gridExtra::grid.arrange(p.tsne, p.tsne.cc, ncol = 2)

# UMAP plots by sample and by cluster
plotReducedDim(denoised.sce, colour_by = "sample", dimred = "UMAP") + ggtitle("Primary data: UMAP (k=180)") + theme(plot.title = element_text(size = 16, face = "bold"))
plotReducedDim(denoised.sce.cc, colour_by = "sample", dimred = "UMAP_cc") + ggtitle("Primary data (cc corrected): UMAP (k=180)") + theme(plot.title = element_text(size = 16, face = "bold"))

plotReducedDim(denoised.sce, colour_by = "cluster", dimred = "UMAP") + ggtitle("Primary data: UMAP with clusters (k=180)") + theme(plot.title = element_text(size = 16, face = "bold"))
plotReducedDim(denoised.sce.cc, colour_by = "cluster.cc", dimred = "UMAP_cc") + ggtitle("Primary data (cc corrected): UMAP with clusters (k=180)") + theme(plot.title = element_text(size = 16, face = "bold"))

```

### 2.2 Clustering comparison
**Calculate graph modularity and ARI of clusterings**
```{r primary-data-clustering-comparison}
# Calculate modularities of the graphs
mod <- modularity(snng, clust)
mod.cc <- modularity(snng.cc, clust.cc)
mod
mod.cc

# Calculate adjusted Rand index
ari <- adj.rand.index(clust, clust.cc)
ari
```


**Calculate the modularity of each cluster and visualize the results**
```{r plot-modularities, eval=FALSE}
# Plot the modularities of clusters
ratio <- clusterModularity(snng, clust, as.ratio = TRUE)
ratio.cc <- clusterModularity(snng.cc, clust.cc, as.ratio = TRUE)

pheatmap(log2(ratio + 1), cluster_cols = FALSE, cluster_rows = FALSE, color=colorRampPalette(c("white", "blue"))(100), main = "Primary data, k = 180")
pheatmap(log2(ratio.cc + 1), cluster_cols = FALSE, cluster_rows = FALSE, color=colorRampPalette(c("white", "blue"))(100), main = "Primary data (cc corrected), k = 180")
```


**Silhouette analysis:**
```{r silhouette-analysis, eval=FALSE}
# Perform silhouette analysis
dist.data <- dist(reducedDim(denoised.sce, "PCA"))
dist.data.cc <- dist(reducedDim(denoised.sce, "PCA_cc"))
sil <- silhouette(clust, dist = dist.data)
sil.cc <- silhouette(clust.cc, dist = dist.data.cc)
plot(sil, main = "Primary data: silhouette plot")
plot(sil, main = "Primary data (cc corrected): silhouette plot")

# Silhouette with subset of cells
denoised.sce.f <- denoised.sce[, seq(1, 21808, by = 80)]
denoised.sce.f.cc <- denoised.sce[, seq(1, 21808, by = 80)]
clust.f <- clust[seq(1, 21808, by = 80)]
clust.f.cc <- clust.cc[seq(1, 21808, by = 80)]
dist.data.f <- dist(reducedDim(denoised.sce.f, "PCA"))
dist.data.f.cc <- dist(reducedDim(denoised.sce.f.cc, "PCA"))
sil.f <- silhouette(clust.f, dist = dist.data.f)
sil.f.cc <- silhouette(clust.f.cc, dist = dist.data.f.cc)

plot(sil.f, main = "Subset of primary data: silhouette plot")
plot(sil.f.cc, main = "Subset of primary data (cc corrected): silhouette plot")

# Plot a heatmap of the differences in clustering between original data and cc-corrected data
tab <- table(clust, clust.cc)
pheatmap(log10(tab+10), main = "Original data vs. CC corrected (k=180)", color=viridis::viridis(100))
```


### 2.3 Primary data: Subclustering and marker gene detection
#### 2.3.1 Subclustering
**Inspect subclusters within the clusters of the primary data**
```{r primary-data-subclustering, eval=FALSE}
# Inspect clusters in each sample
table(denoised.sce$sample, denoised.sce$cluster)
table(denoised.sce.cc$sample, denoised.sce.cc$cluster)

# Perform subclustering
subclust <- quickSubCluster(denoised.sce, groups = denoised.sce$sample, normalize = FALSE)
subclust.cc <- quickSubCluster(denoised.sce.cc, groups = denoised.sce.cc$sample, normalize = FALSE)

# Plot sublusters
p.sub.control <- plotReducedDim(subclust[["control"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data, Control sample - UMAP plot: subclusters")
p.sub.control.cc <- plotReducedDim(subclust.cc[["control"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data (cc corrected), Control sample - UMAP plot: subclusters")

p.sub.cap50 <- plotReducedDim(subclust[["cap50"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data, Cap50 sample - UMAP plot: subclusters")
p.sub.cap50.cc <- plotReducedDim(subclust.cc[["cap50"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data (cc corrected), Cap50 sample - UMAP plot: subclusters")

p.sub.cap50r4h <- plotReducedDim(subclust[["cap50_r4h"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data, R4H sample - UMAP plot: subclusters")
p.sub.cap50r4h.cc <- plotReducedDim(subclust.cc[["cap50_r4h"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data (cc corrected), R4H sample - UMAP plot: subclusters")

p.sub.cap50r8h <- plotReducedDim(subclust[["cap50_r8h"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data, R8H sample - UMAP plot: subclusters")
p.sub.cap50r8h.cc <- plotReducedDim(subclust.cc[["cap50_r8h"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data (cc corrected), R8H sample - UMAP plot: subclusters")

p.sub.cap50r16h <- plotReducedDim(subclust[["cap50_r16h"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data, R16H sample - UMAP plot: subclusters")
p.sub.cap50r16h.cc <- plotReducedDim(subclust.cc[["cap50_r16h"]], colour_by = "subcluster", dimred = "UMAP") + ggtitle("Primary data (cc corrected), R16H sample - UMAP plot: subclusters")

gridExtra::grid.arrange(p.sub.control, p.sub.control.cc, ncol = 2)
gridExtra::grid.arrange(p.sub.cap50, p.sub.cap50.cc, ncol = 2)
gridExtra::grid.arrange(p.sub.cap50r4h, p.sub.cap50r4h.cc, ncol = 2)
gridExtra::grid.arrange(p.sub.cap50r8h, p.sub.cap50r8h.cc, ncol = 2)
gridExtra::grid.arrange(p.sub.cap50r16h, p.sub.cap50r16h.cc, ncol = 2)

```

#### 2.3.2 Marker gene detection
**This code chunk is not evaluated (as it would produce quite a few plots) but it can be run if one wants to see the marker genes for each sub-cluster** 
```{r marker-gene-detection, eval=FALSE}
# Find marker genes for Control sample
markers.control <- findMarkers(subclust[["control"]], groups = subclust[["control"]]$subcluster, pval.type = "any", lfc = 1)
markers.control.cc <- findMarkers(subclust.cc[["control"]], groups = subclust.cc[["control"]]$subcluster, pval.type = "any", lfc = 1)

Plot marker genes
for(i in 1:length(markers.control)) {
  print(i)
  chosen <- markers.control[[i]][c("Tnnt1", "Tnnt2", "Tnnt3", "Myog", "Tpm1", "Tpm2"), ]
  logFCs <- getMarkerEffects(chosen)
  pheatmap(logFCs, breaks=seq(-5, 5, length.out=101), main = paste("Control Subcluster:", i))
}

for(i in 1:length(markers.control.cc)) {
  print(i)
  chosen <- markers.control.cc[[i]][c("Tnnt1", "Tnnt2", "Tnnt3", "Myog", "Tpm1", "Tpm2"), ]
  logFCs <- getMarkerEffects(chosen)
  pheatmap(logFCs, breaks=seq(-5, 5, length.out=101), main = paste("Control Subcluster (cc):", i))
}

```

### 2.4 Create Cerebro files (.crb)
**This code chunk is used for producing cerebro files and it is by default not evaluated**
```{r cerebro-files, eval=FALSE}
sce.seurat <- as.Seurat(denoised.sce, counts = "counts", data = "logcounts")
sce.cc.seurat <- as.Seurat(denoised.sce.cc, counts = "scaled_logcounts", data = "scaled_logcounts")

exportFromSeurat(object = sce.seurat,
                 file = "./../saved/primary_data.crb", 
                 experiment_name = "Primary data (no correction)",
                 organism = "mm",
                 column_sample = "sample",
                 column_cluster = "cluster",
                 column_nUMI = "sum", 
                 column_cell_cycle_seurat = "Phase",
                 column_nGene = "detected", 
                 add_all_meta_data = TRUE)


exportFromSeurat(object = sce.cc.seurat,
                 file = "./../saved/primary_data_cc_corrected.crb", 
                 experiment_name = "Primary data (cell cycle corrected)",
                 organism = "mm",
                 column_sample = "sample",
                 column_cluster = "cluster.cc",
                 column_nUMI = "sum", 
                 column_cell_cycle_seurat = "Phase",
                 column_nGene = "detected", 
                 add_all_meta_data = TRUE)
```


## 3. Trajectory inference
### 3.1 De Micheli et al. (2020) reference data
```{r deMicheli-traj-inference}
# Create expression and sample objects
expression.deMicheli <- as.matrix(t(logcounts(denoised.sce.deMicheli)))
group_name.deMicheli <- as.factor(denoised.sce.deMicheli$cell_type)
cc.phases.deMicheli <- as.factor(denoised.sce.deMicheli$Phase)

# Trajectory building
space.deMicheli <- reduce_dimensionality(expression.deMicheli, dist = "spearman", ndim = 2)
traj.deMicheli <- infer_trajectory(space.deMicheli)

# Plot trajectory by cell types
draw_trajectory_plot(space = space.deMicheli, 
                     progression_group = group_name.deMicheli,
                     path = traj.deMicheli$path,
                     contour = TRUE) + ggtitle("De Micheli et al. dataset: All genes (ndim = 2) - Cell types")

# Plot trajectory by cc phases
draw_trajectory_plot(space = space.deMicheli, 
                     progression_group = cc.phases.deMicheli,
                     path = traj.deMicheli$path,
                     contour = TRUE) + ggtitle("De Micheli et al. dataset: All genes (ndim = 2) - Cell cycle phases")

# Abundance of some marker genes
pax7 <- expression.deMicheli[, "Pax7"]
cdk1 <- expression.deMicheli[, "Cdk1"]
myod <- expression.deMicheli[, "Myod1"]
myog <- expression.deMicheli[, "Myog"]
btg2 <- expression.deMicheli[, "Btg2"]
cdc20 <- expression.deMicheli[, "Cdc20"]
cdkn1c <- expression.deMicheli[, "Cdkn1c"]

pax7.plot <- draw_trajectory_plot(space = space.deMicheli, 
                                  progression_group = pax7,
                                  path = traj.deMicheli$path,
                                  contour = FALSE) + 
  ggtitle("De Micheli et al. dataset: Pax7 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

cdk1.plot <- draw_trajectory_plot(space = space.deMicheli, 
                                  progression_group = cdk1,
                                  path = traj.deMicheli$path,
                                  contour = FALSE) +
  ggtitle("De Micheli et al. dataset: Cdk1 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))


myog.plot <- draw_trajectory_plot(space = space.deMicheli, 
                                  progression_group = myog,
                                  path = traj.deMicheli$path,
                                  contour = FALSE) + 
  ggtitle("De Micheli et al. dataset: Myog gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

myod.plot <- draw_trajectory_plot(space = space.deMicheli, 
                                  progression_group = myod,
                                  path = traj.deMicheli$path,
                                  contour = FALSE) + 
  ggtitle("De Micheli et al. dataset: Myod gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))


btg2.plot <- draw_trajectory_plot(space = space.deMicheli, 
                                  progression_group = btg2,
                                  path = traj.deMicheli$path,
                                  contour = FALSE) +
  ggtitle("De Micheli et al. dataset: Btg2 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))


cdc20.plot <- draw_trajectory_plot(space = space.deMicheli, 
                                  progression_group = cdc20,
                                  path = traj.deMicheli$path,
                                  contour = FALSE) + 
  ggtitle("De Micheli et al. dataset: Cdc20 gene") + 
  theme(plot.title = element_text(size = 10, face = "bold"))


cdkn1c.plot <- draw_trajectory_plot(space = space.deMicheli, 
                                  progression_group = cdkn1c,
                                  path = traj.deMicheli$path,
                                  contour = FALSE) + 
  ggtitle("De Micheli et al. dataset: Cdkn1c gene") + 
  theme(plot.title = element_text(size = 10, face = "bold"))


gridExtra::grid.arrange(pax7.plot, cdk1.plot, myog.plot, myod.plot, btg2.plot, cdc20.plot, cdkn1c.plot, ncol = 2)

```

**Trajectory inference without cell cycle genes**
```{r deMicheli-traj-inference-cc-genes}
# Remove cell cycle genes from the data
denoised.sce.deMicheli.cc.genes <- denoised.sce.deMicheli[!rownames(denoised.sce.deMicheli) %in% cc.genes.mouse, ]

# Create expression and sample objects
expression.deMicheli.cc.genes <- as.matrix(t(logcounts(denoised.sce.deMicheli.cc.genes)))
group_name.deMicheli.cc.genes <- as.factor(denoised.sce.deMicheli.cc.genes$cell_type)

# Trajectory building
space.deMicheli.cc.genes <- reduce_dimensionality(expression.deMicheli.cc.genes, dist = "spearman", ndim = 2)
traj.deMicheli.cc.genes <- infer_trajectory(space.deMicheli.cc.genes)

# Plot trajectory by cell types
draw_trajectory_plot(space = space.deMicheli.cc.genes, 
                     progression_group = group_name.deMicheli.cc.genes,
                     path = traj.deMicheli.cc.genes$path,
                     contour = TRUE) + ggtitle("De Micheli et al. dataset: All genes except cell cycle marker genes, (ndim = 2) - Cell types")
```


**Trajectory inference for cell cycle corrected expression values:**
```{r deMicheli-traj-inference-cc}
# Create a sce object from the CC corrected alternative experiment
denoised.sce.deMicheli.cc <- as(altExp(denoised.sce.deMicheli, "CC_corrected"), "SingleCellExperiment")


# Create expression and sample objects
expression.deMicheli.cc <- as.matrix(t(assay(denoised.sce.deMicheli.cc, "scaled_logcounts")))
group_name.deMicheli.cc <- as.factor(denoised.sce.deMicheli.cc$cell_type)
cc.phases.deMicheli <- as.factor(denoised.sce.deMicheli.cc$Phase)

# Trajectory building
space.deMicheli.cc <- reduce_dimensionality(expression.deMicheli.cc, dist = "spearman", ndim = 2)
traj.deMicheli.cc <- infer_trajectory(space.deMicheli.cc)

draw_trajectory_plot(space = space.deMicheli.cc, 
                     progression_group = group_name.deMicheli.cc,
                     path = traj.deMicheli.cc$path,
                     contour = TRUE) + ggtitle("De Micheli et al. dataset (cc corrected): HVGs (ndim = 2) - Cell types")

draw_trajectory_plot(space = space.deMicheli.cc, 
                     progression_group = cc.phases.deMicheli,
                     path = traj.deMicheli.cc$path,
                     contour = TRUE) + ggtitle("De Micheli et al. dataset (cc corrected): HVGs (ndim = 2) - Cell cycle phases")
```



### 3.2 Dell'Orso et al. (2019) reference data
```{r dellOrso-traj-inference}
# Create expression and sample objects
expression.dellOrso <- as.matrix(t(logcounts(denoised.sce.dellOrso)))
group_name.dellOrso <- as.factor(denoised.sce.dellOrso$cell_type)
cc.phases.dellOrso <- as.factor(denoised.sce.dellOrso$Phase)

# Trajectory inference
space.dellOrso <- reduce_dimensionality(expression.dellOrso, dist = "spearman", ndim = 2)
traj.dellOrso <- infer_trajectory(space.dellOrso)

# plot by cell type
draw_trajectory_plot(space = space.dellOrso, 
                     progression_group = group_name.dellOrso,
                     path = traj.dellOrso$path,
                     contour = TRUE) + ggtitle("Dell'Orso et al. dataset: All genes, ndim = 2 - Cell types")

# Plot by cell cycle phase
draw_trajectory_plot(space = space.dellOrso, 
                     progression_group = cc.phases.dellOrso,
                     path = traj.dellOrso$path,
                     contour = TRUE) + ggtitle("Dell'Orso et al. dataset: All genes, ndim = 2 - Cell cycle phases")


# Abundance of some marker genes
pax7 <- expression.dellOrso[, "Pax7"]
hes1 <- expression.dellOrso[, "Hes1"]
myf5 <- expression.dellOrso[, "Myf5"]
myod <- expression.dellOrso[, "Myod1"]
ccne1 <- expression.dellOrso[, "Ccne1"]
eifs <- expression.dellOrso[, "Eif1"]
ccnd1 <- expression.dellOrso[, "Ccnd1"]
ccnd2 <- expression.dellOrso[, "Ccnd2"]
myog <- expression.dellOrso[, "Myog"]
tnnt2 <- expression.dellOrso[, "Tnnt2"]
cdkn1a <- expression.dellOrso[, "Cdkn1a"]
cdkn1b <- expression.dellOrso[, "Cdkn1b"]

pax7.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = pax7,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Pax7 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

hes1.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = hes1,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Hes1 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))


myf5.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = myf5,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Myf5 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))


myod.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = myod,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Myod gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

ccne1.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = ccne1,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Ccne1 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

eifs.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = eifs,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: eIFs gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

ccnd1.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = ccnd1,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Ccnd1 gene") +
  theme(plot.title = element_text(size = 12, face = "bold"))

ccnd2.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = ccnd2,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Ccnd2 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

myog.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = myog,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Myog gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

tnnt2.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = tnnt2,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Tnnt2 gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

cdkn1a.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = cdkn1a,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Cdkn1a gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))

cdkn1b.plot <- draw_trajectory_plot(space = space.dellOrso, 
                                  progression_group = cdkn1b,
                                  path = traj.dellOrso$path,
                                  contour = FALSE) +
  ggtitle("Dell'Orso et al. dataset: Cdkn1b gene") +
  theme(plot.title = element_text(size = 10, face = "bold"))


gridExtra::grid.arrange(pax7.plot, hes1.plot,
                        myf5.plot, myod.plot,
                        ccne1.plot, eifs.plot,
                        ccnd1.plot, ccnd2.plot,
                        myog.plot, tnnt2.plot,
                        cdkn1a.plot, cdkn1b.plot)

```


**Trajectory inference without the Primary Myoblast cells**
```{r dellOrso-traj-inference-without-primary-mb}
# Remove PMs from the dellOrso data
denoised.sce.dellOrso.f <- denoised.sce.dellOrso[, denoised.sce.dellOrso$cell_type != "Primary_MB"]

# Create expression and sample objects
expression.dellOrso.f <- as.matrix(t(logcounts(denoised.sce.dellOrso.f)))
group_name.dellOrso.f <- as.factor(denoised.sce.dellOrso.f$cell_type)

# Trajectory inference
space.dellOrso.f <- reduce_dimensionality(expression.dellOrso.f, dist = "spearman", ndim = 2)
traj.dellOrso.f <- infer_trajectory(space.dellOrso.f)

# plot by cell type
draw_trajectory_plot(space = space.dellOrso.f, 
                     progression_group = group_name.dellOrso.f,
                     path = traj.dellOrso.f$path,
                     contour = TRUE) + ggtitle("Dell'Orso et al. dataset: Only homeostatic and injured 60h MuSCs, All genes, ndim = 2")

```

**Trajectory inference without cell cycle genes**
```{r dellOrso-traj-inference-cc-genes}
# Remove cell cycle genes from the data
denoised.sce.dellOrso.cc.genes <- denoised.sce.dellOrso[!rownames(denoised.sce.dellOrso) %in% cc.genes.mouse, ]

# Create expression and sample objects
expression.dellOrso.cc.genes <- as.matrix(t(logcounts(denoised.sce.dellOrso.cc.genes)))
group_name.dellOrso.cc.genes <- as.factor(denoised.sce.dellOrso.cc.genes$cell_type)

# Trajectory inference
space.dellOrso.cc.genes <- reduce_dimensionality(expression.dellOrso.cc.genes, dist = "spearman", ndim = 2)
traj.dellOrso.cc.genes <- infer_trajectory(space.dellOrso.cc.genes)

# plot by cell type
draw_trajectory_plot(space = space.dellOrso.cc.genes, 
                     progression_group = group_name.dellOrso.cc.genes,
                     path = traj.dellOrso.cc.genes$path,
                     contour = TRUE) + ggtitle("Dell'Orso et al. dataset: All genes except cell cycle marker genes, ndim = 2 - Cell types")
```



**Trajectory inference for cell cycle corrected expression values:**
```{r dellOrso-traj-inference-cc}
# Create a sce object from the CC corrected alternative experiment
denoised.sce.dellOrso.cc <- as(altExp(denoised.sce.dellOrso, "CC_corrected"), "SingleCellExperiment")

# Create expression and sample objects
expression.dellOrso.cc <- as.matrix(t(assay(denoised.sce.dellOrso.cc, "scaled_logcounts")))
group_name.dellOrso.cc <- as.factor(denoised.sce.dellOrso.cc$cell_type)
cc.phases.dellOrso <- as.factor(denoised.sce.dellOrso.cc$Phase)

# Trajectory building
space.dellOrso.cc <- reduce_dimensionality(expression.dellOrso.cc, dist = "spearman", ndim = 2)
traj.dellOrso.cc <- infer_trajectory(space.dellOrso.cc)

draw_trajectory_plot(space = space.dellOrso.cc, 
                     progression_group = group_name.dellOrso.cc,
                     path = traj.dellOrso.cc$path,
                     contour = TRUE) + ggtitle("Dell'Orso et al. dataset (cc corrected): HVGs (ndim = 2) - Cell types")

draw_trajectory_plot(space = space.dellOrso.cc, 
                     progression_group = cc.phases.dellOrso,
                     path = traj.dellOrso.cc$path,
                     contour = TRUE) + ggtitle("Dell'Orso et al. dataset (cc corrected): HVGs (ndim = 2) - Cell cycle phases")
```


## 4. Identifying candidate marker genes
### 4.1 De Micheli et al. (2020) reference data

**Shortcut for fetching the gimps & modules generated by Puhti supercomputer**
```{r shortcut-gimp-modules}
gimp.deMicheli <- readRDS("./../saved/gimp_deMicheli")
gimp.dellOrso <- readRDS("./../saved/gimp_dellOrso")
modules.deMicheli <- readRDS("./../saved/modules_deMicheli")
modules.dellOrso <- readRDS("./../saved/modules_dellOrso")

# Generate objects for de Micheli dataset
gimp.deMicheli$qvalue <- p.adjust(p = gimp.deMicheli$pvalue, method = "BH")
gene_sel.deMicheli <- gimp.deMicheli$gene[gimp.deMicheli$qvalue < .05]
expr_sel.deMicheli <- scale_quantile(expression.deMicheli[, gene_sel.deMicheli])

# Generate objects for dellOrso dataset
gimp.dellOrso$qvalue <- p.adjust(gimp.dellOrso$pvalue, "BH", length(gimp.dellOrso$pvalue))
gene_sel.dellOrso <- gimp.dellOrso$gene[gimp.dellOrso$qvalue < .05]
expr_sel.dellOrso <- scale_quantile(expression.dellOrso[, gene_sel.dellOrso])
```

```{r marker-genes-deMicheli, eval=FALSE}
# Identify candidate marker genes
gimp.deMicheli <- gene_importances(x = expression.deMicheli,
                                   time = traj.deMicheli$time,
                                   num_permutations = 10,
                                   num_threads = 8)

gimp.deMicheli$qvalue <- p.adjust(p = gimp.deMicheli$pvalue, method = "BH")
gene_sel.deMicheli <- gimp.deMicheli$gene[gimp.deMicheli$qvalue < .05]
expr_sel.deMicheli <- expression.deMicheli[, gene_sel.deMicheli]

# Group genes to modules
modules.deMicheli <- extract_modules(scale_quantile(expr_sel.deMicheli),
                                     time = traj.deMicheli$time,
                                     verbose = FALSE)

#saveRDS(modules.deMicheli, file = "./../saved/modules_deMicheli")
#saveRDS(gimp.deMicheli, file = "./../saved/gimp_deMicheli")

```


**Visualization of gene importances and modules**
```{r DeMicheli-gimp-modules-visualization, eval=FALSE}
# Visualize expression of selected genes
pdf("./../saved/gimp_heatmap_deMicheli.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli, 
                        time = traj.deMicheli$time,
                        progression_group = group_name.deMicheli,
                        show_labels_row = TRUE,
                        fontsize_row = 2)
dev.off()

# Visualize modules
pdf("./../saved/modules_heatmap_deMicheli.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli, 
                        time = traj.deMicheli$time,
                        progression_group = group_name.deMicheli,
                        modules = modules.deMicheli,
                        show_labels_row = TRUE,
                        fontsize_row = 3)
dev.off()
```


### 4.2 Dell'Orso et al. (2019) reference data
```{r marker-genes-dellOrso, eval=FALSE}
# Identify candidate marker genes
gimp.dellOrso <- gene_importances(x = expression.dellOrso,
                                  time = traj.dellOrso$time,
                                  num_permutations = 10,
                                  num_threads = 8)

gimp.dellOrso$qvalue <- p.adjust(gimp.dellOrso$pvalue, "BH", length(gimp.dellOrso$pvalue))
gene_sel.dellOrso <- gimp.dellOrso$gene[gimp.dellOrso$qvalue < .05]
expr_sel.dellOrso <- scale_quantile(expression.dellOrso[, gene_sel.dellOrso])

# Group genes to modules
modules.dellOrso <- extract_modules(scale_quantile(expr_sel.dellOrso),
                                     time = traj.dellOrso$time,
                                     verbose = FALSE)

#saveRDS(modules.dellOrso, file = "./../saved/modules_dellOrso")
#saveRDS(gimp.dellOrso, file = "./../saved/gimp_dellOrso")
```

**Visualization of gene importances and modules**
```{r DellOrso-gimp-modules-visualization, eval=FALSE}
# Visualize expression of selected genes
pdf("./../saved/gimp_heatmap_dellOrso.pdf")
draw_trajectory_heatmap(expr_sel.dellOrso, 
                        time = traj.dellOrso$time,
                        progression_group = group_name.dellOrso,
                        show_labels_row = TRUE,
                        fontsize_row = 1)
dev.off()

# Visualize modules
pdf("./../saved/modules_heatmap_dellOrso.pdf")
draw_trajectory_heatmap(expr_sel.dellOrso, 
                        time = traj.dellOrso$time,
                        progression_group = group_name.dellOrso,
                        modules = modules.dellOrso,
                        show_labels_row = TRUE,
                        fontsize_row = 1)
dev.off()
```

## 5. Mapping of primary data to reference datasets
### 5.1 Assign reference data cell types to primary data and inspect results

**Predict cell types (shortcut)**
```{r predict-cell-types-shortcut}
# Load pre-saved prediction objects
pred.deMicheli <- readRDS("./../saved/pred_deMicheli")
pred.dellOrso <- readRDS("./../saved/pred_dellOrso")

# Assign cell types to primary data
denoised.sce$cell_type_deMicheli <- pred.deMicheli$pruned.labels
denoised.sce$cell_type_dellOrso <- pred.dellOrso$pruned.labels

```

**Predict cell types**
```{r predict-cell-types, eval=FALSE}
## De Micheli
pred.deMicheli <- SingleR(test = denoised.sce,
                          assay.type.test = "logcounts",
                          ref = denoised.sce.deMicheli,
                          labels = denoised.sce.deMicheli$cell_type)

denoised.sce$cell_type_deMicheli <- pred.deMicheli$pruned.labels
#saveRDS(pred.deMicheli, file = "./../saved/pred_deMicheli")

## Dell'Orso
pred.dellOrso <- SingleR(test = denoised.sce,
                         assay.type.test = "logcounts",
                         ref = denoised.sce.dellOrso,
                         labels = denoised.sce.dellOrso$cell_type)

denoised.sce$cell_type_dellOrso <- pred.dellOrso$pruned.labels
#saveRDS(pred.dellOrso, file = "./../saved/pred_dellOrso")

```


```{r cell-mapping-inspection}
# Inspection of predictions
table(pred.deMicheli$pruned.labels)
table(pred.dellOrso$labels)
tab.deMicheli <- t(table(pred.deMicheli$pruned.labels, denoised.sce$sample)[, c("control", "cap50", "cap50_r4h","cap50_r8h","cap50_r16h")])
tab.dellOrso <- t(table(pred.dellOrso$pruned.labels, denoised.sce$sample)[, c("control", "cap50", "cap50_r4h","cap50_r8h","cap50_r16h")])
tab.deMicheli
tab.dellOrso

## 'melt' tables
tab.deMicheli.melt <- melt(t(tab.deMicheli))
colnames(tab.deMicheli.melt) <- c("cell_type", "sample", "count")
tab.dellOrso.melt <- melt(t(tab.dellOrso))
colnames(tab.dellOrso.melt) <- c("cell_type", "sample", "count")

p.deMicheli <- ggplot(tab.deMicheli.melt, mapping = aes(x = sample, y = count, fill = cell_type)) + 
  geom_bar(position = "dodge", stat = "identity") + ggtitle("Primary data - De Micheli cell types")

p.dellOrso <- ggplot(tab.dellOrso.melt, mapping = aes(x = sample, y = count, fill = cell_type)) + 
  geom_bar(position = "dodge", stat = "identity") + ggtitle("Primary data - Dell'Orso cell types")

gridExtra::grid.arrange(p.deMicheli, p.dellOrso, ncol = 2)
```


**Create diagnstic plots for the cell type annotations**
```{r deMicheli-annotation-diagnostic-plots, eval=FALSE}
# Visualize the score matrix
pdf("./../saved/deMicheli_score_heatmap.pdf")
plotScoreHeatmap(pred.deMicheli, 
                 annotation_col = as.data.frame(colData(denoised.sce)[, "sample", drop = FALSE]))
dev.off()

# Delta distribution
pdf("./../saved/deMicheli_score_distribution.pdf")
plotScoreDistribution(pred.deMicheli, show = "delta.med")
dev.off()

# Get marker genes
all.markers.deMicheli <- metadata(pred.deMicheli)$de.genes

# Write markers for each sample into a csv file
all.markers.deMicheli.df <- data.table("d0_mature_skeletal_muscle" = unique(unlist(all.markers.deMicheli$d0_mature_skeletal_muscle)),
                                       "d0_MuSCs_and_progenitors" = unique(unlist(all.markers.deMicheli$d0_MuSCs_and_progenitors)),
                                       "d5_mature_skeletal_muscle" = unique(unlist(all.markers.deMicheli$d5_mature_skeletal_muscle)),
                                       "d5_MuSCs_and_progenitors" = unique(unlist(all.markers.deMicheli$d5_MuSCs_and_progenitors)),
                                       "d7_mature_skeletal_muscle" = unique(unlist(all.markers.deMicheli$d7_mature_skeletal_muscle)),
                                       "d7_MuSCs_and_progenitors" = unique(unlist(all.markers.deMicheli$d7_MuSCs_and_progenitors)))
write.csv(all.markers.deMicheli.df, file ="./../saved/all_markers_deMicheli.csv")

# Examine the expression of the marker genes for each label in the test dataset
pdf("./../saved/heatmap_deMicheli_d0_muscle.pdf")
plotHeatmap(denoised.sce,
            order_columns_by="cell_type_deMicheli",
            features=unique(unlist(all.markers.deMicheli$d0_mature_skeletal_muscle)),
            fontsize_row = 2)
dev.off()
pdf("./../saved/heatmap_deMicheli_d0_musc.pdf")
plotHeatmap(denoised.sce,
            order_columns_by="cell_type_deMicheli",
            features=unique(unlist(all.markers.deMicheli$d0_MuSCs_and_progenitors)),
            fontsize_row = 2)
dev.off()
pdf("./../saved/heatmap_deMicheli_d5_muscle.pdf")
plotHeatmap(denoised.sce,
            order_columns_by="cell_type_deMicheli",
            features=unique(unlist(all.markers.deMicheli$d5_mature_skeletal_muscle)),
            fontsize_row = 2)
dev.off()
pdf("./../saved/heatmap_deMicheli_d5_musc.pdf")
plotHeatmap(denoised.sce,
            order_columns_by="cell_type_deMicheli",
            features=unique(unlist(all.markers.deMicheli$d5_MuSCs_and_progenitors)),
            fontsize_row = 2)
dev.off()
pdf("./../saved/heatmap_deMicheli_d7_muscle.pdf")
plotHeatmap(denoised.sce,
            order_columns_by="cell_type_deMicheli",
            features=unique(unlist(all.markers.deMicheli$d7_mature_skeletal_muscle)),
            fontsize_row = 2)
dev.off()
pdf("./../saved/heatmap_deMicheli_d7_musc.pdf")
plotHeatmap(denoised.sce,
            order_columns_by="cell_type_deMicheli",
            features=unique(unlist(all.markers.deMicheli$d7_MuSCs_and_progenitors)),
            fontsize_row = 2)
dev.off()
```

```{r dellOrso-annotation-diagnostic-plots, eval=FALSE}
# Visualize the score matrix
pdf("./../saved/dellOrso_score_heatmap.pdf")
plotScoreHeatmap(pred.dellOrso, 
                 annotation_col = as.data.frame(colData(denoised.sce)[, "sample", drop = FALSE]))
dev.off()

# Delta distribution
pdf("./../saved/dellOrso_score_distribution.pdf")
plotScoreDistribution(pred.dellOrso, show = "delta.med")
dev.off()

# Get marker genes
all.markers.dellOrso <- metadata(pred.dellOrso)$de.genes

# Write markers for each sample into a csv file
all.markers.dellOrso.df <- data.table("Homeostatic_MuSCs" = unique(unlist(all.markers.dellOrso$homeostatic_MuSCs)),
                                      "inj_60h_MuSCs" = unique(unlist(all.markers.dellOrso$inj_60h_MuSCs)),
                                      "Primary_MB" = unique(unlist(all.markers.dellOrso$Primary_MB)))

write.csv(all.markers.dellOrso.df, file = "./../saved/all_markers_dellOrso.csv")

# Examine the expression of the marker genes for each label in the test dataset
pdf("./../saved/heatmap_dellOrso_homeostatic_muscs.pdf")
plotHeatmap(denoised.sce, order_columns_by="cell_type_dellOrso",
    features=unique(unlist(all.markers.dellOrso.df$homeostatic_MuSCs)),
    fontsize_row = 2)
dev.off()

pdf("./../saved/heatmap_dellOrso_inj_60h_muscs.pdf")
plotHeatmap(denoised.sce, order_columns_by="cell_type_dellOrso",
    features=unique(unlist(all.markers.dellOrso.df$inj_60h_MuSCs)),
    fontsize_row = 2)
dev.off()

pdf("./../saved/heatmap_dellOrso_primary_mb.pdf")
plotHeatmap(denoised.sce, order_columns_by="cell_type_dellOrso",
    features=unique(unlist(all.markers.dellOrso.df$Primary_MB)),
    fontsize_row = 2)
dev.off()


```


#### 5.2 Assign cell types from reference to another reference dataset
**Shortcut for loading the prediction objects**
```{r cell-type-reference-datasets-shortcut}
# Load the pre-saved reference data cell annotation predictions
pred.reference <- readRDS("./../saved/pred_reference")
pred.reference2 <- readRDS("./../saved/pred_reference2")
```


```{r cell-type-reference-datasets, eval=FALSE}
# Dell'Orso as reference
pred.reference <- SingleR(test = denoised.sce.deMicheli,
                          assay.type.test = "logcounts",
                          ref = denoised.sce.dellOrso,
                          labels = denoised.sce.dellOrso$cell_type)
#saveRDS(pred.reference, file = "./../saved/pred_reference")

# De Micheli as reference
pred.reference2 <- SingleR(test = denoised.sce.dellOrso,
                          assay.type.test = "logcounts",
                          ref = denoised.sce.deMicheli,
                          labels = denoised.sce.deMicheli$cell_type)
#saveRDS(pred.reference2, file = "./../saved/pred_reference2")

```
```{r cell-type-reference-datasets-inspection}
# Dell'Orso as reference:
table(pred.reference$pruned.labels)
t(table(pred.reference$pruned.labels, denoised.sce.deMicheli$cell_type))

# DeMicheli as reference
table(pred.reference2$pruned.labels)
table(pred.reference2$pruned.labels, denoised.sce.dellOrso$cell_type)
```



### 5.2 Build trajectories of reference data with the primary data cell types
```{r trajectories-with-primary-data-samples, eval=FALSE}
# De Micheli
## Predict primary data samples for reference data
pred.deMicheli2 <- SingleR(test = denoised.sce.deMicheli,
                           assay.type.test = "logcounts",
                           ref = denoised.sce,
                           labels = denoised.sce$sample)

## Assign predicted primary data samples to reference datasets
denoised.sce.deMicheli$primary_data_sample <- pred.deMicheli2$pruned.labels
primary.group.deMicheli <- as.factor(denoised.sce.deMicheli$primary_data_sample)

## Plot trajectories with primary data samples
draw_trajectory_plot(space = space.deMicheli, 
                     progression_group = primary.group.deMicheli,
                     path = traj.deMicheli$path,
                     contour = TRUE) + ggtitle("De Micheli et al. trajectory: Primary data samples")

# Dell'Orso
## Predict primary data samples for reference data
pred.dellOrso2 <- SingleR(test = denoised.sce.dellOrso,
                          assay.type.test = "logcounts",
                          ref = denoised.sce,
                          labels = denoised.sce$sample)

## Assign predicted primary data samples to reference datasets
denoised.sce.dellOrso$primary_data_sample <- pred.dellOrso2$pruned.labels
primary.group.dellOrso <- as.factor(denoised.sce.dellOrso$primary_data_sample)

## Plot trajectories with primary data samples
draw_trajectory_plot(space = space.dellOrso, 
                     progression_group = primary.group.dellOrso,
                     path = traj.dellOrso$path,
                     contour = TRUE) + ggtitle("Dell'Orso et al. trajectory: Primary data samples")
```



