---
title: "Metabolic_pathways_pt2"
author: "Ville Lehtonen"
date: "11/9/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Set working dir to the source file location
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
```

# Metabolic pathways
Here, we perform the aggregation of gene expression values into KEGG module / pathway expression and calculate the module / pathway importances along the trajectories.

## 1. Load packages and data
### 1.1 Load packages
```{r load-packages, message=FALSE}
library(SCORPIUS)
library(SingleCellExperiment)
library(stats)
library(data.table)
library(dplyr)


library(gridExtra)
library(ggplot2)
library(gplots)
```

### 1.2 Load pre-calculated data
**Single Cell Experiments**
```{r load-sce-objects}
# Load datasets with metabolic genes only
denoised.sce.m <- readRDS("./../saved/denoised_sce_m")
denoised.sce.deMicheli.m <- readRDS("./../saved/denoised_sce_deMicheli_m")
denoised.sce.dellOrso.m <- readRDS("./../saved/denoised_sce_dellOrso_m")
```

**Trajectories**
```{r load-trajectories}
traj.primary <- readRDS("./../saved/traj_primary")
traj.deMicheli <- readRDS("./../saved/traj_deMicheli")
traj.dellOrso <- readRDS("./../saved/traj_dellOrso")
```

**Aggregated sce objects**
```{r load-aggr-sce-objects}
# Primary
sce.primary.module <- readRDS("./../saved/sce_primary_module")
sce.primary.pathway <- readRDS("./../saved/sce_primary_pathway")

# De Micheli
sce.deMicheli.module <- readRDS("./../saved/sce_deMicheli_module")
sce.deMicheli.pathway <- readRDS("./../saved/sce_deMicheli_pathway")

# Dell Orso
sce.dellOrso.module <- readRDS("./../saved/sce_dellOrso_module")
sce.dellOrso.pathway <- readRDS("./../saved/sce_dellOrso_pathway")

# Inspect dimensions of objects
dim(sce.primary.module)
dim(sce.primary.pathway)
dim(sce.deMicheli.module)
dim(sce.deMicheli.pathway)
dim(sce.dellOrso.module)
dim(sce.dellOrso.pathway)
```


**KEGG module and pathway id:name mappings**
Data available at https://www.genome.jp/dbget-bin/www_bget?pathway:mmu01100
```{r load-kegg-module-and-pathway-csv-files}
kegg.modules <- read.csv("./../saved/kegg_metabolic_modules.csv", header = FALSE)
colnames(kegg.modules) <- c("module", "module_name")

kegg.pathways <- read.csv("./../saved/kegg_metabolic_pathways.csv", header = FALSE)
colnames(kegg.pathways) <- c("pathway", "pathway_name")
```


## 2. Perform aggregation of gene expression into pathway expression
### 2.1 Primary dataset
```{r primary-data-KEGG-module-pathway-aggregation, eval=FALSE}
###################################################################
# Get a list of all metabolic pathways/modules present in the data
###################################################################
all_pathways <- c()
all_modules <- c()

for (pathway_set in rowData(denoised.sce.m)$pathway) {
  new_pathways <- strsplit(pathway_set, split = ", ")[[1]] %>% lapply(trimws) %>% unlist()
  all_pathways <- unique(c(all_pathways, new_pathways))
}

for (module_set in rowData(denoised.sce.m)$module) {
  new_modules <- strsplit(module_set, split = ", ")[[1]] %>% lapply(trimws) %>% unlist()
  all_modules <- unique(c(all_modules, new_modules))
}

################################
# Module expression aggregation
################################
# Create a new data frame for storing aggregated expression values for each module
module.expression.profile <- data.frame()

# Aggregate expression for each module
for (module in all_modules) {
  print(paste(module))
  # Get all genes linked to a module and create a subsetted sce object from these
  module.genes <- grepl(module, rowData(denoised.sce.m)$module, fixed = TRUE)
  sce.module.genes <- denoised.sce.m[module.genes, ]
  # Aggregate expression of module genes and transform into a 1-row dataframe
  module.expression <- t(as.data.frame(colSums(as.matrix(logcounts(sce.module.genes)), na.rm = TRUE), drop=FALSE))
  # Divide the expression of the module in each cell by the number of genes contributing to this module (to get the avg. expression)
  module.expression <- module.expression / nrow(sce.module.genes)
  # Add a new aggregated module expression row into the module.expression.profile dataframe
  module.expression.profile <- rbind(module.expression.profile, module.expression)
}
# Add correct rownames to the expression object
rownames(module.expression.profile) <- all_modules

# Create a SingleCellExperiment object to store the data
sce.primary.module <- SingleCellExperiment(assays = list(logcounts = as.matrix(module.expression.profile)),
                                           colData = colData(denoised.sce.m))


################################
# Pathway expression aggregation
################################
# Create a new data frame for storing aggregated expression values for each module
pathway.expression.profile <- data.frame()

# Aggregate expression for each pathway
for (pathway in all_pathways) {
  print(paste(pathway))
  # Get all genes linked to a pathway and create a subsetted sce object from these
  pathway.genes <- grepl(pathway, rowData(denoised.sce.m)$pathway, fixed = TRUE)
  sce.pathway.genes <- denoised.sce.m[pathway.genes, ]
  # Aggregate expression of pathway genes and transform into a 1-row dataframe
  pathway.expression <- t(as.data.frame(colSums(as.matrix(logcounts(sce.pathway.genes)), na.rm = TRUE), drop = FALSE))
  # Divide the expression of the pathway in each cell by the number of genes contributing to this module (to get the avg. expression)
  pathway.expression <- pathway.expression / nrow(sce.pathway.genes)
  # Add a new aggregated pathway expression row into the pathway.expression.profile dataframe
  pathway.expression.profile <- rbind(pathway.expression.profile, pathway.expression)
}

# Add correct rownames to the expression object
rownames(pathway.expression.profile) <- all_pathways

# Create a SingleCellExperiment object to store the data
sce.primary.pathway <- SingleCellExperiment(assays = list(logcounts = as.matrix(pathway.expression.profile)),
                                            colData = colData(denoised.sce.m))


###########################################
# Save aggregated module & pathway objects
###########################################
saveRDS(sce.primary.module, file = "./../saved/sce_primary_module")
saveRDS(sce.primary.pathway, file = "./../saved/sce_primary_pathway")

# Inspect dimensions of module and pathway objects
dim(sce.primary.module)
dim(sce.primary.pathway)
```

### 2.2 De Micheli dataset
```{r deMicheli-data-KEGG-module-pathway-aggregation, eval=FALSE}
###################################################################
# Get a list of all metabolic pathways/modules present in the data
###################################################################
all_pathways <- c()
all_modules <- c()

for (pathway_set in rowData(denoised.sce.deMicheli.m)$pathway) {
  new_pathways <- strsplit(pathway_set, split = ", ")[[1]] %>% lapply(trimws) %>% unlist()
  all_pathways <- unique(c(all_pathways, new_pathways))
}

for (module_set in rowData(denoised.sce.deMicheli.m)$module) {
  new_modules <- strsplit(module_set, split = ", ")[[1]] %>% lapply(trimws) %>% unlist()
  all_modules <- unique(c(all_modules, new_modules))
}

################################
# Module expression aggregation
################################
# Create a new data frame for storing aggregated expression values for each module
module.expression.profile <- data.frame()

# Aggregate expression for each module
for (module in all_modules) {
  print(paste(module))
  # Get all genes linked to a module and create a subsetted sce object from these
  module.genes <- grepl(module, rowData(denoised.sce.deMicheli.m)$module, fixed = TRUE)
  sce.module.genes <- denoised.sce.deMicheli.m[module.genes, ]
  # Aggregate expression of module genes and transform into a 1-row dataframe
  module.expression <- t(as.data.frame(colSums(as.matrix(logcounts(sce.module.genes)), na.rm = TRUE), drop=FALSE))
  # Divide the expression of the module in each cell by the number of genes contributing to this module (to get the avg. expression)
  module.expression <- module.expression / nrow(sce.module.genes)
  # Add a new aggregated module expression row into the module.expression.profile dataframe
  module.expression.profile <- rbind(module.expression.profile, module.expression)
}
# Add correct rownames to the expression object
rownames(module.expression.profile) <- all_modules

# Create a SingleCellExperiment object to store the data
sce.deMicheli.module <- SingleCellExperiment(assays = list(logcounts = as.matrix(module.expression.profile)),
                                             colData = colData(denoised.sce.deMicheli.m))


################################
# Pathway expression aggregation
################################
# Create a new data frame for storing aggregated expression values for each module
pathway.expression.profile <- data.frame()

# Aggregate expression for each pathway
for (pathway in all_pathways) {
  print(paste(pathway))
  # Get all genes linked to a pathway and create a subsetted sce object from these
  pathway.genes <- grepl(pathway, rowData(denoised.sce.deMicheli.m)$pathway, fixed = TRUE)
  sce.pathway.genes <- denoised.sce.deMicheli.m[pathway.genes, ]
  # Aggregate expression of pathway genes and transform into a 1-row dataframe
  pathway.expression <- t(as.data.frame(colSums(as.matrix(logcounts(sce.pathway.genes)), na.rm = TRUE), drop = FALSE))
  # Divide the expression of the pathway in each cell by the number of genes contributing to this module (to get the avg. expression)
  pathway.expression <- pathway.expression / nrow(sce.pathway.genes)
  # Add a new aggregated pathway expression row into the pathway.expression.profile dataframe
  pathway.expression.profile <- rbind(pathway.expression.profile, pathway.expression)
}

# Add correct rownames to the expression object
rownames(pathway.expression.profile) <- all_pathways

# Create a SingleCellExperiment object to store the data
sce.deMicheli.pathway <- SingleCellExperiment(assays = list(logcounts = as.matrix(pathway.expression.profile)),
                                            colData = colData(denoised.sce.deMicheli.m))


###########################################
# Save aggregated module & pathway objects
###########################################
saveRDS(sce.deMicheli.module, file = "./../saved/sce_deMicheli_module")
saveRDS(sce.deMicheli.pathway, file = "./../saved/sce_deMicheli_pathway")

# Inspect dimensions of module and pathway objects
dim(sce.deMicheli.module)
dim(sce.deMicheli.pathway)
```

### 2.3 Dell'Orso dataset
```{r dellOrso-data-KEGG-module-pathway-aggregation, eval=FALSE}
###################################################################
# Get a list of all metabolic pathways/modules present in the data
###################################################################
all_pathways <- c()
all_modules <- c()

for (pathway_set in rowData(denoised.sce.dellOrso.m)$pathway) {
  new_pathways <- strsplit(pathway_set, split = ", ")[[1]] %>% lapply(trimws) %>% unlist()
  all_pathways <- unique(c(all_pathways, new_pathways))
}

for (module_set in rowData(denoised.sce.dellOrso.m)$module) {
  new_modules <- strsplit(module_set, split = ", ")[[1]] %>% lapply(trimws) %>% unlist()
  all_modules <- unique(c(all_modules, new_modules))
}

################################
# Module expression aggregation
################################
# Create a new data frame for storing aggregated expression values for each module
module.expression.profile <- data.frame()

# Aggregate expression for each module
for (module in all_modules) {
  print(paste(module))
  # Get all genes linked to a module and create a subsetted sce object from these
  module.genes <- grepl(module, rowData(denoised.sce.dellOrso.m)$module, fixed = TRUE)
  sce.module.genes <- denoised.sce.dellOrso.m[module.genes, ]
  # Aggregate expression of module genes and transform into a 1-row dataframe
  module.expression <- t(as.data.frame(colSums(as.matrix(logcounts(sce.module.genes)), na.rm = TRUE), drop=FALSE))
  # Divide the expression of the module in each cell by the number of genes contributing to this module (to get the avg. expression)
  module.expression <- module.expression / nrow(sce.module.genes)
  # Add a new aggregated module expression row into the module.expression.profile dataframe
  module.expression.profile <- rbind(module.expression.profile, module.expression)
}
# Add correct rownames to the expression object
rownames(module.expression.profile) <- all_modules

# Create a SingleCellExperiment object to store the data
sce.dellOrso.module <- SingleCellExperiment(assays = list(logcounts = as.matrix(module.expression.profile)),
                                             colData = colData(denoised.sce.dellOrso.m))


################################
# Pathway expression aggregation
################################
# Create a new data frame for storing aggregated expression values for each module
pathway.expression.profile <- data.frame()

# Aggregate expression for each pathway
for (pathway in all_pathways) {
  print(paste(pathway))
  # Get all genes linked to a pathway and create a subsetted sce object from these
  pathway.genes <- grepl(pathway, rowData(denoised.sce.dellOrso.m)$pathway, fixed = TRUE)
  sce.pathway.genes <- denoised.sce.dellOrso.m[pathway.genes, ]
  # Aggregate expression of pathway genes and transform into a 1-row dataframe
  pathway.expression <- t(as.data.frame(colSums(as.matrix(logcounts(sce.pathway.genes)), na.rm = TRUE), drop = FALSE))
  # Divide the expression of the pathway in each cell by the number of genes contributing to this module (to get the avg. expression)
  pathway.expression <- pathway.expression / nrow(sce.pathway.genes)
  # Add a new aggregated pathway expression row into the pathway.expression.profile dataframe
  pathway.expression.profile <- rbind(pathway.expression.profile, pathway.expression)
}

# Add correct rownames to the expression object
rownames(pathway.expression.profile) <- all_pathways

# Create a SingleCellExperiment object to store the data
sce.dellOrso.pathway <- SingleCellExperiment(assays = list(logcounts = as.matrix(pathway.expression.profile)),
                                            colData = colData(denoised.sce.dellOrso.m))


###########################################
# Save aggregated module & pathway objects
###########################################
saveRDS(sce.dellOrso.module, file = "./../saved/sce_dellOrso_module")
saveRDS(sce.dellOrso.pathway, file = "./../saved/sce_dellOrso_pathway")

# Inspect dimensions of module and pathway objects
dim(sce.dellOrso.module)
dim(sce.dellOrso.pathway)
```


## 3. Add KEGG module and pathway information to the rowData of the sce objects
```{r add-KEGG-module-and-pathway-information-into-the-sce-objects, eval=FALSE}
# Modify the KEGG information tables so that rownames are module/pathway IDs and the df contains only 1 column which is module / pathway name
rownames(kegg.modules) <- kegg.modules$module
kegg.modules <- kegg.modules[, "module_name", drop=FALSE]

rownames(kegg.pathways) <- kegg.pathways$pathway
kegg.pathways <- kegg.pathways[, "pathway_name", drop=FALSE]

# Add module / pathway names to the sce objects
###############
## Primary data
###############
kegg.modules.primary <- kegg.modules[rownames(kegg.modules) %in% rownames(sce.primary.module), , drop=FALSE]
kegg.modules.primary.sorted <- kegg.modules.primary[row.names(sce.primary.module), , drop=FALSE]
rowData(sce.primary.module)$module_name <- kegg.modules.primary.sorted$module_name

kegg.pathways.primary <- kegg.pathways[rownames(kegg.pathways) %in% rownames(sce.primary.pathway), , drop=FALSE]
kegg.pathways.primary.sorted <- kegg.pathways.primary[row.names(sce.primary.pathway), , drop=FALSE]
rowData(sce.primary.pathway)$pathway_name <- kegg.pathways.primary.sorted$pathway_name

##################
## De Micheli data
##################
kegg.modules.deMicheli <- kegg.modules[rownames(kegg.modules) %in% rownames(sce.deMicheli.module), , drop=FALSE]
kegg.modules.deMicheli.sorted <- kegg.modules.deMicheli[row.names(sce.deMicheli.module), , drop=FALSE]
rowData(sce.deMicheli.module)$module_name <- kegg.modules.deMicheli.sorted$module_name

kegg.pathways.deMicheli <- kegg.pathways[rownames(kegg.pathways) %in% rownames(sce.deMicheli.pathway), , drop=FALSE]
kegg.pathways.deMicheli.sorted <- kegg.pathways.deMicheli[row.names(sce.deMicheli.pathway), , drop=FALSE]
rowData(sce.deMicheli.pathway)$pathway_name <- kegg.pathways.deMicheli.sorted$pathway_name

##################
## Dell'Orso data
##################
kegg.modules.dellOrso <- kegg.modules[rownames(kegg.modules) %in% rownames(sce.dellOrso.module), , drop=FALSE]
kegg.modules.dellOrso.sorted <- kegg.modules.dellOrso[row.names(sce.dellOrso.module), , drop=FALSE]
rowData(sce.dellOrso.module)$module_name <- kegg.modules.dellOrso.sorted$module_name

kegg.pathways.dellOrso <- kegg.pathways[rownames(kegg.pathways) %in% rownames(sce.dellOrso.pathway), , drop=FALSE]
kegg.pathways.dellOrso.sorted <- kegg.pathways.dellOrso[row.names(sce.dellOrso.pathway), , drop=FALSE]
rowData(sce.dellOrso.pathway)$pathway_name <- kegg.pathways.dellOrso.sorted$pathway_name


# Overwrite saved sce objects with new objects containing rowData
saveRDS(sce.dellOrso.module, file = "./../saved/sce_dellOrso_module")
saveRDS(sce.dellOrso.pathway, file = "./../saved/sce_dellOrso_pathway")
saveRDS(sce.deMicheli.module, file = "./../saved/sce_deMicheli_module")
saveRDS(sce.deMicheli.pathway, file = "./../saved/sce_deMicheli_pathway")
saveRDS(sce.primary.module, file = "./../saved/sce_primary_module")
saveRDS(sce.primary.pathway, file = "./../saved/sce_primary_pathway")

```

## 4. Calculate module / pathway importances
### 4.1 Load pre-computed module / pathway importance objects
```{r load-pre-saved-module-and-pathway-importances}
#########################
# Primary data - modules
#########################
# Define expression and group objects
expression.primary.module <- as.matrix(t(logcounts(sce.primary.module)))
group_name.primary.module <- as.factor(sce.primary.module$sample)

# Get gimp & module objects
gimp.primary.module <- readRDS("./../saved/gimp_primary_module")
modules.primary.module <- readRDS("./../saved/modules_primary_module")
gimp.primary.module$qvalue <- p.adjust(p = gimp.primary.module$pvalue, method = "BH")
gene_sel.primary.module <- gimp.primary.module$gene[gimp.primary.module$qvalue < .05]
expr_sel.primary.module <- scale_quantile(expression.primary.module[, gene_sel.primary.module])

#########################
# Primary data - pathways
#########################
# Define expression and group objects
expression.primary.pathway <- as.matrix(t(logcounts(sce.primary.pathway)))
group_name.primary.pathway <- as.factor(sce.primary.pathway$sample)

# Get gimp & module objects
gimp.primary.pathway <- readRDS("./../saved/gimp_primary_pathway")
modules.primary.pathway <- readRDS("./../saved/modules_primary_pathway")
gimp.primary.pathway$qvalue <- p.adjust(p = gimp.primary.pathway$pvalue, method = "BH")
gene_sel.primary.pathway <- gimp.primary.pathway$gene[gimp.primary.pathway$qvalue < .05]
expr_sel.primary.pathway <- scale_quantile(expression.primary.pathway[, gene_sel.primary.pathway])

###########################
# De Micheli data - modules
###########################
# Define expression and group objects
expression.deMicheli.module <- as.matrix(t(logcounts(sce.deMicheli.module)))
group_name.deMicheli.module <- as.factor(sce.deMicheli.module$cell_type)

# Get gimp & module objects
gimp.deMicheli.module <- readRDS("./../saved/gimp_deMicheli_module")
modules.deMicheli.module <- readRDS("./../saved/modules_deMicheli_module")
gimp.deMicheli.module$qvalue <- p.adjust(p = gimp.deMicheli.module$pvalue, method = "BH")
gene_sel.deMicheli.module <- gimp.deMicheli.module$gene[gimp.deMicheli.module$qvalue < .05]
expr_sel.deMicheli.module <- scale_quantile(expression.deMicheli.module[, gene_sel.deMicheli.module])

############################
# De Micheli data - pathways
############################
# Define expression and group objects
expression.deMicheli.pathway <- as.matrix(t(logcounts(sce.deMicheli.pathway)))
group_name.deMicheli.pathway <- as.factor(sce.deMicheli.pathway$cell_type)

# Get gimp & module objects
gimp.deMicheli.pathway <- readRDS("./../saved/gimp_deMicheli_pathway")
modules.deMicheli.pathway <- readRDS("./../saved/modules_deMicheli_pathway")
gimp.deMicheli.pathway$qvalue <- p.adjust(p = gimp.deMicheli.pathway$pvalue, method = "BH")
gene_sel.deMicheli.pathway <- gimp.deMicheli.pathway$gene[gimp.deMicheli.pathway$qvalue < .05]
expr_sel.deMicheli.pathway <- scale_quantile(expression.deMicheli.pathway[, gene_sel.deMicheli.pathway])


#############################################################
# De Micheli data - modules (without mature skeletal muscle)
#############################################################
# Remove muscle cells and define expression and group objects
muscle.cells.module <- sce.deMicheli.module$cell_type == "d0_mature_skeletal_muscle" | sce.deMicheli.module$cell_type =="d5_mature_skeletal_muscle" | sce.deMicheli.module$cell_type == "d7_mature_skeletal_muscle"
sce.deMicheli.module.f <- sce.deMicheli.module[, !muscle.cells.module]

expression.deMicheli.module.f <- as.matrix(t(logcounts(sce.deMicheli.module.f)))
group_name.deMicheli.module.f <- as.factor(sce.deMicheli.module.f$cell_type)

# Get gimp & module objects
gimp.deMicheli.module.f <- readRDS("./../saved/gimp_deMicheli_module_no_muscle")
modules.deMicheli.module.f <- readRDS("./../saved/modules_deMicheli_module_no_muscle")
gimp.deMicheli.module.f$qvalue <- p.adjust(p = gimp.deMicheli.module.f$pvalue, method = "BH")
gene_sel.deMicheli.module.f <- gimp.deMicheli.module.f$gene[gimp.deMicheli.module.f$qvalue < .05]
expr_sel.deMicheli.module.f <- scale_quantile(expression.deMicheli.module.f[, gene_sel.deMicheli.module.f])


##############################################################
# De Micheli data - pathways (without mature skeletal muscle)
#############################################################
# Remove muscle cells and define expression and group objects
muscle.cells.pathway <- sce.deMicheli.pathway$cell_type == "d0_mature_skeletal_muscle" | sce.deMicheli.pathway$cell_type =="d5_mature_skeletal_muscle" | sce.deMicheli.pathway$cell_type == "d7_mature_skeletal_muscle"
sce.deMicheli.pathway.f <- sce.deMicheli.pathway[, !muscle.cells.pathway]

expression.deMicheli.pathway.f <- as.matrix(t(logcounts(sce.deMicheli.pathway.f)))
group_name.deMicheli.pathway.f <- as.factor(sce.deMicheli.pathway.f$cell_type)

# Get gimp & module objects
gimp.deMicheli.pathway.f <- readRDS("./../saved/gimp_deMicheli_pathway_no_muscle")
modules.deMicheli.pathway.f <- readRDS("./../saved/modules_deMicheli_pathway_no_muscle")
gimp.deMicheli.pathway.f$qvalue <- p.adjust(p = gimp.deMicheli.pathway.f$pvalue, method = "BH")
gene_sel.deMicheli.pathway.f <- gimp.deMicheli.pathway.f$gene[gimp.deMicheli.pathway.f$qvalue < .05]
expr_sel.deMicheli.pathway.f <- scale_quantile(expression.deMicheli.pathway.f[, gene_sel.deMicheli.pathway.f])


###########################
# Dell'Orso data - modules
###########################
# Define expression and group objects
expression.dellOrso.module <- as.matrix(t(logcounts(sce.dellOrso.module)))
group_name.dellOrso.module <- as.factor(sce.dellOrso.module$cell_type)

# Get gimp & module objects
gimp.dellOrso.module <- readRDS("./../saved/gimp_dellOrso_module")
modules.dellOrso.module <- readRDS("./../saved/modules_dellOrso_module")
gimp.dellOrso.module$qvalue <- p.adjust(p = gimp.dellOrso.module$pvalue, method = "BH")
gene_sel.dellOrso.module <- gimp.dellOrso.module$gene[gimp.dellOrso.module$qvalue < .05]
expr_sel.dellOrso.module <- scale_quantile(expression.dellOrso.module[, gene_sel.dellOrso.module])

############################
# Dell'Orso data - pathways
############################
# Define expression and group objects
expression.dellOrso.pathway <- as.matrix(t(logcounts(sce.dellOrso.pathway)))
group_name.dellOrso.pathway <- as.factor(sce.dellOrso.pathway$cell_type)

# Get gimp & module objects
gimp.dellOrso.pathway <- readRDS("./../saved/gimp_dellOrso_pathway")
modules.dellOrso.pathway <- readRDS("./../saved/modules_dellOrso_pathway")
gimp.dellOrso.pathway$qvalue <- p.adjust(p = gimp.dellOrso.pathway$pvalue, method = "BH")
gene_sel.dellOrso.pathway <- gimp.dellOrso.pathway$gene[gimp.dellOrso.pathway$qvalue < .05]
expr_sel.dellOrso.pathway <- scale_quantile(expression.dellOrso.pathway[, gene_sel.dellOrso.pathway])

########################################
# Dell'Orso data - modules (without PM)
########################################
# Remove primary myoblast cells and define expression and group objects
primary.mbs <- sce.dellOrso.module$cell_type == "Primary_MB"
sce.dellOrso.module.f <- sce.dellOrso.module[, !muscle.cells.module]
expression.dellOrso.module.f <- as.matrix(t(logcounts(sce.dellOrso.module.f)))
group_name.dellOrso.module.f <- as.factor(sce.dellOrso.module.f$cell_type)

# Get gimp & module objects
gimp.dellOrso.module.f <- readRDS("./../saved/gimp_dellOrso_module_no_pm")
modules.dellOrso.module.f <- readRDS("./../saved/modules_dellOrso_module_no_pm")
gimp.dellOrso.module.f$qvalue <- p.adjust(p = gimp.dellOrso.module.f$pvalue, method = "BH")
gene_sel.dellOrso.module.f <- gimp.dellOrso.module.f$gene[gimp.dellOrso.module.f$qvalue < .05]
expr_sel.dellOrso.module.f <- scale_quantile(expression.dellOrso.module.f[, gene_sel.dellOrso.module.f])


#########################################
# Dell'Orso data - pathways (without PM)
#########################################
# Remove primary myoblast cells and define expression and group objects
primary.mbs <- sce.dellOrso.pathway$cell_type == "Primary_MB"
sce.dellOrso.pathway.f <- sce.dellOrso.pathway[, !muscle.cells.pathway]
expression.dellOrso.pathway.f <- as.matrix(t(logcounts(sce.dellOrso.pathway.f)))
group_name.dellOrso.pathway.f <- as.factor(sce.dellOrso.pathway.f$cell_type)

# Get gimp & module objects
gimp.dellOrso.pathway.f <- readRDS("./../saved/gimp_dellOrso_pathway_no_pm")
modules.dellOrso.pathway.f <- readRDS("./../saved/modules_dellOrso_pathway_no_pm")
gimp.dellOrso.pathway.f$qvalue <- p.adjust(p = gimp.dellOrso.pathway.f$pvalue, method = "BH")
gene_sel.dellOrso.pathway.f <- gimp.dellOrso.pathway.f$gene[gimp.dellOrso.pathway.f$qvalue < .05]
expr_sel.dellOrso.pathway.f <- scale_quantile(expression.dellOrso.pathway.f[, gene_sel.dellOrso.pathway.f])
```


### 4.2 Perform computation of module/pathway importances for all datasets (optional)
**These code chunks should only be computed if no pre-computed objects are available**
#### 4.2.1 Primary data - modules
```{r compute-primary-module-importances, eval=FALSE}
# Load required objects (if required)
if(!exists("sce.primary.module")) {sce.primary.module <- readRDS("./../saved/sce_primary_module")}

# Create expression and sample objects
expression.primary.module <- as.matrix(t(logcounts(sce.primary.module)))
group_name.primary.module <- as.factor(sce.primary.module$sample)

# Compute gene importances
gimp.primary.module <- gene_importances(x = expression.primary.module,
                                        time = traj.primary$time,
                                        num_permutations = 10,
                                        num_threads = 8)

gimp.primary.module$qvalue <- p.adjust(p = gimp.primary.module$pvalue, method = "BH")
gene_sel.primary.module <- gimp.primary.module$gene[gimp.primary.module$qvalue < .05]
expr_sel.primary.module <- scale_quantile(expression.primary.module[, gene_sel.primary.module])

saveRDS(gimp.primary.module, file = "./../saved/gimp_primary_module")

# Extract modules
modules.primary.module <- extract_modules(expr_sel.primary.module,
                                          time = traj.primary$time,
                                          verbose = FALSE)

saveRDS(modules.primary.module, file = "./../saved/modules_primary_module")
```

#### 4.2.2 Primary data - pathways
```{r, eval=FALSE}
# TODO: copy from puhti script
```
#### 4.2.3 De Micheli data - modules
```{r, eval=FALSE}
# TODO: copy from puhti script
```
#### 4.2.4 De Micheli data - pathways
```{r, eval=FALSE}
# TODO: copy from puhti script
```
#### 4.2.5 De Micheli data - modules (without mature skeletal muscle)
```{r, eval=FALSE}
# TODO: copy from puhti script
```
#### 4.2.6 De Micheli data - pathways (without mature skeletal muscle)
```{r, eval=FALSE}
# TODO: copy from puhti script
```
#### 4.2.7 Dell'Orso data - modules
```{r, eval=FALSE}
# TODO: copy from puhti script
```
#### 4.2.8 Dell'Orso data - pathways
```{r, eval=FALSE}
# TODO: copy from puhti script
```
#### 4.2.9 Dell'Orso data - modules (without PMs)
```{r, eval=FALSE}
# TODO: copy from puhti script
```
#### 4.2.10 Dell'Orso data - pathways (without PMs)
```{r, eval=FALSE}
# TODO: copy from puhti script
```


### 4.3 Visualization of module / pathway importances
#### 4.3.1 Primary data - modules
```{r trajectory-heatmap-primary-data-modules}
# Define annotation row
ann.row <- as.data.frame(rowData(sce.primary.module)[rownames(sce.primary.module) %in% colnames(expr_sel.primary.module), , drop=FALSE])
ann.row$module_name <- as.factor(ann.row$module_name)

# Plot heatmap
p.primary.modules <- draw_trajectory_heatmap(expr_sel.primary.module, 
                        time = traj.primary$time,
                        progression_group = group_name.primary.module,
                        modules = modules.primary.module,
                        annotation_row = ann.row,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)

p.primary.modules
```

#### 4.3.2 Primary data - pathways
```{r trajectory-heatmap-primary-data-pathways}
# Define annotation row
ann.row <- as.data.frame(rowData(sce.primary.pathway)[rownames(sce.primary.pathway) %in% colnames(expr_sel.primary.pathway), , drop=FALSE])
ann.row$pathway_name <- as.factor(ann.row$pathway_name)

# Plot heatmap
p.primary.pathways <- draw_trajectory_heatmap(expr_sel.primary.pathway, 
                        time = traj.primary$time,
                        progression_group = group_name.primary.pathway,
                        modules = modules.primary.pathway,
                        annotation_row = ann.row,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)

p.primary.pathways
```

#### 4.3.3 De Micheli data - modules
```{r trajectory-heatmap-deMicheli-data-modules}
# Define annotation row
ann.row <- as.data.frame(rowData(sce.deMicheli.module)[rownames(sce.deMicheli.module) %in% colnames(expr_sel.deMicheli.module), , drop=FALSE])
ann.row$module_name <- as.factor(ann.row$module_name)

# Plot heatmap
p.deMicheli.modules <- draw_trajectory_heatmap(expr_sel.deMicheli.module, 
                        time = traj.deMicheli$time,
                        progression_group = group_name.deMicheli.module,
                        modules = modules.deMicheli.module,
                        annotation_row = ann.row,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)

p.deMicheli.modules
```

#### 4.3.4 De Micheli data - pathways
```{r trajectory-heatmap-deMicheli-data-pathways}
# Define annotation row
ann.row <- as.data.frame(rowData(sce.deMicheli.pathway)[rownames(sce.deMicheli.pathway) %in% colnames(expr_sel.deMicheli.pathway), , drop=FALSE])
ann.row$pathway_name <- as.factor(ann.row$pathway_name)

# Plot heatmap
p.deMicheli.pathways <- draw_trajectory_heatmap(expr_sel.deMicheli.pathway, 
                        time = traj.deMicheli$time,
                        progression_group = group_name.deMicheli.pathway,
                        modules = modules.deMicheli.pathway,
                        annotation_row = ann.row,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)

p.deMicheli.pathways
```

#### 4.3.5 De Micheli data - modules (without mature skeletal muscle)
```{r trajectory-heatmap-deMicheli-data-modules-no-muscle}
# define time without mature skeletal muscle
traj.deMicheli.time.module.f <-  traj.deMicheli$time[names(traj.deMicheli$time) %in% colnames(sce.deMicheli.module.f)]

# Define annotation row
ann.row <- as.data.frame(rowData(sce.deMicheli.module.f)[rownames(sce.deMicheli.module.f) %in% colnames(expr_sel.deMicheli.module.f), , drop=FALSE])
ann.row$module_name <- as.factor(ann.row$module_name)

# Draw heatmap
p.deMicheli.modules.f <- draw_trajectory_heatmap(expr_sel.deMicheli.module.f, 
                        time = traj.deMicheli.time.module.f,
                        progression_group = group_name.deMicheli.module.f,
                        modules = modules.deMicheli.module.f,
                        annotation_row = ann.row,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)

p.deMicheli.modules.f
```

#### 4.3.6 De Micheli data - pathways (without mature skeletal muscle)
```{r trajectory-heatmap-deMicheli-data-pathways-no-muscle}
# define time without mature skeletal muscle
traj.deMicheli.time.pathway.f <-  traj.deMicheli$time[names(traj.deMicheli$time) %in% colnames(sce.deMicheli.pathway.f)]

# Define annotation row
ann.row <- as.data.frame(rowData(sce.deMicheli.pathway.f)[rownames(sce.deMicheli.pathway.f) %in% colnames(expr_sel.deMicheli.pathway.f), , drop=FALSE])
ann.row$pathway_name <- as.factor(ann.row$pathway_name)

# Draw heatmap
p.deMicheli.pathways.f <-  draw_trajectory_heatmap(expr_sel.deMicheli.pathway.f, 
                        time = traj.deMicheli.time.pathway.f,
                        progression_group = group_name.deMicheli.pathway.f,
                        modules = modules.deMicheli.pathway.f,
                        annotation_row = ann.row,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)

p.deMicheli.pathways.f
```

#### 4.3.7 Dell'Orso data - modules
```{r trajectory-heatmap-dellOrso-data-modules}
# Define annotation row
ann.row <- as.data.frame(rowData(sce.dellOrso.module)[rownames(sce.dellOrso.module) %in% colnames(expr_sel.dellOrso.module), , drop=FALSE])
ann.row$module_name <- as.factor(ann.row$module_name)

# Draw trajectory heatmap
p.dellOrso.modules <- draw_trajectory_heatmap(expr_sel.dellOrso.module, 
                        time = traj.dellOrso$time,
                        progression_group = group_name.dellOrso.module,
                        modules = modules.dellOrso.module,
                        annotation_row = ann.row,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)

p.dellOrso.modules
```

#### 4.3.8 Dell'Orso data - pathways
```{r trajectory-heatmap-dellOrso-data-pathways}
# Define annotation row
ann.row <- as.data.frame(rowData(sce.dellOrso.pathway)[rownames(sce.dellOrso.pathway) %in% colnames(expr_sel.dellOrso.pathway), , drop=FALSE])
ann.row$pathway_name <- as.factor(ann.row$pathway_name)

# Draw trajectory heatmap
p.dellOrso.pathways <- draw_trajectory_heatmap(expr_sel.dellOrso.pathway, 
                        time = traj.dellOrso$time,
                        progression_group = group_name.dellOrso.pathway,
                        modules = modules.dellOrso.pathway,
                        annotation_row = ann.row,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)

p.dellOrso.pathways
```


#### 4.3.9 Dell'Orso data - modules (without PMs)
```{r trajectory-heatmap-dellOrso-data-modules-no-PMs}
# define time without PMs
traj.dellOrso.time.module.f <-  traj.dellOrso$time[names(traj.dellOrso$time) %in% colnames(sce.dellOrso.module.f)]

# Define annotation row
ann.row <- as.data.frame(rowData(sce.dellOrso.module.f)[rownames(sce.dellOrso.module.f) %in% colnames(expr_sel.dellOrso.module.f), , drop=FALSE])
ann.row$module_name <- as.factor(ann.row$module_name)

# Draw heatmap
p.dellOrso.modules.f <- draw_trajectory_heatmap(expr_sel.dellOrso.module.f, 
                        time = traj.dellOrso.time.module.f,
                        progression_group = group_name.dellOrso.module.f,
                        modules = modules.dellOrso.module.f,
                        annotation_row = ann.row,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)

p.dellOrso.modules.f
```

#### 4.3.10 Dell'Orso data - pathways (without PMs)
```{r trajectory-heatmap-dellOrso-data-pathways-no-PMs}
# define time without PMs
traj.dellOrso.time.pathway.f <-  traj.dellOrso$time[names(traj.dellOrso$time) %in% colnames(sce.dellOrso.pathway.f)]

# Define annotation row
ann.row <- as.data.frame(rowData(sce.dellOrso.pathway.f)[rownames(sce.dellOrso.pathway.f) %in% colnames(expr_sel.dellOrso.pathway.f), , drop=FALSE])
ann.row$pathway_name <- as.factor(ann.row$pathway_name)

# Draw heatmap
p.dellOrso.pathways.f <- draw_trajectory_heatmap(expr_sel.dellOrso.pathway.f, 
                        time = traj.dellOrso.time.pathway.f,
                        progression_group = group_name.dellOrso.pathway.f,
                        modules = modules.dellOrso.pathway.f,
                        annotation_row = ann.row,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)

p.dellOrso.pathways.f
```


#### 4.3.11 Save all heatmaps in pdf format
```{r save-all-heatmaps-in-pdf-format, eval=FALSE}
all.plots <- arrangeGrob(p.primary.modules[[4]],
                         p.primary.pathways[[4]],
                         p.deMicheli.modules[[4]],
                         p.deMicheli.pathways[[4]],
                         p.deMicheli.modules.f[[4]],
                         p.deMicheli.pathways.f[[4]],
                         p.dellOrso.modules[[4]],
                         p.dellOrso.pathways[[4]],
                         p.dellOrso.modules.f[[4]],
                         p.dellOrso.pathways.f[[4]],
                         ncol = 1)

ggsave(filename = "./../saved/module_and_pathway_importances_for_all_datasets.pdf",
       plot = all.plots, limitsize = FALSE, width = 13, height = 100)

```


## 5. Venn diagram of shared metabolic modules / pathways between datasets
```{r comparison-of-metabolic-genes-between-datasets}
# Load metabolic genes if not defined
if(!exists("modules.primary.module")) {modules.primary.module <- readRDS("./../saved/modules_primary_module")}
if(!exists("modules.primary.pathway")) {modules.primary.pathway <- readRDS("./../saved/modules_primary_pathway")}

if(!exists("modules.deMicheli.module")) {modules.deMicheli.module <- readRDS("./../saved/modules_deMicheli_module")}
if(!exists("modules.deMicheli.pathway")) {modules.deMicheli.pathway <- readRDS("./../saved/modules_deMicheli_pathway")}
if(!exists("modules.deMicheli.module.f")) {modules.deMicheli.module.f <- readRDS("./../saved/modules_deMicheli_module_no_muscle")}
if(!exists("modules.deMicheli.pathway.f")) {modules.deMicheli.pathway.f <- readRDS("./../saved/modules_deMicheli_pathway_no_muscle")}

if(!exists("modules.dellOrso.module")) {modules.dellOrso.module <- readRDS("./../saved/modules_dellOrso_module")}
if(!exists("modules.dellOrso.pathway")) {modules.dellOrso.pathway <- readRDS("./../saved/modules_dellOrso_pathway")}
if(!exists("modules.dellOrso.module.f")) {modules.dellOrso.module.f <- readRDS("./../saved/modules_dellOrso_module_no_pm")}
if(!exists("modules.dellOrso.pathway.f")) {modules.dellOrso.pathway.f <- readRDS("./../saved/modules_dellOrso_pathway_no_pm")}

venn.modules <- venn(data = list(primary_data = modules.primary.module$feature,
                                 de_Micheli_data = modules.deMicheli.module$feature,
                                 dellOrso_data = modules.dellOrso.module$feature))

# Print the intersections between datasets
print(attr(venn.modules ,"intersections"))


venn.pathways <- venn(data = list(primary_data = modules.primary.pathway$feature,
                                 de_Micheli_data = modules.deMicheli.pathway$feature,
                                 dellOrso_data = modules.dellOrso.pathway$feature))

# Print the intersections between datasets
print(attr(venn.pathways ,"intersections"))


venn.modules.f <- venn(data = list(primary_data = modules.primary.module$feature,
                                 de_Micheli_data = modules.deMicheli.module.f$feature,
                                 dellOrso_data = modules.dellOrso.module.f$feature))

# Print the intersections between datasets
print(attr(venn.modules.f ,"intersections"))


venn.pathways.f <- venn(data = list(primary_data = modules.primary.pathway$feature,
                                 de_Micheli_data = modules.deMicheli.pathway.f$feature,
                                 dellOrso_data = modules.dellOrso.pathway.f$feature))

# Print the intersections between datasets
print(attr(venn.pathways.f ,"intersections"))
```
