---
title: "Metabolic_pathways_pt2"
author: "Ville Lehtonen"
date: "11/9/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Set working dir to the source file location
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
```

# Metabolic pathways
Here, we perform the aggregation of gene expression values into KEGG module / pathway expression and perform various types of analyses on this aggregated dataset.

## 1. Load packages and data
### 1.1 Load packages
```{r load-packages, message=FALSE}
library(SCORPIUS)
library(scater)
library(scran)
library(SingleCellExperiment)
library(dplyr)
library(SingleR)
library(igraph)
library(stats)
library(cluster)
library(pheatmap)
library(fossil)
library(Seurat)
library(cerebroApp)
library(reshape)
library(data.table)
library(scRNAseq)
library(org.Mm.eg.db)
library(tidyr)
library(BBmisc)
library(KEGGREST)
library(RColorBrewer)
library(ggpubr)
library(grid)
library(gridExtra)
library(plyr)
library(stringr)
library(ggrepel)
library(VennDiagram)
```

### 1.2 Load pre-calculated data
**Single Cell Experiments**
```{r load-sce-objects}
# Load datasets with metabolic genes only
denoised.sce.m <- readRDS("./../saved/denoised_sce_m")
denoised.sce.deMicheli.m <- readRDS("./../saved/denoised_sce_deMicheli_m")
denoised.sce.dellOrso.m <- readRDS("./../saved/denoised_sce_dellOrso_m")
```

**Trajectories**
```{r load-trajectories}
traj.primary <- readRDS("./../saved/traj_primary")
```



## 2. Perform aggregation of gene expression into pathway expression
### 2.1 Primary dataset
```{r primary-data-KEGG-module-pathway-aggregation}
###################################################################
# Get a list of all metabolic pathways/modules present in the data
###################################################################
all_pathways <- c()
all_modules <- c()

for (pathway_set in rowData(denoised.sce.m)$pathway) {
  new_pathways <- strsplit(pathway_set, split = ", ")[[1]] %>% lapply(trimws) %>% unlist()
  all_pathways <- unique(c(all_pathways, new_pathways))
}

for (module_set in rowData(denoised.sce.m)$module) {
  new_modules <- strsplit(module_set, split = ", ")[[1]] %>% lapply(trimws) %>% unlist()
  all_modules <- unique(c(all_modules, new_modules))
}

################################
# Module expression aggregation
################################
# Create a new data frame for storing aggregated expression values for each module
module.expression.profile <- data.frame()

# Aggregate expression for each module
for (module in all_modules) {
  print(paste(module))
  # Get all genes linked to a module and create a subsetted sce object from these
  module.genes <- grepl(module, rowData(denoised.sce.m)$module, fixed = TRUE)
  sce.module.genes <- denoised.sce.m[module.genes, ]
  # Aggregate expression of module genes and transform into a 1-row dataframe
  module.expression <- t(as.data.frame(colSums(as.matrix(logcounts(sce.module.genes)), na.rm = TRUE), drop=FALSE))
  # Add a new aggregated module expression row into the module.expression.profile dataframe
  module.expression.profile <- rbind(module.expression.profile, module.expression)
}
# Add correct rownames to the expression object
rownames(module.expression.profile) <- all_modules

# Create a SingleCellExperiment object to store the data
sce.primary.module <- SingleCellExperiment(assays = list(logcounts = as.matrix(module.expression.profile)),
                                           colData = colData(denoised.sce.m))


################################
# Pathway expression aggregation
################################
# Create a new data frame for storing aggregated expression values for each module
pathway.expression.profile <- data.frame()

# Aggregate expression for each pathway
for (pathway in all_pathways) {
  print(paste(pathway))
  # Get all genes linked to a pathway and create a subsetted sce object from these
  pathway.genes <- grepl(pathway, rowData(denoised.sce.m)$pathway, fixed = TRUE)
  sce.pathway.genes <- denoised.sce.m[pathway.genes, ]
  # Aggregate expression of pathway genes and transform into a 1-row dataframe
  pathway.expression <- t(as.data.frame(colSums(as.matrix(logcounts(sce.pathway.genes)), na.rm = TRUE), drop = FALSE))
  # Add a new aggregated pathway expression row into the pathway.expression.profile dataframe
  pathway.expression.profile <- rbind(pathway.expression.profile, pathway.expression)
}

# Add correct rownames to the expression object
rownames(pathway.expression.profile) <- all_pathways

# Create a SingleCellExperiment object to store the data
sce.primary.pathway <- SingleCellExperiment(assays = list(logcounts = as.matrix(pathway.expression.profile)),
                                            colData = colData(denoised.sce.m))


###########################################
# Save aggregated module & pathway objects
###########################################
saveRDS(sce.primary.module, file = "./../saved/sce_primary_module")
saveRDS(sce.primary.pathway, file = "./../saved/sce_primary_pathway")
```

### 2.2 De Micheli dataset
```{r}

#TODO

```

### 2.3 Dell'Orso dataset
```{r}

#TODO

```


## 3. Calculate module / pathway importances
### 3.1 Primary data
#### 3.1.1 Module importances
```{r}
# Load required objects (if required)
if(!exists("sce.primary.module")) {sce.primary.module <- readRDS("./../saved/sce_primary_module")}

# Create expression and sample objects
expression.primary.module <- as.matrix(t(logcounts(sce.primary.module)))
group_name.primary.module <- as.factor(sce.primary.module$sample)

# Compute gene importances
gimp.primary.module <- gene_importances(x = expression.primary.module,
                                        time = traj.primary$time,
                                        num_permutations = 10,
                                        num_threads = 8)

gimp.primary.module$qvalue <- p.adjust(p = gimp.primary.module$pvalue, method = "BH")
gene_sel.primary.module <- gimp.primary.module$gene[gimp.primary.module$qvalue < .05]
expr_sel.primary.module <- scale_quantile(expression.primary.module[, gene_sel.primary.module])

saveRDS(gimp.primary.module, file = "./../saved/gimp_primary_module")

# Extract modules
modules.primary.module <- extract_modules(expr_sel.primary.module,
                                          time = traj.primary$time,
                                          verbose = FALSE)

saveRDS(modules.primary.module, file = "./../saved/modules_primary_module")
```

#### 3.1.2 Pathway importances
```{r}

#TODO

```


### 3.2 De Micheli data
#### 3.2.1 Module importances
```{r}

#TODO

```

#### 3.2.2 Pathway importances
```{r}

#TODO

```


### 3.3 Dell'Orso data
#### 3.3.1 Module importances
```{r}

#TODO

```

#### 3.3.2 Pathway importances
```{r}

#TODO

```



