---
title: "Metabolic_pathways"
author: "Ville Lehtonen"
date: "11/3/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Set working dir to the source file location
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
```

# Metabolic pathways
Here, we inspect the different metabolic modules / pathways that cells adopt as they progress through the differentiation trajectory. This is done by:
 - Fetching metabolic module / pathway information from KEGG and creating new sce objects containing only metabolic genes and metabolic module and pathway information on their rowData
 - Using Scorpius to calculate gene importances and grouping genes into modules using only a subset of metabolic genes
 - Using singleR to annotate cells using only a subset of metabolic genes
 - Performing DE analysis using only metabolic genes

The aggregation of gene expression values into modules / pathways, and the subsequent computing of module/pathway importances etc. are separated into another file (Metablic_pathways_pt2.Rmd)

## 1. Load packages and data
### 1.1 Load packages
```{r load-packages, message=FALSE}
# TODO: Remove unused libraries
library(SCORPIUS)
library(scater)
library(scran)
library(SingleCellExperiment)
library(dplyr)
library(SingleR)
library(igraph)
library(stats)
library(cluster)
library(pheatmap)
library(fossil)
library(Seurat)
library(reshape)
library(data.table)
library(scRNAseq)
library(org.Mm.eg.db)
library(tidyr)
library(BBmisc)
library(KEGGREST)
library(RColorBrewer)
library(ggpubr)
library(grid)
library(gridExtra)
library(plyr)
library(stringr)
library(ggrepel)
library(VennDiagram)
```

### 1.2 Load pre-calculated data
**Single Cell Experiments**
```{r load-objects}
# Load pre-processed sce objects
denoised.sce <- readRDS("./../saved/denoised_sce")
denoised.sce.deMicheli <- readRDS("./../saved/denoised_sce_deMicheli")
denoised.sce.dellOrso <- readRDS("./../saved/denoised_sce_dellOrso")

# Load datasets with metabolic genes only
denoised.sce.m <- readRDS("./../saved/denoised_sce_m")
denoised.sce.deMicheli.m <- readRDS("./../saved/denoised_sce_deMicheli_m")
denoised.sce.dellOrso.m <- readRDS("./../saved/denoised_sce_dellOrso_m")

# Define group names for trajectories
group_name.primary.m <- as.factor(denoised.sce.m$sample)
group_name.deMicheli.m <- as.factor(denoised.sce.deMicheli.m$cell_type)
group_name.dellOrso.m <- as.factor(denoised.sce.dellOrso.m$cell_type)

```

**Trajectories**
```{r load-pre-calculated-trajectories}
# Load pre-calculated trajectory objects
traj.primary <- readRDS("./../saved/traj_primary")
traj.deMicheli <- readRDS("./../saved/traj_deMicheli")
traj.dellOrso <- readRDS("./../saved/traj_dellOrso")
```


### 1.2 Build trajectories for datasets (Optional)
#### 1.2.1 Primary data
```{r trajectory-primary-data, eval=FALSE}
# Create expression and sample objects
expression.primary <- as.matrix(t(logcounts(denoised.sce)))
group_name.primary <- as.factor(denoised.sce$sample)

# Trajectory inference
set.seed(123)
space.primary <- reduce_dimensionality(expression.primary, dist = "spearman", ndim = 2)
set.seed(123)
traj.primary <- infer_trajectory(space.primary)
traj.primary <- reverse_trajectory(traj.primary)
```

#### 1.2.2 De Micheli data
```{r trajectory-deMicheli, eval=FALSE}
# Create expression and sample objects
expression.deMicheli <- as.matrix(t(logcounts(denoised.sce.deMicheli)))
group_name.deMicheli <- as.factor(denoised.sce.deMicheli$cell_type)

# Trajectory building
set.seed(123)
space.deMicheli <- reduce_dimensionality(expression.deMicheli, dist = "spearman", ndim = 2)
set.seed(123)
traj.deMicheli <- infer_trajectory(space.deMicheli)
traj.deMicheli <- reverse_trajectory(traj.deMicheli)
```

#### 1.2.3 Dell'Orso data
```{r trajectory-dellOrso, eval=FALSE}
# Create expression and sample objects
expression.dellOrso <- as.matrix(t(logcounts(denoised.sce.dellOrso)))
group_name.dellOrso <- as.factor(denoised.sce.dellOrso$cell_type)

# Trajectory building
set.seed(123)
space.dellOrso <- reduce_dimensionality(expression.dellOrso, dist = "spearman", ndim = 2)
set.seed(123)
traj.dellOrso <- infer_trajectory(space.dellOrso)
traj.dellOrso <- reverse_trajectory(traj.dellOrso)
```


## 2. Mapping of KEGG modules & pathways to the datasets

### 2.1 Fetch KEGG modules & pathway information
#### 2.1.2 Fetch KEGG information from pre-computed files
```{r load-fetched-KEGG-information}
# Fetch KEGG entries information
df.kegg.entries <- readRDS("./../saved/df_kegg_entries")
```

#### 2.1.2 Fetch information from KEGG (optional)
**Get all metabolic pathway associated KEGG modules and pathways**
The data is available at https://www.genome.jp/dbget-bin/www_bget?pathway:mmu01100

```{r module-pathway-information-from-KEGG}
kegg.modules <- read.csv("./../saved/kegg_metabolic_modules.csv", header = FALSE)
colnames(kegg.modules) <- c("module", "module_name")

kegg.pathways <- read.csv("./../saved/kegg_metabolic_pathways.csv", header = FALSE)
colnames(kegg.pathways) <- c("pathway", "pathway_name")

```

**Fetch entry, gene, module, and pathway information from KEGG**
```{r fetch-kegg-information, eval=FALSE, message=FALSE}
# Fetch all the modules (and KEGG entries) for mouse
mouse.org.id <- "T01002"
mouse.modules <- keggLink(target = "module", source = mouse.org.id)

# Filter non-metabolic-module associated entries
mouse.modules <- mouse.modules[substring(mouse.modules, 4) %in% kegg.modules$module]

# Build a dataframe to store module, gene, and pathway information for each KEGG entry
df.kegg.entries <- data.frame(entry = names(mouse.modules),
                              module = NA,
                              genes = NA,
                              pathway = NA,
                              module_info = NA,
                              pathway_info = NA)

# For each KEGG entry, fetch the information with keggGet()
for (i in 1:nrow(df.kegg.entries)) {
  print(i)
  entry <- df.kegg.entries[i,]$entry
  query <- keggGet(entry)
  genes <- query[[1]]$NAME
  # add module IDs
  module.list <- names(query[[1]]$MODULE)
  module.list <- module.list[module.list %in% kegg.modules$module]
  module <- paste(module.list, collapse = ", ")
  # add pathway IDs
  pathway.list <- names(query[[1]]$PATHWAY)
  pathway.list <- pathway.list[pathway.list %in% kegg.pathways$pathway]
  pathway <- paste(pathway.list, collapse = ", ")
  # Add module information
  module_info.list <- query[[1]]$MODULE
  module_info.list <- module_info.list[names(module_info.list) %in% module.list]
  module_info <- paste(module_info.list, collapse = ", ")
  # Add pathway information
  pathway_info.list <- query[[1]]$PATHWAY
  pathway_info.list <- pathway_info.list[names(pathway_info.list) %in% pathway.list]
  pathway_info <- paste(pathway_info.list, collapse = ", ")
  # Update information to dataframe
  df.kegg.entries[i,]$genes <- genes
  df.kegg.entries[i,]$module <- module
  df.kegg.entries[i,]$pathway <- pathway
  df.kegg.entries[i,]$module_info <- module_info
  df.kegg.entries[i,]$pathway_info <- pathway_info
}

# Save objects
saveRDS(df.kegg.entries, file = "./../saved/df_kegg_entries")
```


### 2.2 Add KEGG modules & pathway information to the datasets (Optional)
#### 2.2.1 Primary data
```{r add-KEGG-information-to-primary-data, eval=FALSE, message=FALSE}
# Load mouse modules if it doesn't exist
if(!exists("df.kegg.entries")) {
  df.kegg.entries <- readRDS("./../saved/df_kegg_entries")
}

denoised.sce.m <- denoised.sce
rowData(denoised.sce.m)$module <- ""
rowData(denoised.sce.m)$pathway <- ""
rowData(denoised.sce.m)$module_info <- ""
rowData(denoised.sce.m)$pathway_info <- ""

for(i in 1:nrow(df.kegg.entries)) {
  print(i)
  # Split the gene names into a list
  genes <- unlist(lapply(strsplit(df.kegg.entries[i,]$genes, split = ",")[[1]], trimws))
  useful_genes <- genes[genes %in% rownames(denoised.sce.m)]
  for(gene in useful_genes) {
    rowData(denoised.sce.m)[gene, ]$module <- df.kegg.entries[i,]$module
    rowData(denoised.sce.m)[gene, ]$pathway <- df.kegg.entries[i,]$pathway
    rowData(denoised.sce.m)[gene, ]$module_info <- df.kegg.entries[i,]$module_info
    rowData(denoised.sce.m)[gene, ]$pathway_info <- df.kegg.entries[i,]$pathway_info
  }
}

genes.to.use <- rownames(denoised.sce.m)[rowData(denoised.sce.m)$module != ""]
denoised.sce.m <- denoised.sce.m[rownames(denoised.sce.m) %in% genes.to.use, ]

# Save sce object containing only metabolic genes and module/pathway information stored on rowData
saveRDS(denoised.sce.m, "./../saved/denoised_sce_m")
```

#### 2.2.2 De Micheli et al. (2020) reference data
```{r add-KEGG-information-to-deMicheli-data, eval=FALSE, message=FALSE}
# Load mouse modules if it doesn't exist
if(!exists("df.kegg.entries")) {
  df.kegg.entries <- readRDS("./../saved/df_kegg_entries")
}

denoised.sce.deMicheli.m <- denoised.sce.deMicheli
rowData(denoised.sce.deMicheli.m)$module <- ""
rowData(denoised.sce.deMicheli.m)$pathway <- ""
rowData(denoised.sce.deMicheli.m)$module_info <- ""
rowData(denoised.sce.deMicheli.m)$pathway_info <- ""

for(i in 1:nrow(df.kegg.entries)) {
  print(i)
  # Split the gene names into a list
  genes <- unlist(lapply(strsplit(df.kegg.entries[i,]$genes, split = ",")[[1]], trimws))
  useful_genes <- genes[genes %in% rownames(denoised.sce.deMicheli.m)]
  for(gene in useful_genes) {
    rowData(denoised.sce.deMicheli.m)[gene, ]$module <- df.kegg.entries[i,]$module
    rowData(denoised.sce.deMicheli.m)[gene, ]$pathway <- df.kegg.entries[i,]$pathway
    rowData(denoised.sce.deMicheli.m)[gene, ]$module_info <- df.kegg.entries[i,]$module_info
    rowData(denoised.sce.deMicheli.m)[gene, ]$pathway_info <- df.kegg.entries[i,]$pathway_info
  }
}

genes.to.use <- rownames(denoised.sce.deMicheli.m)[rowData(denoised.sce.deMicheli.m)$module != ""]
denoised.sce.deMicheli.m <- denoised.sce.deMicheli.m[rownames(denoised.sce.deMicheli.m) %in% genes.to.use, ]

# Inspect number of genes
dim(denoised.sce.deMicheli.m)

# Save sce object containing only metabolic genes and module/pathway information stored on rowData
saveRDS(denoised.sce.deMicheli.m, "./../saved/denoised_sce_deMicheli_m")
```


#### 2.2.3 Dell'Orso et al. (2019) reference data
```{r add-KEGG-information-to-dellOrso-data, eval=FALSE, message=FALSE}
# Load mouse modules if it doesn't exist
if(!exists("df.kegg.entries")) {
  df.kegg.entries <- readRDS("./../saved/df_kegg_entries")
}

denoised.sce.dellOrso.m <- denoised.sce.dellOrso
rowData(denoised.sce.dellOrso.m)$module <- ""
rowData(denoised.sce.dellOrso.m)$pathway <- ""
rowData(denoised.sce.dellOrso.m)$module_info <- ""
rowData(denoised.sce.dellOrso.m)$pathway_info <- ""

for(i in 1:nrow(df.kegg.entries)) {
  print(i)
  # Split the gene names into a list
  genes <- unlist(lapply(strsplit(df.kegg.entries[i,]$genes, split = ",")[[1]], trimws))
  useful_genes <- genes[genes %in% rownames(denoised.sce.dellOrso.m)]
  for(gene in useful_genes) {
    rowData(denoised.sce.dellOrso.m)[gene, ]$module <- df.kegg.entries[i,]$module
    rowData(denoised.sce.dellOrso.m)[gene, ]$pathway <- df.kegg.entries[i,]$pathway
    rowData(denoised.sce.dellOrso.m)[gene, ]$module_info <- df.kegg.entries[i,]$module_info
    rowData(denoised.sce.dellOrso.m)[gene, ]$pathway_info <- df.kegg.entries[i,]$pathway_info
  }
}

genes.to.use <- rownames(denoised.sce.dellOrso.m)[rowData(denoised.sce.dellOrso.m)$module != ""]
denoised.sce.dellOrso.m <- denoised.sce.dellOrso.m[rownames(denoised.sce.dellOrso.m) %in% genes.to.use, ]

# Inspect number of genes
dim(denoised.sce.dellOrso.m)

# Save sce object containing only metabolic genes and module/pathway information stored on rowData
saveRDS(denoised.sce.dellOrso.m, "./../saved/denoised_sce_dellOrso_m")
```


## 3. Gene importances using only metabolic genes
### 3.1 Load pre-computed gene importances for all datasets
```{r load-pre-computed-gimp-and-module-objects-for-all-datasets}
################
# Primary data
################
# Load gimp & modules
gimp.primary.m <- readRDS("./../saved/gimp_primary_m")
modules.primary.m <- readRDS("./../saved/modules_primary_m")

gimp.primary.m$qvalue <- p.adjust(p = gimp.primary.m$pvalue, method = "BH")
gene_sel.primary.m <- gimp.primary.m$gene[gimp.primary.m$qvalue < .05]

if(!exists("expression.primary.m")) {expression.primary.m <- as.matrix(t(logcounts(denoised.sce.m)))}
expr_sel.primary.m <- scale_quantile(expression.primary.m[, gene_sel.primary.m])

#############################
# De Micheli data: all cells
#############################
# Load gimp & modules
gimp.deMicheli.m <- readRDS("./../saved/gimp_deMicheli_m")
modules.deMicheli.m <- readRDS("./../saved/modules_deMicheli_m")

gimp.deMicheli.m$qvalue <- p.adjust(p = gimp.deMicheli.m$pvalue, method = "BH")
gene_sel.deMicheli.m <- gimp.deMicheli.m$gene[gimp.deMicheli.m$qvalue < .05]

if(!exists("expression.deMicheli.m")) {expression.deMicheli.m <- as.matrix(t(logcounts(denoised.sce.deMicheli.m)))}
expr_sel.deMicheli.m <- scale_quantile(expression.deMicheli.m[, gene_sel.deMicheli.m])

#################################################
# De Micheli data: without mature skeletal muscle
#################################################
# Load gimp & modules
gimp.deMicheli.m.f <- readRDS("./../saved/gimp_deMicheli_m_no_muscle")
modules.deMicheli.m.f <- readRDS("./../saved/modules_deMicheli_m_no_muscle")

gimp.deMicheli.m.f$qvalue <- p.adjust(p = gimp.deMicheli.m.f$pvalue, method = "BH")
gene_sel.deMicheli.m.f <- gimp.deMicheli.m.f$gene[gimp.deMicheli.m.f$qvalue < .05]

if(!exists("expression.deMicheli.m.f")) {
  muscle_cells.m <- denoised.sce.deMicheli.m$cell_type == "d0_mature_skeletal_muscle" | denoised.sce.deMicheli.m$cell_type =="d5_mature_skeletal_muscle" | denoised.sce.deMicheli.m$cell_type == "d7_mature_skeletal_muscle"
  denoised.sce.deMicheli.m.f <- denoised.sce.deMicheli.m[, !muscle_cells.m]
  expression.deMicheli.m.f <- as.matrix(t(logcounts(denoised.sce.deMicheli.m.f)))
  }
expr_sel.deMicheli.m.f <- scale_quantile(expression.deMicheli.m.f[, gene_sel.deMicheli.m.f])
group_name.deMicheli.m.f <- as.factor(denoised.sce.deMicheli.m.f$cell_type)
traj.deMicheli.f.time <- traj.deMicheli$time[names(traj.deMicheli$time) %in% colnames(denoised.sce.deMicheli.m.f)]

  
#############################
# Dell'Orso data: all cells
#############################
# Read gimp & modules from a file
gimp.dellOrso.m <- readRDS("./../saved/gimp_dellOrso_m")
modules.dellOrso.m <- readRDS("./../saved/modules_dellOrso_m")

# Create expression and sample objects
gimp.dellOrso.m$qvalue <- p.adjust(p = gimp.dellOrso.m$pvalue, method = "BH")
gene_sel.dellOrso.m <- gimp.dellOrso.m$gene[gimp.dellOrso.m$qvalue < .05]

if(!exists("expression.dellOrso.m")) {expression.dellOrso.m <- as.matrix(t(logcounts(denoised.sce.dellOrso.m)))}
expr_sel.dellOrso.m <- scale_quantile(expression.dellOrso.m[, gene_sel.dellOrso.m])

###########################################
# Dell'Orso data: without primary myoblasts
###########################################
# Read gimp & modules from a file
gimp.dellOrso.m.f <- readRDS("./../saved/gimp_dellOrso_m_no_primaryMB")
modules.dellOrso.m.f <- readRDS("./../saved/modules_dellOrso_m_no_primaryMB")

# Create expression and sample objects
gimp.dellOrso.m.f$qvalue <- p.adjust(p = gimp.dellOrso.m.f$pvalue, method = "BH")
gene_sel.dellOrso.m.f <- gimp.dellOrso.m.f$gene[gimp.dellOrso.m.f$qvalue < .05]

if(!exists("expression.dellOrso.m.f")) {
  primary_mbs.m <- denoised.sce.dellOrso.m$cell_type == "Primary_MB"
  denoised.sce.dellOrso.m.f <- denoised.sce.dellOrso.m[, !primary_mbs.m]
  expression.dellOrso.m.f <- as.matrix(t(logcounts(denoised.sce.dellOrso.m.f)))
  }
expr_sel.dellOrso.m.f <- scale_quantile(expression.dellOrso.m.f[, gene_sel.dellOrso.m.f])
group_name.dellOrso.m.f <- as.factor(denoised.sce.dellOrso.m.f$cell_type)
traj.dellOrso.f.time <- traj.dellOrso$time[names(traj.dellOrso$time) %in% colnames(denoised.sce.dellOrso.m.f)]

```


### 3.1 Primary data
### 3.1.2 Compute gene importances
```{r gimp-modules-primary-metabolic-genes, eval=FALSE}
# Read trajectory from a file (if not defined)
if(!exists("traj.primary")) {
  traj.primary <- readRDS("./../saved/traj_primary")
}

# Create expression and sample objects
expression.primary.m <- as.matrix(t(logcounts(denoised.sce.m)))
group_name.primary.m <- as.factor(denoised.sce.m$sample)

# Gene importances
gimp.primary.m <- gene_importances(x = expression.primary.m,
                                           time = traj.primary$time,
                                           num_permutations = 10,
                                           num_threads = 8)

gimp.primary.m$qvalue <- p.adjust(p = gimp.primary.m$pvalue, method = "BH")
gene_sel.primary.m <- gimp.primary.m$gene[gimp.primary.m$qvalue < .05]
expr_sel.primary.m <- scale_quantile(expression.primary.m[, gene_sel.primary.m])

#saveRDS(gimp.primary.m, file = "./../saved/gimp_primary_m")

# Extract modules
modules.primary.m <- extract_modules(expr_sel.primary.m,
                                     time = traj.primary$time,
                                     verbose = FALSE)
#saveRDS(modules.primary.m, "./../saved/modules_primary_m")
```


### 3.1.3 Visualize gene importances
**Make a trajectory heatmap**
```{r primary-data-gimp-modules-metabolic-genes}
draw_trajectory_heatmap(expr_sel.primary.m, 
                        time = traj.primary$time,
                        progression_group = group_name.primary.m,
                        modules = modules.primary.m,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
```

**Save as pdf**
```{r primary-data-gimp-modules-metabolic-genes-pdf, eval=FALSE}
# Gimp
pdf("./../saved/gimp_heatmap_primary_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.primary.m, 
                        time = traj.primary$time,
                        progression_group = group_name.primary.m,
                        show_labels_row = TRUE,
                        fontsize_row = 5)
dev.off()

# Modules
pdf("./../saved/modules_heatmap_primary_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.primary.m, 
                        time = traj.primary$time,
                        progression_group = group_name.primary.m,
                        modules = modules.primary.m,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()
```


### 3.2 De Micheli data
#### 3.2.1 Gimp: All cell types
##### 3.2.1.1 Compute gene importances
```{r gimp-modules-deMicheli-metabolic-genes, eval=FALSE}
# Read trajectory from a file (if not defined)
if(!exists("traj.deMicheli")) {
  traj.deMicheli <- readRDS("./../saved/traj_deMicheli")
}

# Create expression and sample objects
expression.deMicheli.m <- as.matrix(t(logcounts(denoised.sce.deMicheli.m)))
group_name.deMicheli.m <- as.factor(denoised.sce.deMicheli.m$cell_type)

# Gene importances
gimp.deMicheli.m <- gene_importances(x = expression.deMicheli.m,
                                     time = traj.deMicheli$time,
                                     num_permutations = 10,
                                     num_threads = 8)

gimp.deMicheli.m$qvalue <- p.adjust(p = gimp.deMicheli.m$pvalue, method = "BH")
gene_sel.deMicheli.m <- gimp.deMicheli.m$gene[gimp.deMicheli.m$qvalue < .05]
expr_sel.deMicheli.m <- scale_quantile(expression.deMicheli.m[, gene_sel.deMicheli.m])

#saveRDS(gimp.deMicheli.m, file = "./../saved/gimp_deMicheli_m")

# Extract modules
modules.deMicheli.m <- extract_modules(expr_sel.deMicheli.m,
                                       time = traj.deMicheli$time,
                                       verbose = FALSE)

#saveRDS(modules.deMicheli.m, "./../saved/modules_deMicheli_m")
```

##### 3.2.1.2 Visualize gene importances
**Make a trajectory heatmap**
```{r deMicheli-gimp-modules-metabolic-genes}
draw_trajectory_heatmap(expr_sel.deMicheli.m, 
                        time = traj.deMicheli$time,
                        progression_group = group_name.deMicheli.m,
                        modules = modules.deMicheli.m,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
```

**Save as pdf**
```{r deMicheli-gimp-modules-metabolic-genes-pdf, eval=FALSE}
# Gimp
pdf("./../saved/gimp_heatmap_deMicheli_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli.m, 
                        time = traj.deMicheli$time,
                        progression_group = group_name.deMicheli.m,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Modules
pdf("./../saved/modules_heatmap_deMicheli_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli.m, 
                        time = traj.deMicheli$time,
                        progression_group = group_name.deMicheli.m,
                        modules = modules.deMicheli.m,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()
```

#### 3.2.2 Gimp: Without mature skeletal muscle
##### 3.2.2.1 Compute gene importances
```{r gimp-modules-deMicheli-metabolic-genes-no-muscle, eval=FALSE}

#TODO (copy from Puhti script)

```


##### 3.2.2.2 Visualize gene importances
**Make a trajectory heatmap**
```{r visualization-deMicheli-gimp-without-skeletal-muscle}
draw_trajectory_heatmap(expr_sel.deMicheli.m.f, 
                        time = traj.deMicheli.f.time,
                        progression_group = group_name.deMicheli.m.f,
                        modules = modules.deMicheli.m.f,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
```


**Save as pdf**
```{r visualization-deMicheli-gimp-without-skeletal-muscle-pdf, eval=FALSE}
# Gimp
pdf("./../saved/gimp_heatmap_deMicheli_metabolic_genes_no_muscle.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli.m.f, 
                        time = traj.deMicheli.f.time,
                        progression_group = group_name.deMicheli.m.f,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Modules
pdf("./../saved/modules_heatmap_deMicheli_metabolic_genes_no_muscle.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli.m.f, 
                        time = traj.deMicheli.f.time,
                        progression_group = group_name.deMicheli.m.f,
                        modules = modules.deMicheli.m.f.kegg,
                        annotation_row = ann_row_deMicheli,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()
```


### 3.3 Dell'Orso data
#### 3.3.1 Gimp: All cell types
##### 3.3.1.1 Compute gene importances
```{r gimp-modules-dellOrso-metabolic-genes, eval=FALSE}
# Read trajectory from a file (if not defined)
if(!exists("traj.dellOrso")) {
  traj.dellOrso <- readRDS("./../saved/traj_dellOrso")
}

# Create expression and sample objects
expression.dellOrso.m <- as.matrix(t(logcounts(denoised.sce.dellOrso.m)))
group_name.dellOrso.m <- as.factor(denoised.sce.dellOrso.m$cell_type)

# Gene importances
gimp.dellOrso.m <- gene_importances(x = expression.dellOrso.m,
                                    time = traj.dellOrso$time,
                                    num_permutations = 10,
                                    num_threads = 8)

gimp.dellOrso.m$qvalue <- p.adjust(p = gimp.dellOrso.m$pvalue, method = "BH")
gene_sel.dellOrso.m <- gimp.dellOrso.m$gene[gimp.dellOrso.m$qvalue < .05]
expr_sel.dellOrso.m <- scale_quantile(expression.dellOrso.m[, gene_sel.dellOrso.m])

#saveRDS(gimp.dellOrso.m, file = "./../saved/gimp_dellOrso_m")

# Extract modules
modules.dellOrso.m <- extract_modules(expr_sel.dellOrso.m,
                                      time = traj.dellOrso$time,
                                      verbose = FALSE)

#saveRDS(modules.dellOrso.m, "./../saved/modules_dellOrso_m")
```

##### 3.3.1.2 Visualize gene importances
**Make a trajectory heatmap**
```{r dellOrso-gimp-modules-metabolic-genes}
draw_trajectory_heatmap(expr_sel.dellOrso.m, 
                        time = reverse_trajectory(traj.dellOrso)$time,
                        progression_group = group_name.dellOrso.m,
                        modules = modules.dellOrso.m,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
```


**Save as pdf**
```{r dellOrso-gimp-modules-metabolic-genes-pdf, eval=FALSE}
# Gimp
pdf("./../saved/gimp_heatmap_dellOrso_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.dellOrso.m, 
                        time = traj.dellOrso$time,
                        progression_group = group_name.dellOrso.m,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Modules
pdf("./../saved/modules_heatmap_dellOrso_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.dellOrso.m, 
                        time = reverse_trajectory(traj.dellOrso)$time,
                        progression_group = group_name.dellOrso.m,
                        modules = modules.dellOrso.m,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()
```


#### 3.3.2 Gimp: Without primary myoblasts
##### 3.3.2.1 Compute gene importances
#### 5.2.2 Compute gene importances (optional)
```{r gimp-modules-dellOrso-metabolic-genes-without-pm, eval=FALSE}

#TODO (copy from Puhti script)

```

##### 3.3.2.2 Visualize gene importances
**Make a trajectory heatmap**
```{r dellOrso-gimp-modules-metabolic-genes-no-pm}
draw_trajectory_heatmap(expr_sel.dellOrso.m.f, 
                        time = traj.dellOrso.f.time,
                        progression_group = group_name.dellOrso.m.f,
                        modules = modules.dellOrso.m.f,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
```

**save as pdf**
```{r dellOrso-gimp-modules-metabolic-genes-no-pm-pdf, eval=FALSE}
# Gimp
pdf("./../saved/gimp_heatmap_dellOrso_metabolic_genes_no_PM.pdf")
draw_trajectory_heatmap(expr_sel.dellOrso.m.f, 
                        time = traj.dellOrso.f.time,
                        progression_group = group_name.dellOrso.m.f,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Modules
pdf("./../saved/modules_heatmap_dellOrso_metabolic_genes_no_PM.pdf")
draw_trajectory_heatmap(expr_sel.dellOrso.m.f, 
                        time = traj.dellOrso.f.time,
                        progression_group = group_name.dellOrso.m.f,
                        modules = modules.dellOrso.m.f,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()
```

### 3.4 Comparison of metabolic genes between the datasets
```{r comparison-of-metabolic-genes-between-datasets}
# Define colour palette for venn diagram
venn.colors <- brewer.pal(3, "Pastel2")

# Load metabolic genes if not defined
if(!exists("modules.primary.m")) {modules.primary.m <- readRDS("./../saved/modules_primary_m")}
if(!exists("modules.deMicheli.m")) {modules.deMicheli.m <- readRDS("./../saved/modules_deMicheli_m_f")}
if(!exists("modules.dellOrso.m")) {modules.dellOrso.m <- readRDS("./../saved/modules_dellOrso_m_f")}

m.genes.primary <- modules.primary.m$feature
m.genes.deMicheli.f <- modules.deMicheli.m$feature
m.genes.dellOrso.f <- modules.dellOrso.m$feature

# Creata a Venn diagram from metabolic genes
venn <- venn.diagram(
  x = list(m.genes.primary, m.genes.deMicheli.f, m.genes.dellOrso.f),
  category.names = c("Primary data" , "de Micheli " , "Dell'Orso"),
  filename = NULL,
  fill = venn.colors,
  fontface = "bold",
  lwd = 2,
  lty = "blank",
  cat.fontface = "bold",
  fontfamily = "sans",
  cat.fontfamily = "sans"
)

p.gg <- as_ggplot(venn)
p.gg

# Inspect the genes
## Intersection (all)
intersect(m.genes.primary, m.genes.deMicheli.f) %>%
  intersect(m.genes.dellOrso.f)

##Intersection: primary - de Micheli
intersect(m.genes.primary, m.genes.deMicheli.f)

## Intersection: primary - dell'Orso
intersect(m.genes.primary, m.genes.dellOrso.f)

## Intersection: de Micheli - Dell'Orso
intersect(m.genes.deMicheli.f, m.genes.dellOrso.f)

## Union (all)
union(m.genes.primary, m.genes.deMicheli.f) %>% 
  union(m.genes.dellOrso.f)

```


## 4. SingleR cell annotations using only metabolic genes
### 4.1 Cell annotation: all cell types
#### 4.1.1 Load pre-computed prediction objects
```{r load-singleR-predictions-metabolic-genes}
# Load prediction objects
pred.deMicheli.m <- readRDS("./../saved/pred_deMicheli_m")
pred.dellOrso.m <- readRDS("./../saved/pred_dellOrso_m")
```

#### 4.1.2 Predict cell types (optional)
```{r predict-cell-types-using-mteabolic-genes, eval=FALSE}

# TODO (copy from Puhti script)

```

#### 4.1.3 Inspect predictions
```{r inspect-singleR-predictions-metabolic-genes}
# De Micheli
tab.deMicheli.m <- t(table(pred.deMicheli.m$pruned.labels, denoised.sce.m$sample)[, c("control", "cap50", "cap50_r4h","cap50_r8h","cap50_r16h")])
tab.deMicheli.m

# Dell'Orso
tab.dellOrso.m <- t(table(pred.dellOrso.m$pruned.labels, denoised.sce.m$sample)[, c("control", "cap50", "cap50_r4h","cap50_r8h","cap50_r16h")])
tab.dellOrso.m
```

#### 4.1.4 Diagnostic plots
**De Micheli**
```{r deMicheli-annotation-diagnostic-plots-metabolic-genes}
# Score heatmap
plotScoreHeatmap(pred.deMicheli.m, annotation_col = as.data.frame(colData(denoised.sce.m)[, "sample", drop = FALSE]), width = 10)

# Remove low-quality cells
to.remove <- pruneScores(pred.deMicheli.m, min.diff.med = 0.2)
table(Label=pred.deMicheli.m$labels, Removed=to.remove)

# Delta distribution
plotScoreDistribution(results = pred.deMicheli.m, show = "delta.med")
```

**Dell'Orso**
```{r dellOrso-annotation-diagnostic-plots-metabolic-genes}
# Score heatmap
plotScoreHeatmap(pred.dellOrso.m, annotation_col = as.data.frame(colData(denoised.sce.m)[, "sample", drop = FALSE]), width = 10)

# Remove low-quality cells
to.remove <- pruneScores(pred.dellOrso.m, min.diff.med = 0.2)
table(Label=pred.dellOrso.m$labels, Removed=to.remove)

# Delta distribution
plotScoreDistribution(results = pred.dellOrso.m, show = "delta.med")
```


### 4.2 Cell annotation: Without mature skeletal muscle and primary myoblasts
#### 4.2.1 Load pre-computed prediction objects
```{r}
# Load prediction objects
pred.deMicheli.m.no.muscle <- readRDS("./../saved/pred_deMicheli_m_no_muscle")
pred.dellOrso.m.no.pm <- readRDS("./../saved/pred_dellOrso_m_no_pm")
```

#### 4.2.2 Predict cell types (optional)
```{r predict-cell-types-using-metabolic-genes-filtered-cell-types, eval=FALSE}

# TODO (copy from Puhti script)

```

#### 4.2.3 Inspect predictions
```{r inspect-singleR-predictions-metabolic-genes-filtered-cell-types}
# De Micheli (no mature skeletal muscle)
tab.deMicheli.m <- t(table(pred.deMicheli.m.no.muscle$pruned.labels, denoised.sce.m$sample)[, c("control", "cap50", "cap50_r4h","cap50_r8h","cap50_r16h")])
tab.deMicheli.m

# Dell'Orso (no primary myoblasts)
tab.dellOrso.m <- t(table(pred.dellOrso.m.no.pm$pruned.labels, denoised.sce.m$sample)[, c("control", "cap50", "cap50_r4h","cap50_r8h","cap50_r16h")])
tab.dellOrso.m
```

#### 4.2.4 Diagnostic plots
**De Micheli**
```{r deMicheli-annotation-diagnostic-plots-metabolic-genes-no-muscle, eval=FALSE}
# Score heatmap
plotScoreHeatmap(pred.deMicheli.m.no.muscle, annotation_col = as.data.frame(colData(denoised.sce.m)[, "sample", drop = FALSE]), width = 10)

# Remove low-quality cells
to.remove <- pruneScores(pred.deMicheli.m.no.muscle, min.diff.med = 0.2)
table(Label=pred.deMicheli.m.no.muscle$labels, Removed=to.remove)

# Delta distribution
plotScoreDistribution(results = pred.deMicheli.m.no.muscle, show = "delta.med")
```

**Dell'Orso**
```{r dellOrso-annotation-diagnostic-plots-metabolic-genes-no-pm, eval=FALSE}
# Score heatmap
plotScoreHeatmap(pred.dellOrso.m.no.pm, annotation_col = as.data.frame(colData(denoised.sce.m)[, "sample", drop = FALSE]), width = 10)

# Remove low-quality cells
to.remove <- pruneScores(pred.dellOrso.m.no.pm, min.diff.med = 0.2)
table(Label=pred.dellOrso.m.no.pm$labels, Removed=to.remove)

# Delta distribution
plotScoreDistribution(results = pred.dellOrso.m.no.pm, show = "delta.med")
```


## 5. Differential expression (DE) analysis with metabolic genes
### 5.1 Marker genes: Control vs. Cap samples
```{r marker-genes-metabolic-primary-data}
markers.primary <- findMarkers(denoised.sce.m,
                               groups = denoised.sce.m$sample,
                               lfc = 0.5)

control.markers <- markers.primary[["control"]]
control.markers.interesting <- control.markers[control.markers$Top <= 10, ]
#control.markers.interesting <- control.markers[control.markers$FDR <= 0.05, ]
logFCs <- getMarkerEffects(control.markers.interesting)

pheatmap(logFCs, breaks=seq(-2.5, 2.5, length.out=101), main = "Marker genes: Control sample vs. Cap samples (metabolic genes)", fontsize_row = 12)
```

### 5.2 Marker genes shown on reference trajectories
#### 5.2.1 De Micheli trajectory
```{r control-marker-genes-on-deMicheli-trajectory, eval=FALSE}
# Control marker genes
marker.genes.control <- rownames(control.markers.interesting)

marker.gene.plots <- list()

# Define de Micheli trajectory space
expression.deMicheli <- as.matrix(t(logcounts(denoised.sce.deMicheli)))
set.seed(123)
space.deMicheli <- reduce_dimensionality(expression.deMicheli, dist = "spearman", ndim = 2)

# Plot trajectory for each marker
for(gene in marker.genes.control) {
  p <- draw_trajectory_plot(space = space.deMicheli, 
                            progression_group = expression.deMicheli[, gene],
                            point_size = 1,
                            point_alpha = 0.5,
                            path = traj.deMicheli$path,
                            contour = FALSE) + 
  ggtitle(paste(gene, "gene")) +
  theme(plot.title = element_text(size = 14, face = "bold"), axis.title = element_text(size = 10))
  marker.gene.plots[[gene]] <- p
}

gridExtra::grid.arrange(marker.gene.plots$Cth, marker.gene.plots$Taldo1, marker.gene.plots$Gclm, marker.gene.plots$Psph, marker.gene.plots$Gal, marker.gene.plots$Pgd,
                        marker.gene.plots$Cox6a2, marker.gene.plots$Idi1, marker.gene.plots$Sat1, marker.gene.plots$Pck2, marker.gene.plots$Maoa, marker.gene.plots$Gbe1,
                        marker.gene.plots$Eno3, marker.gene.plots$Atp6v1g1, marker.gene.plots$Hgsnat, marker.gene.plots$Atp6v0b, marker.gene.plots$Hexa, marker.gene.plots$Uap1l1,
                        marker.gene.plots$Aldh2, ncol = 4)
```

#### 5.2.2 Dell'Orso trajectory
```{r control-marker-genes-on-dellOrso-trajectory, eval=FALSE}
# Control marker genes
marker.genes.control <- rownames(control.markers.interesting)

marker.gene.plots <- list()

# Define de Micheli trajectory space
expression.dellOrso <- as.matrix(t(logcounts(denoised.sce.dellOrso)))
set.seed(123)
space.dellOrso <- reduce_dimensionality(expression.dellOrso, dist = "spearman", ndim = 2)

# Plot trajectory for each marker
for(gene in marker.genes.control) {
  p <- draw_trajectory_plot(space = space.dellOrso, 
                            progression_group = expression.dellOrso[, gene],
                            point_size = 1,
                            point_alpha = 0.5,
                            path = traj.dellOrso$path,
                            contour = FALSE) + 
  ggtitle(paste(gene, "gene")) +
  theme(plot.title = element_text(size = 14, face = "bold"), axis.title = element_text(size = 10))
  marker.gene.plots[[gene]] <- p
}

gridExtra::grid.arrange(marker.gene.plots$Cth, marker.gene.plots$Taldo1, marker.gene.plots$Gclm, marker.gene.plots$Psph, marker.gene.plots$Gal, marker.gene.plots$Pgd,
                        marker.gene.plots$Cox6a2, marker.gene.plots$Idi1, marker.gene.plots$Sat1, marker.gene.plots$Pck2, marker.gene.plots$Maoa, marker.gene.plots$Gbe1,
                        marker.gene.plots$Eno3, marker.gene.plots$Atp6v1g1, marker.gene.plots$Hgsnat, marker.gene.plots$Atp6v0b, marker.gene.plots$Hexa, marker.gene.plots$Uap1l1,
                        marker.gene.plots$Aldh2, ncol = 4)
```
