---
title: "Metabolic_pathways"
author: "Ville Lehtonen"
date: "11/3/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Set working dir to the source file location
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
```

# Metabolic pathways
Here, we inspect the different metabolic modules / pathways that cells adopt as they progress through the differentiation trajectory

## 1. Load packages and data
### 1.1 Load packages
```{r load-packages, message=FALSE}
library(SCORPIUS)
library(scater)
library(scran)
library(SingleCellExperiment)
library(dplyr)
library(SingleR)
library(igraph)
library(stats)
library(cluster)
library(pheatmap)
library(fossil)
library(Seurat)
library(cerebroApp)
library(reshape)
library(data.table)
library(scRNAseq)
library(org.Mm.eg.db)
library(tidyr)
library(BBmisc)
library(KEGGREST)
library(RColorBrewer)
library(ggpubr)
library(grid)
library(gridExtra)
library(plyr)
library(stringr)
library(ggrepel)
library(VennDiagram)
```

### 1.2 Load pre-calculated data
**Single Cell Experiments**
```{r load-objects}
# Load pre-processed sce objects
denoised.sce <- readRDS("./../saved/denoised_sce")
denoised.sce.deMicheli <- readRDS("./../saved/denoised_sce_deMicheli")
denoised.sce.dellOrso <- readRDS("./../saved/denoised_sce_dellOrso")

# Load datasets with metabolic genes only
denoised.sce.m <- readRDS("./../saved/denoised_sce_m")
denoised.sce.deMicheli.m <- readRDS("./../saved/denoised_sce_deMicheli_m")
denoised.sce.dellOrso.m <- readRDS("./../saved/denoised_sce_dellOrso_m")

# Define group names for trajectories
group_name.primary.m <- as.factor(denoised.sce.m$sample)
group_name.deMicheli.m <- as.factor(denoised.sce.deMicheli.m$cell_type)
group_name.dellOrso.m <- as.factor(denoised.sce.dellOrso.m$cell_type)
```

**Trajectories**
```{r load-pre-calculated-trajectories}
# Load pre-calculated trajectory objects
traj.primary <- readRDS("./../saved/traj_primary")
traj.deMicheli <- readRDS("./../saved/traj_deMicheli")
traj.dellOrso <- readRDS("./../saved/traj_dellOrso")
```


### 1.2 Build trajectories for datasets (Optional)
#### 1.2.1 Primary data
```{r trajectory-primary-data, eval=FALSE}
# Create expression and sample objects
expression.primary <- as.matrix(t(logcounts(denoised.sce)))
group_name.primary <- as.factor(denoised.sce$sample)

# Trajectory inference
set.seed(123)
space.primary <- reduce_dimensionality(expression.primary, dist = "spearman", ndim = 2)
set.seed(123)
traj.primary <- infer_trajectory(space.primary)
traj.primary <- reverse_trajectory(traj.primary)
```

#### 1.2.2 De Micheli data
```{r trajectory-deMicheli, eval=FALSE}
# Create expression and sample objects
expression.deMicheli <- as.matrix(t(logcounts(denoised.sce.deMicheli)))
group_name.deMicheli <- as.factor(denoised.sce.deMicheli$cell_type)

# Trajectory building
set.seed(123)
space.deMicheli <- reduce_dimensionality(expression.deMicheli, dist = "spearman", ndim = 2)
set.seed(123)
traj.deMicheli <- infer_trajectory(space.deMicheli)
traj.deMicheli <- reverse_trajectory(traj.deMicheli)
```

#### 1.2.3 Dell'Orso data
```{r trajectory-dellOrso, eval=FALSE}
# Create expression and sample objects
expression.dellOrso <- as.matrix(t(logcounts(denoised.sce.dellOrso)))
group_name.dellOrso <- as.factor(denoised.sce.dellOrso$cell_type)

# Trajectory building
set.seed(123)
space.dellOrso <- reduce_dimensionality(expression.dellOrso, dist = "spearman", ndim = 2)
set.seed(123)
traj.dellOrso <- infer_trajectory(space.dellOrso)
traj.dellOrso <- reverse_trajectory(traj.dellOrso)
```


## 2. Mapping of KEGG modules & pathways to the datasets

### 2.1 Fetch KEGG modules & pathway information
#### 2.1.2 Fetch KEGG information from pre-computed files
```{r load-fetched-KEGG-information}
# Fetch KEGG entries information
df.kegg.entries <- readRDS("./../saved/df_kegg_entries")
```

#### 2.1.2 Fetch information from KEGG (optional)
```{r fetch-kegg-information, eval=FALSE, message=FALSE}
# Fetch all the modules (and KEGG entries) for mouse
mouse.org.id <- "T01002"
mouse.modules <- keggLink(target = "module", source = mouse.org.id)

# Build a dataframe to store module, gene, and pathway information for each KEGG entry
df.kegg.entries <- data.frame(entry = names(mouse.modules),
                              module = NA,
                              genes = NA,
                              pathway = NA,
                              module_info = NA,
                              pathway_info = NA)

# remove duplicate rows from data
df.kegg.entries <- distinct(df.kegg.entries)

# For each KEGG entry, fetch the information with keggGet()
for (i in 1:nrow(df.kegg.entries)) {
  print(i)
  entry <- df.kegg.entries[i,]$entry
  query <- keggGet(entry)
  
  # Check if entry is associated with metabolic pathways (mmu01100)
  if("mmu01100" %in% names(query[[1]]$PATHWAY)) {
    genes <- query[[1]]$NAME
    module <- paste(names(query[[1]]$MODULE), collapse = ", ")
    pathway <- paste(names(query[[1]]$PATHWAY), collapse = ", ")
    module_info <- paste(unlist(query[[1]]$MODULE), collapse = ", ")
    pathway_info <- paste(unlist(query[[1]]$PATHWAY), collapse = ", ")
    df.kegg.entries[i,]$genes <- genes
    df.kegg.entries[i,]$module <- module
    df.kegg.entries[i,]$pathway <- pathway
    df.kegg.entries[i,]$module_info <- module_info
    df.kegg.entries[i,]$pathway_info <- pathway_info
  }
}

# Filter out entries with no association to metabolic pathways
non.metabolic <- is.na(df.kegg.entries$module)
df.kegg.entries <- df.kegg.entries[!non.metabolic, ]

# Save objects
#saveRDS(df.kegg.entries, file = "./../saved/df_kegg_entries")
```


### 2.2 Add KEGG modules & pathway information to the datasets (Optional)
#### 2.2.1 Primary data
```{r add-KEGG-information-to-primary-data, eval=FALSE, message=FALSE}
# Load mouse modules if it doesn't exist
if(!exists("df.kegg.entries")) {
  df.kegg.entries <- readRDS("./../saved/df_kegg_entries")
}

denoised.sce.m <- denoised.sce
rowData(denoised.sce.m)$module <- ""
rowData(denoised.sce.m)$pathway <- ""
rowData(denoised.sce.m)$module_info <- ""
rowData(denoised.sce.m)$pathway_info <- ""

for(i in 1:nrow(df.kegg.entries)) {
  print(i)
  # Split the gene names into a list
  genes <- unlist(lapply(strsplit(df.kegg.entries[i,]$genes, split = ",")[[1]], trimws))
  useful_genes <- genes[genes %in% rownames(denoised.sce.m)]
  for(gene in useful_genes) {
    rowData(denoised.sce.m)[gene, ]$module <- df.kegg.entries[i,]$module
    rowData(denoised.sce.m)[gene, ]$pathway <- df.kegg.entries[i,]$pathway
    rowData(denoised.sce.m)[gene, ]$module_info <- df.kegg.entries[i,]$module_info
    rowData(denoised.sce.m)[gene, ]$pathway_info <- df.kegg.entries[i,]$pathway_info
  }
}

genes.to.use <- rownames(denoised.sce.m)[rowData(denoised.sce.m)$module != ""]
denoised.sce.m <- denoised.sce.m[rownames(denoised.sce.m) %in% genes.to.use, ]

# Save sce object containing only metabolic genes and module/pathway information stored on rowData
#saveRDS(denoised.sce.m, "./../saved/denoised_sce_m")
```

#### 2.2.2 De Micheli et al. (2020) reference data
```{r add-KEGG-information-to-deMicheli-data, eval=FALSE, message=FALSE}
# Load mouse modules if it doesn't exist
if(!exists("df.kegg.entries")) {
  df.kegg.entries <- readRDS("./../saved/df_kegg_entries")
}

denoised.sce.deMicheli.m <- denoised.sce.deMicheli
rowData(denoised.sce.deMicheli.m)$module <- ""
rowData(denoised.sce.deMicheli.m)$pathway <- ""
rowData(denoised.sce.deMicheli.m)$module_info <- ""
rowData(denoised.sce.deMicheli.m)$pathway_info <- ""

for(i in 1:nrow(df.kegg.entries)) {
  print(i)
  # Split the gene names into a list
  genes <- unlist(lapply(strsplit(df.kegg.entries[i,]$genes, split = ",")[[1]], trimws))
  useful_genes <- genes[genes %in% rownames(denoised.sce.deMicheli.m)]
  for(gene in useful_genes) {
    rowData(denoised.sce.deMicheli.m)[gene, ]$module <- df.kegg.entries[i,]$module
    rowData(denoised.sce.deMicheli.m)[gene, ]$pathway <- df.kegg.entries[i,]$pathway
    rowData(denoised.sce.deMicheli.m)[gene, ]$module_info <- df.kegg.entries[i,]$module_info
    rowData(denoised.sce.deMicheli.m)[gene, ]$pathway_info <- df.kegg.entries[i,]$pathway_info
  }
}

genes.to.use <- rownames(denoised.sce.deMicheli.m)[rowData(denoised.sce.deMicheli.m)$module != ""]
denoised.sce.deMicheli.m <- denoised.sce.deMicheli.m[rownames(denoised.sce.deMicheli.m) %in% genes.to.use, ]

# Save sce object containing only metabolic genes and module/pathway information stored on rowData
#saveRDS(denoised.sce.deMicheli.m, "./../saved/denoised_sce_deMicheli_m")
```


#### 2.2.3 Dell'Orso et al. (2019) reference data
```{r add-KEGG-information-to-dellOrso-data, eval=FALSE, message=FALSE}
# Load mouse modules if it doesn't exist
if(!exists("df.kegg.entries")) {
  df.kegg.entries <- readRDS("./../saved/df_kegg_entries")
}

denoised.sce.dellOrso.m <- denoised.sce.dellOrso
rowData(denoised.sce.dellOrso.m)$module <- ""
rowData(denoised.sce.dellOrso.m)$pathway <- ""
rowData(denoised.sce.dellOrso.m)$module_info <- ""
rowData(denoised.sce.dellOrso.m)$pathway_info <- ""

for(i in 1:nrow(df.kegg.entries)) {
  print(i)
  # Split the gene names into a list
  genes <- unlist(lapply(strsplit(df.kegg.entries[i,]$genes, split = ",")[[1]], trimws))
  useful_genes <- genes[genes %in% rownames(denoised.sce.dellOrso.m)]
  for(gene in useful_genes) {
    rowData(denoised.sce.dellOrso.m)[gene, ]$module <- df.kegg.entries[i,]$module
    rowData(denoised.sce.dellOrso.m)[gene, ]$pathway <- df.kegg.entries[i,]$pathway
    rowData(denoised.sce.dellOrso.m)[gene, ]$module_info <- df.kegg.entries[i,]$module_info
    rowData(denoised.sce.dellOrso.m)[gene, ]$pathway_info <- df.kegg.entries[i,]$pathway_info
  }
}

genes.to.use <- rownames(denoised.sce.dellOrso.m)[rowData(denoised.sce.dellOrso.m)$module != ""]
denoised.sce.dellOrso.m <- denoised.sce.dellOrso.m[rownames(denoised.sce.dellOrso.m) %in% genes.to.use, ]

# Save sce object containing only metabolic genes and module/pathway information stored on rowData
#saveRDS(denoised.sce.dellOrso.m, "./../saved/denoised_sce_dellOrso_m")
```


## 3. Gene importances using only metabolic genes
#### 3.1 Primary data
##### 3.1.1 Load gene importances
```{r load-gimp-primary-metabolic-genes}
# Load gimp & modules
gimp.primary.m <- readRDS("./../saved/gimp_primary_m")
modules.primary.m <- readRDS("./../saved/modules_primary_m")

gimp.primary.m$qvalue <- p.adjust(p = gimp.primary.m$pvalue, method = "BH")
gene_sel.primary.m <- gimp.primary.m$gene[gimp.primary.m$qvalue < .05]

if(!exists("expression.primary.m")) {expression.primary.m <- as.matrix(t(logcounts(denoised.sce.m)))}
expr_sel.primary.m <- scale_quantile(expression.primary.m[, gene_sel.primary.m])
```


#### 3.1.2 Compute gene importances (optional)
```{r gimp-modules-primary-metabolic-genes, eval=FALSE}
# Read trajectory from a file (if not defined)
if(!exists("traj.primary")) {
  traj.primary <- readRDS("./../saved/traj_primary")
}

# Create expression and sample objects
expression.primary.m <- as.matrix(t(logcounts(denoised.sce.m)))
group_name.primary.m <- as.factor(denoised.sce.m$sample)

# Gene importances
gimp.primary.m <- gene_importances(x = expression.primary.m,
                                           time = traj.primary$time,
                                           num_permutations = 10,
                                           num_threads = 8)

gimp.primary.m$qvalue <- p.adjust(p = gimp.primary.m$pvalue, method = "BH")
gene_sel.primary.m <- gimp.primary.m$gene[gimp.primary.m$qvalue < .05]
expr_sel.primary.m <- scale_quantile(expression.primary.m[, gene_sel.primary.m])

#saveRDS(gimp.primary.m, file = "./../saved/gimp_primary_m")

# Extract modules
modules.primary.m <- extract_modules(expr_sel.primary.m,
                                     time = traj.primary$time,
                                     verbose = FALSE)
#saveRDS(modules.primary.m, "./../saved/modules_primary_m")
```

#### 3.1.3 Visualization of gene importances
```{r primary-data-gimp-modules-metabolic-genes, eval=FALSE}
# Get rowdata (incl. KEGG modules for each gene) for the metabolic genes
rowdata.primary <- rowData(denoised.sce.m)
genes.f <- rownames(rowdata.primary) %in% colnames(expr_sel.primary.m)
rowdata.primary.f <- rowdata.primary[genes.f, ]

# Create an annotation row df
ann_row <- data.frame(metabolic_module = factor(rowdata.primary.f$module))
rownames(ann_row) <- rownames(rowdata.primary.f)

# Order genes to same order as in modules object
ann_row <- ann_row[modules.primary.m$feature, ,drop=FALSE]

# Add metabolic module information to the modules object
modules.primary.m$metabolic_module <- ann_row$metabolic_module

# Arrange modules object by module and by metabolic_module
modules.primary.m.kegg <- arrange(modules.primary.m,
                                  as.character(modules.primary.m$module),
                                  as.character(modules.primary.m$metabolic_module))

# save modules object
#saveRDS(modules.primary.m.kegg, file = "./../saved/modules_primary_m_kegg")

# Create annotation row color palette (43 colors)
ann_colors <- colorRampPalette(grDevices::rainbow(length(unique(ann_row$metabolic_module))))
ann_colors <- ann_colors(length(unique(ann_row$metabolic_module)))
names(ann_colors) <- unique(ann_row$metabolic_module)
ann_colors <- list(metabolic_module = ann_colors)

# Gimp
pdf("./../saved/gimp_heatmap_primary_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.primary.m, 
                        time = traj.primary$time,
                        progression_group = group_name.primary.m,
                        show_labels_row = TRUE,
                        fontsize_row = 5)
dev.off()


# Modules
pdf("./../saved/modules_heatmap_primary_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.primary.m, 
                        time = traj.primary$time,
                        progression_group = group_name.primary.m,
                        modules = modules.primary.m.kegg,
                        show_labels_row = TRUE,
                        annotation_row = ann_row,
                        #annotation_colors = ann_colors,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Save modules object as csv
write.csv(as.data.frame(modules.primary.m.kegg), file = "./../saved/modules_primary_m_kegg.csv")
```

### 3.2 De Micheli et al. (2020) reference data
#### 3.2.1 Load gene importances
```{r load-gimp-deMicheli-metabolic-genes}
# Load gimp & modules
gimp.deMicheli.m <- readRDS("./../saved/gimp_deMicheli_m")
modules.deMicheli.m <- readRDS("./../saved/modules_deMicheli_m")

gimp.deMicheli.m$qvalue <- p.adjust(p = gimp.deMicheli.m$pvalue, method = "BH")
gene_sel.deMicheli.m <- gimp.deMicheli.m$gene[gimp.deMicheli.m$qvalue < .05]

if(!exists("expression.deMicheli.m")) {expression.deMicheli.m <- as.matrix(t(logcounts(denoised.sce.deMicheli.m)))}
expr_sel.deMicheli.m <- scale_quantile(expression.deMicheli.m[, gene_sel.deMicheli.m])
```

#### 3.2.2 Compute gene importances (optional)
```{r gimp-modules-deMicheli-metabolic-genes, eval=FALSE}
# Read trajectory from a file (if not defined)
if(!exists("traj.deMicheli")) {
  traj.deMicheli <- readRDS("./../saved/traj_deMicheli")
}

# Create expression and sample objects
expression.deMicheli.m <- as.matrix(t(logcounts(denoised.sce.deMicheli.m)))
group_name.deMicheli.m <- as.factor(denoised.sce.deMicheli.m$cell_type)

# Gene importances
gimp.deMicheli.m <- gene_importances(x = expression.deMicheli.m,
                                     time = traj.deMicheli$time,
                                     num_permutations = 10,
                                     num_threads = 8)

gimp.deMicheli.m$qvalue <- p.adjust(p = gimp.deMicheli.m$pvalue, method = "BH")
gene_sel.deMicheli.m <- gimp.deMicheli.m$gene[gimp.deMicheli.m$qvalue < .05]
expr_sel.deMicheli.m <- scale_quantile(expression.deMicheli.m[, gene_sel.deMicheli.m])

#saveRDS(gimp.deMicheli.m, file = "./../saved/gimp_deMicheli_m")

# Extract modules
modules.deMicheli.m <- extract_modules(expr_sel.deMicheli.m,
                                       time = traj.deMicheli$time,
                                       verbose = FALSE)

#saveRDS(modules.deMicheli.m, "./../saved/modules_deMicheli_m")
```


##### 3.2.3 Visualization of gene importances (pdf format)
```{r deMicheli-gimp-modules-metabolic-genes, eval=FALSE}
# Get rowdata (incl. KEGG modules for each gene) for the metabolic genes
rowdata.deMicheli <- rowData(denoised.sce.deMicheli.m)
genes.deMicheli.f <- rownames(rowdata.deMicheli) %in% colnames(expr_sel.deMicheli.m)
rowdata.deMicheli.f <- rowdata.deMicheli[genes.deMicheli.f, ]

# Create an annotation row df
ann_row_deMicheli <- data.frame(metabolic_module = factor(rowdata.deMicheli.f$module))
rownames(ann_row_deMicheli) <- rownames(rowdata.deMicheli.f)

# Order genes to same order as in modules object
ann_row_deMicheli <- ann_row_deMicheli[modules.deMicheli.m$feature, ,drop=FALSE]

# Add metabolic module information to the modules object
modules.deMicheli.m$metabolic_module <- ann_row_deMicheli$metabolic_module

# Arrange modules object by module and by metabolic_module
modules.deMicheli.m.kegg <- arrange(modules.deMicheli.m,
                                    as.character(modules.deMicheli.m$module),
                                    as.character(modules.deMicheli.m$metabolic_module))

# save modules object
#saveRDS(modules.deMicheli.m.kegg, file = "./../saved/modules_deMicheli_m_kegg")

# Gimp
pdf("./../saved/gimp_heatmap_deMicheli_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli.m, 
                        time = traj.deMicheli$time,
                        progression_group = group_name.deMicheli.m,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Modules
pdf("./../saved/modules_heatmap_deMicheli_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli.m, 
                        time = traj.deMicheli$time,
                        progression_group = group_name.deMicheli.m,
                        modules = modules.deMicheli.m.kegg,
                        annotation_row = ann_row_deMicheli,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Save modules object as csv
write.csv(as.data.frame(modules.deMicheli.m.kegg), file = "./../saved/modules_deMicheli_m_kegg.csv")
```


### 3.3 Dell'Orso et al. (2019) reference data
#### 3.3.1 Load gene importances
```{r load-gimp-dellOrso-metabolic-genes}
# Read gimp & modules from a file
gimp.dellOrso.m <- readRDS("./../saved/gimp_dellOrso_m")
modules.dellOrso.m <- readRDS("./../saved/modules_dellOrso_m")

# Create expression and sample objects
gimp.dellOrso.m$qvalue <- p.adjust(p = gimp.dellOrso.m$pvalue, method = "BH")
gene_sel.dellOrso.m <- gimp.dellOrso.m$gene[gimp.dellOrso.m$qvalue < .05]

if(!exists("expression.dellOrso.m")) {expression.dellOrso.m <- as.matrix(t(logcounts(denoised.sce.dellOrso.m)))}
expr_sel.dellOrso.m <- scale_quantile(expression.dellOrso.m[, gene_sel.dellOrso.m])
```


#### 3.3.2 Compute gene importances (optional)
```{r gimp-modules-dellOrso-metabolic-genes, eval=FALSE}
# Read trajectory from a file (if not defined)
if(!exists("traj.dellOrso")) {
  traj.dellOrso <- readRDS("./../saved/traj_dellOrso")
}

# Create expression and sample objects
expression.dellOrso.m <- as.matrix(t(logcounts(denoised.sce.dellOrso.m)))
group_name.dellOrso.m <- as.factor(denoised.sce.dellOrso.m$cell_type)

# Gene importances
gimp.dellOrso.m <- gene_importances(x = expression.dellOrso.m,
                                    time = traj.dellOrso$time,
                                    num_permutations = 10,
                                    num_threads = 8)

gimp.dellOrso.m$qvalue <- p.adjust(p = gimp.dellOrso.m$pvalue, method = "BH")
gene_sel.dellOrso.m <- gimp.dellOrso.m$gene[gimp.dellOrso.m$qvalue < .05]
expr_sel.dellOrso.m <- scale_quantile(expression.dellOrso.m[, gene_sel.dellOrso.m])

#saveRDS(gimp.dellOrso.m, file = "./../saved/gimp_dellOrso_m")

# Extract modules
modules.dellOrso.m <- extract_modules(expr_sel.dellOrso.m,
                                      time = traj.dellOrso$time,
                                      verbose = FALSE)

#saveRDS(modules.dellOrso.m, "./../saved/modules_dellOrso_m")
```


#### 3.3.3 Visualization of gene importances
```{r dellOrso-gimp-modules-metabolic-genes, eval=FALSE}
# Get rowdata (incl. KEGG modules for each gene) for the metabolic genes
rowdata.dellOrso <- rowData(denoised.sce.dellOrso.m)
genes.dellOrso.f <- rownames(rowdata.dellOrso) %in% colnames(expr_sel.dellOrso.m)
rowdata.dellOrso.f <- rowdata.dellOrso[genes.dellOrso.f, ]

# Create an annotation row df
ann_row_dellOrso <- data.frame(metabolic_module = factor(rowdata.dellOrso.f$module))
rownames(ann_row_dellOrso) <- rownames(rowdata.dellOrso.f)

# Order genes to same order as in modules object
ann_row_dellOrso <- ann_row_dellOrso[modules.dellOrso.m$feature, ,drop=FALSE]

# Add metabolic module information to the modules object
modules.dellOrso.m$metabolic_module <- ann_row_dellOrso$metabolic_module

# Arrange modules object by module and by metabolic_module
modules.dellOrso.m.kegg <- arrange(modules.dellOrso.m,
                                   as.character(modules.dellOrso.m$module),
                                   as.character(modules.dellOrso.m$metabolic_module))

# save modules object
#saveRDS(modules.dellOrso.m.kegg, file = "./../saved/modules_dellOrso_m_kegg")

# Gimp
pdf("./../saved/gimp_heatmap_dellOrso_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.dellOrso.m, 
                        time = traj.dellOrso$time,
                        progression_group = group_name.dellOrso.m,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Modules
pdf("./../saved/modules_heatmap_dellOrso_metabolic_genes.pdf")
draw_trajectory_heatmap(expr_sel.dellOrso.m, 
                        time = reverse_trajectory(traj.dellOrso)$time,
                        progression_group = group_name.dellOrso.m,
                        modules = modules.dellOrso.m,
                        annotation_row = ann_row_dellOrso,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Save modules object as csv
write.csv(as.data.frame(modules.dellOrso.m.kegg), file = "./../saved/modules_dellOrso_m_kegg.csv")
```

## 4. Visualization (barplots) of metabolic modules & pathways vs. Scorpius modules
### 4.1 Modules
#### 4.1.1 Primary data
```{r primary-data-kegg-module-visualization, message=FALSE, warning=FALSE}
# Load kegg modules object (if not defined)
if(!exists("modules.primary.m.kegg")){modules.primary.m.kegg <- readRDS("./../saved/modules_primary_m_kegg")}

# Create contingency table
tab.modules.primary <- table(modules.primary.m.kegg$metabolic_module, modules.primary.m.kegg$module)

# Melt table
df.modules.primary <- melt(tab.modules.primary)
colnames(df.modules.primary) <- c("KEGG_module", "Scorpius_module", "frequency")
df.modules.primary$KEGG_module <- unfactor(df.modules.primary$KEGG_module)

# Handle rows where there are multiple KEGG modules mapping to one scorpius module
df.modules.primary.separated <- distinct(as.data.frame(separate_rows(df.modules.primary, KEGG_module, sep = ", ")))

# Plot metabolic modules vs. Scorpius modules
p.primary <- ggplot(df.modules.primary.separated, aes(x = as.factor(Scorpius_module), y = frequency)) +
  geom_bar(aes(fill = KEGG_module), stat = "identity", position = position_dodge(1)) +
  ggtitle("Frequency of KEGG modules in different trajectory modules",
          subtitle = "Primary dataset") +
  theme(legend.position="right") +
  geom_label_repel(aes(label=ifelse(frequency>=3, KEGG_module, "")), position = position_dodge(width=0.9), vjust=-0.5)

p.primary

```

#### 4.1.2 De Micheli et al. (2020) reference data
```{r deMicheli-data-kegg-module-visualization, message=FALSE, warning=FALSE}
# Load kegg modules object (if not defined)
if(!exists("modules.deMicheli.m.kegg")){modules.deMicheli.m.kegg <- readRDS("./../saved/modules_deMicheli_m_kegg")}

# Create contingency table
tab.modules.deMicheli <- table(modules.deMicheli.m.kegg$metabolic_module, modules.deMicheli.m.kegg$module)

# Melt table
df.modules.deMicheli <- melt(tab.modules.deMicheli)
colnames(df.modules.deMicheli) <- c("KEGG_module", "Scorpius_module", "frequency")
df.modules.deMicheli$KEGG_module <- unfactor(df.modules.deMicheli$KEGG_module)

# Handle rows where there are multiple KEGG modules mapping to one scorpius module
df.modules.deMicheli.separated <- distinct(as.data.frame(separate_rows(df.modules.deMicheli, KEGG_module, sep = ", ")))

# Plot metabolic modules vs. Scorpius modules
p.deMicheli <- ggplot(df.modules.deMicheli.separated, aes(x = as.factor(Scorpius_module), y = frequency)) +
  geom_bar(aes(fill = KEGG_module), stat = "identity", position = position_dodge(1)) +
  ggtitle("Frequency of KEGG modules in different trajectory modules",
          subtitle = "De Micheli et al. reference dataset") +
  theme(legend.position="right") +
  geom_label_repel(aes(label=ifelse(frequency>=3, KEGG_module, "")), position = position_dodge(width=0.9), vjust=-0.5)

p.deMicheli
```


#### 4.1.3 Dell'Orso et al. (2019) reference data
```{r dellOrso-data-kegg-module-visualization, message=FALSE, warning=FALSE}
# Load kegg modules object (if not defined)
if(!exists("modules.dellOrso.m.kegg")){modules.dellOrso.m.kegg <- readRDS("./../saved/modules_dellOrso_m_kegg")}

# Create contingency table
tab.modules.dellOrso <- table(modules.dellOrso.m.kegg$metabolic_module, modules.dellOrso.m.kegg$module)

# Melt table
df.modules.dellOrso <- melt(tab.modules.dellOrso)
colnames(df.modules.dellOrso) <- c("KEGG_module", "Scorpius_module", "frequency")
df.modules.dellOrso$KEGG_module <- unfactor(df.modules.dellOrso$KEGG_module)

# Handle rows where there are multiple KEGG modules mapping to one scorpius module
df.modules.dellOrso.separated <- distinct(as.data.frame(separate_rows(df.modules.dellOrso, KEGG_module, sep = ", ")))

# Plot metabolic modules vs. Scorpius modules
p.dellOrso <- ggplot(df.modules.dellOrso.separated, aes(x = as.factor(Scorpius_module), y = frequency)) +
  geom_bar(aes(fill = KEGG_module), stat = "identity", position = position_dodge(1)) +
  ggtitle("Frequency of KEGG modules in different trajectory modules",
          subtitle = "Dell'Orso et al. referene dataset") +
  theme(legend.position="right") +
  geom_label_repel(aes(label=ifelse(frequency>=3, KEGG_module, "")), position = position_dodge(width=0.9), vjust=-0.5)

p.dellOrso
```

**Save plots as pdf**
```{r kegg-module-plots-as-pdf, eval=FALSE}
pdf("./../saved/KEGG_modules_in_trajectories", width = 11, height = 5)
p.primary
p.deMicheli
p.dellOrso
dev.off()
```

### 4.2 Pathways
#### 4.2.1 Create mapping between genes,modules, and pathways
```{r mapping-pathways}
# Load scorpius modules objects (if needed)
if(!exists("modules.primary.m.kegg")){modules.primary.m.kegg <- readRDS("./../saved/modules_primary_m_kegg")}
if(!exists("modules.deMicheli.m.kegg")){modules.deMicheli.m.kegg <- readRDS("./../saved/modules_deMicheli_m_kegg")}
if(!exists("modules.dellOrso.m.kegg")){modules.dellOrso.m.kegg <- readRDS("./../saved/modules_dellOrso_m_kegg")}

###############
# primary data
###############
rowData.primary <- rowData(denoised.sce.m)
genes.primary <- rownames(rowData.primary) %in% colnames(expr_sel.primary.m)
rowData.primary.f <- rowData.primary[genes.primary, ]

# Create mapping between genes and pathways
pathway.mapping.primary <- data.frame(pathway = factor(rowData.primary.f$pathway))
rownames(pathway.mapping.primary) <- rownames(rowData.primary.f)

# Order genes
pathway.mapping.primary <- pathway.mapping.primary[modules.primary.m.kegg$feature, , drop = FALSE]

# Add pathway information to modules object
modules.primary.m.kegg$pathway <- pathway.mapping.primary$pathway

# Arrange rows
modules.primary.m.kegg.pathway <- arrange(modules.primary.m.kegg,
                                           as.character(modules.primary.m.kegg$module),
                                           as.character(modules.primary.m.kegg$metabolic_module))

#saveRDS(modules.primary.m.kegg.pathway, file = "./../saved/modules_primary_m_kegg_pathway")

#########################
# De Micheli et al. data
#########################
rowData.deMicheli <- rowData(denoised.sce.deMicheli.m)
genes.deMicheli <- rownames(rowData.deMicheli) %in% colnames(expr_sel.deMicheli.m)
rowData.deMicheli.f <- rowData.deMicheli[genes.deMicheli, ]

# Create mapping between genes and pathways
pathway.mapping.deMicheli <- data.frame(pathway = factor(rowData.deMicheli.f$pathway))
rownames(pathway.mapping.deMicheli) <- rownames(rowData.deMicheli.f)

# Order genes
pathway.mapping.deMicheli <- pathway.mapping.deMicheli[modules.deMicheli.m.kegg$feature, , drop = FALSE]

# Add pathway information to modules object
modules.deMicheli.m.kegg$pathway <- pathway.mapping.deMicheli$pathway

# Arrange rows
modules.deMicheli.m.kegg.pathway <- arrange(modules.deMicheli.m.kegg,
                                            as.character(modules.deMicheli.m.kegg$module),
                                            as.character(modules.deMicheli.m.kegg$metabolic_module))

#saveRDS(modules.deMicheli.m.kegg.pathway, file = "./../saved/modules_deMicheli_m_kegg_pathway")

#########################
# Dell'Orso et al. data
#########################
rowData.dellOrso <- rowData(denoised.sce.dellOrso.m)
genes.dellOrso <- rownames(rowData.dellOrso) %in% colnames(expr_sel.deMicheli.m)
rowData.dellOrso.f <- rowData.dellOrso[genes.dellOrso, ]

# Create mapping between genes and pathways
pathway.mapping.dellOrso <- data.frame(pathway = factor(rowData.dellOrso.f$pathway))
rownames(pathway.mapping.dellOrso) <- rownames(rowData.dellOrso.f)

# Order genes
pathway.mapping.dellOrso <- pathway.mapping.dellOrso[modules.dellOrso.m.kegg$feature, , drop = FALSE]

# Add pathway information to modules object
modules.dellOrso.m.kegg$pathway <- pathway.mapping.dellOrso$pathway

# Arrange rows
modules.dellOrso.m.kegg.pathway <- arrange(modules.dellOrso.m.kegg,
                                           as.character(modules.dellOrso.m.kegg$module),
                                           as.character(modules.dellOrso.m.kegg$metabolic_module))

#saveRDS(modules.dellOrso.m.kegg.pathway, file = "./../saved/modules_dellOrso_m_kegg_pathway")
```


#### 4.2.2 Primary data
```{r primary-data-kegg-pathway-visualization, message=FALSE, warning=FALSE}
# Load kegg modules object (if not defined)
if(!exists("modules.primary.m.kegg.pathway")){modules.primary.m.kegg.pathway <- readRDS("./../saved/modules_primary_m_kegg_pathway")}

# Create contingency table
tab.pathways.primary <- table(modules.primary.m.kegg.pathway$pathway, modules.primary.m.kegg.pathway$module)

# Melt table
df.pathways.primary <- melt(tab.pathways.primary)
colnames(df.pathways.primary) <- c("KEGG_pathway", "Scorpius_module", "frequency")
df.pathways.primary$KEGG_pathway <- unfactor(df.pathways.primary$KEGG_pathway)

# Handle rows where there are multiple KEGG modules mapping to one scorpius module
df.pathways.primary.separated <- distinct(as.data.frame(separate_rows(df.pathways.primary, KEGG_pathway, sep = ", ")))

# Plot metabolic modules vs. Scorpius modules
p.primary <- ggplot(df.pathways.primary.separated, aes(x = as.factor(Scorpius_module), y = frequency)) +
  geom_bar(aes(fill = KEGG_pathway), stat = "identity", position = position_dodge(1)) +
  ggtitle("Frequency of KEGG pathways in different trajectory modules",
          subtitle = "Primary dataset") +
  theme(legend.position="right") +
  geom_label_repel(aes(label=ifelse(frequency>=3, KEGG_pathway, "")), position = position_dodge(width=0.9), vjust=-0.5)

p.primary
```


#### 4.2.3 De Micheli et al. (2020) reference data
```{r deMicheli-data-kegg-pathway-visualization, message=FALSE, warning=FALSE}
# Load kegg modules object (if not defined)
if(!exists("modules.deMicheli.m.kegg.pathway")){modules.deMicheli.m.kegg.pathway <- readRDS("./../saved/modules_deMicheli_m_kegg_pathway")}

# Create contingency table
tab.pathways.deMicheli <- table(modules.deMicheli.m.kegg.pathway$pathway, modules.deMicheli.m.kegg.pathway$module)

# Melt table
df.pathways.deMicheli <- melt(tab.pathways.deMicheli)
colnames(df.pathways.deMicheli) <- c("KEGG_pathway", "Scorpius_module", "frequency")
df.pathways.deMicheli$KEGG_pathway <- unfactor(df.pathways.deMicheli$KEGG_pathway)

# Handle rows where there are multiple KEGG modules mapping to one scorpius module
df.pathways.deMicheli.separated <- distinct(as.data.frame(separate_rows(df.pathways.deMicheli, KEGG_pathway, sep = ", ")))

# Plot metabolic modules vs. Scorpius modules
p.deMicheli <- ggplot(df.pathways.deMicheli.separated, aes(x = as.factor(Scorpius_module), y = frequency)) +
  geom_bar(aes(fill = KEGG_pathway), stat = "identity", position = position_dodge(1)) +
  ggtitle("Frequency of KEGG pathways in different trajectory modules",
          subtitle = "De Micheli et al. reference dataset") +
  theme(legend.position="right") +
  geom_label_repel(aes(label=ifelse(frequency>=10, KEGG_pathway, "")), position = position_dodge(width=0.9), vjust=-0.5)

p.deMicheli
```


#### 4.2.4 Dell'Orso et al. (2019) reference data
```{r dellOrso-data-kegg-pathway-visualization, message=FALSE, warning=FALSE}
# Load kegg modules object (if not defined)
if(!exists("modules.dellOrso.m.kegg.pathway")){modules.dellOrso.m.kegg.pathway <- readRDS("./../saved/modules_dellOrso_m_kegg_pathway")}

# Create contingency table
tab.pathways.dellOrso <- table(modules.dellOrso.m.kegg.pathway$pathway, modules.dellOrso.m.kegg.pathway$module)

# Melt table
df.pathways.dellOrso <- melt(tab.pathways.dellOrso)
colnames(df.pathways.dellOrso) <- c("KEGG_pathway", "Scorpius_module", "frequency")
df.pathways.dellOrso$KEGG_pathway <- unfactor(df.pathways.dellOrso$KEGG_pathway)

# Handle rows where there are multiple KEGG modules mapping to one scorpius module
df.pathways.dellOrso.separated <- distinct(as.data.frame(separate_rows(df.pathways.dellOrso, KEGG_pathway, sep = ", ")))

# Plot metabolic modules vs. Scorpius modules
p.dellOrso <- ggplot(df.pathways.dellOrso.separated, aes(x = as.factor(Scorpius_module), y = frequency)) +
  geom_bar(aes(fill = as.factor(KEGG_pathway)), stat = "identity", position = position_dodge(1)) +
  ggtitle("Frequency of KEGG pathways in different trajectory modules",
          subtitle = "Dell'Orso et al. referene dataset") +
  theme(legend.position="right") +
  geom_label_repel(aes(label=ifelse(frequency>=6, KEGG_pathway, "")), position = position_dodge(width=0.9), vjust=-0.5)

p.dellOrso
```

**Save plots as pdf**
```{r kegg-pathway-plots-as-pdf, eval=FALSE}
pdf("./../saved/KEGG_pathways_in_trajectories", width = 11, height = 5)
p.primary
p.deMicheli
p.dellOrso
dev.off()
```

## 5. Gene importances using only metabolic genes and subset of cell types
### 5.1 De Micheli et al. (2020) reference data: Without skeletal muscle
#### 5.1.1 Load gene importances
```{r load-gimp-deMicheli-metabolic-genes-no-muscle}
# Load gimp & modules
gimp.deMicheli.m.f <- readRDS("./../saved/gimp_deMicheli_m_no_muscle")
modules.deMicheli.m.f <- readRDS("./../saved/modules_deMicheli_m_no_muscle")

gimp.deMicheli.m.f$qvalue <- p.adjust(p = gimp.deMicheli.m.f$pvalue, method = "BH")
gene_sel.deMicheli.m.f <- gimp.deMicheli.m.f$gene[gimp.deMicheli.m.f$qvalue < .05]

if(!exists("expression.deMicheli.m.f")) {
  muscle_cells.m <- denoised.sce.deMicheli.m$cell_type == "d0_mature_skeletal_muscle" | denoised.sce.deMicheli.m$cell_type =="d5_mature_skeletal_muscle" | denoised.sce.deMicheli.m$cell_type == "d7_mature_skeletal_muscle"
  denoised.sce.deMicheli.m.f <- denoised.sce.deMicheli.m[, !muscle_cells.m]
  expression.deMicheli.m.f <- as.matrix(t(logcounts(denoised.sce.deMicheli.m.f)))
  }
expr_sel.deMicheli.m.f <- scale_quantile(expression.deMicheli.m.f[, gene_sel.deMicheli.m.f])
```

#### 5.1.2 Compute gene importances (optional)
```{r}



#TODO (copy from Puhti script)



```

#### 5.1.3 Visualization of gene importances
```{r visualization-deMicheli-gimp-without-skeletal-muscle}
# Define group name and filtered trajectory
group_name.deMicheli.m.f <- as.factor(denoised.sce.deMicheli.m.f$cell_type)
traj.deMicheli.f.time <- traj.deMicheli$time[names(traj.deMicheli$time) %in% colnames(denoised.sce.deMicheli.m.f)]

# Get rowdata (incl. KEGG modules for each gene) for the metabolic genes
rowdata.deMicheli <- rowData(denoised.sce.deMicheli.m.f)
genes.deMicheli.f <- rownames(rowdata.deMicheli) %in% colnames(expr_sel.deMicheli.m.f)
rowdata.deMicheli.f <- rowdata.deMicheli[genes.deMicheli.f, ]

# Create an annotation row df
ann_row_deMicheli <- data.frame(metabolic_module = factor(rowdata.deMicheli.f$module))
rownames(ann_row_deMicheli) <- rownames(rowdata.deMicheli.f)

# Order genes to same order as in modules object
ann_row_deMicheli <- ann_row_deMicheli[modules.deMicheli.m.f$feature, ,drop=FALSE]

# Add metabolic module information to the modules object
modules.deMicheli.m.f$metabolic_module <- ann_row_deMicheli$metabolic_module

# Arrange modules object by module and by metabolic_module
modules.deMicheli.m.f.kegg <- arrange(modules.deMicheli.m.f,
                                    as.character(modules.deMicheli.m.f$module),
                                    as.character(modules.deMicheli.m.f$metabolic_module))

# save modules object
#saveRDS(modules.deMicheli.m.f.kegg, file = "./../saved/modules_deMicheli_m_f_kegg")

# Gimp
pdf("./../saved/gimp_heatmap_deMicheli_metabolic_genes_no_muscle.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli.m.f, 
                        time = traj.deMicheli.f.time,
                        progression_group = group_name.deMicheli.m.f,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Modules
pdf("./../saved/modules_heatmap_deMicheli_metabolic_genes_no_muscle.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli.m.f, 
                        time = traj.deMicheli.f.time,
                        progression_group = group_name.deMicheli.m.f,
                        modules = modules.deMicheli.m.f.kegg,
                        annotation_row = ann_row_deMicheli,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Save modules object as csv
write.csv(as.data.frame(group_name.deMicheli.m.f), file = "./../saved/modules_deMicheli_m_f_kegg.csv")

```


### 5.2 Dell'Orso et al. (2019) reference data: Without PMs
#### 5.2.1 Load gene importances
```{r load-gimp-dellOrso-metabolic-genes-no-pm}
# Read gimp & modules from a file
gimp.dellOrso.m.f <- readRDS("./../saved/gimp_dellOrso_m_no_primaryMB")
modules.dellOrso.m.f <- readRDS("./../saved/modules_dellOrso_m_no_primaryMB")

# Create expression and sample objects
gimp.dellOrso.m.f$qvalue <- p.adjust(p = gimp.dellOrso.m.f$pvalue, method = "BH")
gene_sel.dellOrso.m.f <- gimp.dellOrso.m.f$gene[gimp.dellOrso.m.f$qvalue < .05]

if(!exists("expression.dellOrso.m.f")) {
  primary_mbs.m <- denoised.sce.dellOrso.m$cell_type == "Primary_MB"
  denoised.sce.dellOrso.m.f <- denoised.sce.dellOrso.m[, !primary_mbs.m]
  expression.dellOrso.m.f <- as.matrix(t(logcounts(denoised.sce.dellOrso.m.f)))
  }
expr_sel.dellOrso.m.f <- scale_quantile(expression.dellOrso.m.f[, gene_sel.dellOrso.m.f])
```


#### 5.2.2 Compute gene importances (optional)
```{r}


#TODO (copy from Puhti script)



```


#### 5.2.3 Visualization of gene importances
```{r dellOrso-gimp-modules-metabolic-genes-no-pm, eval=FALSE}
# Define group name and filtered trajectory
group_name.dellOrso.m.f <- as.factor(denoised.sce.dellOrso.m.f$cell_type)
traj.dellOrso.f.time <- traj.dellOrso$time[names(traj.dellOrso$time) %in% colnames(denoised.sce.dellOrso.m.f)]

# Get rowdata (incl. KEGG modules for each gene) for the metabolic genes
rowdata.dellOrso <- rowData(denoised.sce.dellOrso.m.f)
genes.dellOrso.f <- rownames(rowdata.dellOrso) %in% colnames(expr_sel.dellOrso.m.f)
rowdata.dellOrso.f <- rowdata.dellOrso[genes.dellOrso.f, ]

# Create an annotation row df
ann_row_dellOrso <- data.frame(metabolic_module = factor(rowdata.dellOrso.f$module))
rownames(ann_row_dellOrso) <- rownames(rowdata.dellOrso.f)

# Order genes to same order as in modules object
ann_row_dellOrso <- ann_row_dellOrso[modules.dellOrso.m.f$feature, ,drop=FALSE]

# Add metabolic module information to the modules object
modules.dellOrso.m.f$metabolic_module <- ann_row_dellOrso$metabolic_module

# Arrange modules object by module and by metabolic_module
modules.dellOrso.m.f.kegg <- arrange(modules.dellOrso.m.f,
                                   as.character(modules.dellOrso.m.f$module),
                                   as.character(modules.dellOrso.m.f$metabolic_module))

# save modules object
#saveRDS(modules.dellOrso.m.f.kegg, file = "./../saved/modules_dellOrso_m_f_kegg")

# Gimp
pdf("./../saved/gimp_heatmap_dellOrso_metabolic_genes_no_PM.pdf")
draw_trajectory_heatmap(expr_sel.dellOrso.m.f, 
                        time = traj.dellOrso.f.time,
                        progression_group = group_name.dellOrso.m.f,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Modules
pdf("./../saved/modules_heatmap_dellOrso_metabolic_genes_no_PM.pdf")
draw_trajectory_heatmap(expr_sel.dellOrso.m.f, 
                        time = traj.dellOrso.f.time,
                        progression_group = group_name.dellOrso.m.f,
                        modules = modules.dellOrso.m.f,
                        annotation_row = ann_row_dellOrso,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Save modules object as csv
write.csv(as.data.frame(modules.dellOrso.m.f.kegg), file = "./../saved/modules_dellOrso_m_f_kegg.csv")
```


### 5.3 Comparison of metabolic genes between the datasets
```{r comparison-of-metabolic-genes-between-datasets}
# Define colour palette for venn diagram
venn.colors <- brewer.pal(3, "Pastel2")

# Load metabolic genes if not defined
if(!exists("modules.primary.m.kegg")) {modules.primary.m.kegg <- readRDS("./../saved/modules_primary_m_kegg")}
if(!exists("modules.deMicheli.m.f.kegg")) {modules.deMicheli.m.f.kegg <- readRDS("./../saved/modules_deMicheli_m_f_kegg")}
if(!exists("modules.dellOrso.m.f.kegg")) {modules.dellOrso.m.f.kegg <- readRDS("./../saved/modules_dellOrso_m_f_kegg")}

m.genes.primary <- modules.primary.m.kegg$feature
m.genes.deMicheli.f <- modules.deMicheli.m.f.kegg$feature
m.genes.dellOrso.f <- modules.dellOrso.m.f.kegg$feature

# Creata a Venn diagram from metabolic genes
venn <- venn.diagram(
  x = list(m.genes.primary, m.genes.deMicheli.f, m.genes.dellOrso.f),
  category.names = c("Primary data" , "de Micheli " , "Dell'Orso"),
  filename = NULL,
  fill = venn.colors,
  fontface = "bold",
  lwd = 2,
  lty = "blank",
  cat.fontface = "bold",
  fontfamily = "sans",
  cat.fontfamily = "sans"
)

p.gg <- as_ggplot(venn)
p.gg

# Inspect the genes
## Intersection (all)
intersect(m.genes.primary, m.genes.deMicheli.f) %>%
  intersect(m.genes.dellOrso.f)

##Intersection: primary - de Micheli
intersect(m.genes.primary, m.genes.deMicheli.f)

## Intersection: primary - dell'Orso
intersect(m.genes.primary, m.genes.dellOrso.f)

## Intersection: de Micheli - Dell'Orso
intersect(m.genes.deMicheli.f, m.genes.dellOrso.f)

## Union (all)
union(m.genes.primary, m.genes.deMicheli.f) %>% 
  union(m.genes.dellOrso.f)

```


## 6. SingleR annotation using only metabolic genes
### 6.1 De Micheli predictions (metabolic genes)
### 6.1.1 All cell types
#### 6.1.1.1 Predictions
```{r singleR-deMicheli-metabolic-genes}
# Load prediction objects
pred.deMicheli.m <- readRDS("./../saved/pred_deMicheli_m")

# Inspect results
tab.deMicheli.m <- t(table(pred.deMicheli.m$pruned.labels, denoised.sce.m$sample)[, c("control", "cap50", "cap50_r4h","cap50_r8h","cap50_r16h")])
tab.deMicheli.m
```

#### 6.1.1.2 Diagnostic plots
```{r deMicheli-annotation-diagnostic-plots-metabolic-genes}
# Score heatmap
plotScoreHeatmap(pred.deMicheli.m, annotation_col = as.data.frame(colData(denoised.sce.m)[, "sample", drop = FALSE]), width = 10)

# Remove low-quality cells
to.remove <- pruneScores(pred.deMicheli.m, min.diff.med = 0.2)
table(Label=pred.deMicheli.m$labels, Removed=to.remove)

# Delta distribution
plotScoreDistribution(results = pred.deMicheli.m, show = "delta.med")
```


### 6.1.2 Without mature skeletal muscle
#### 6.1.2.1 Predictions
```{r singleR-deMicheli-metabolic-genes-no-muscle}
# Load prediction objects
pred.deMicheli.m.no.muscle <- readRDS("./../saved/pred_deMicheli_m_no_muscle")

# Inspect results
tab.deMicheli.m <- t(table(pred.deMicheli.m.no.muscle$pruned.labels, denoised.sce.m$sample)[, c("control", "cap50", "cap50_r4h","cap50_r8h","cap50_r16h")])
tab.deMicheli.m
```
#### 6.1.2.2 Diagnostic plots
```{r deMicheli-annotation-diagnostic-plots-metabolic-genes-no-muscle}
# Score heatmap
plotScoreHeatmap(pred.deMicheli.m.no.muscle, annotation_col = as.data.frame(colData(denoised.sce.m)[, "sample", drop = FALSE]), width = 10)

# Remove low-quality cells
to.remove <- pruneScores(pred.deMicheli.m.no.muscle, min.diff.med = 0.2)
table(Label=pred.deMicheli.m.no.muscle$labels, Removed=to.remove)

# Delta distribution
plotScoreDistribution(results = pred.deMicheli.m.no.muscle, show = "delta.med")
```


## 6.2 Dell'Orso predictions (metabolic genes)
### 6.2.1 All cell types
#### 6.2.1.1 Predictions

**Load pre-calculated predictions**
```{r load-singleR-dellOrso-metabolic-genes}
# Load prediction objects
pred.dellOrso.m <- readRDS("./../saved/pred_dellOrso_m")

# Inspect results
tab.dellOrso.m <- t(table(pred.dellOrso.m$pruned.labels, denoised.sce.m$sample)[, c("control", "cap50", "cap50_r4h","cap50_r8h","cap50_r16h")])
tab.dellOrso.m
```
**Perform singleR predictions**
```{r singleR-dellOrso-metabolic-genes}



# TODO: copy from Puhti script



```



#### 6.2.1.2 Diagnostic plots
```{r dellOrso-annotation-diagnostic-plots-metabolic-genes}
# Score heatmap
plotScoreHeatmap(pred.dellOrso.m, annotation_col = as.data.frame(colData(denoised.sce.m)[, "sample", drop = FALSE]), width = 10)

# Remove low-quality cells
to.remove <- pruneScores(pred.dellOrso.m, min.diff.med = 0.2)
table(Label=pred.dellOrso.m$labels, Removed=to.remove)

# Delta distribution
plotScoreDistribution(results = pred.dellOrso.m, show = "delta.med")
```


### 6.2.2 Without Primary Myoblasts (PMs)
#### 6.2.2.1 Predictions
**Load pre-calculated predictions**
```{r load-singleR-dellOrso-metabolic-genes-no-pm}
# Load prediction objects
pred.dellOrso.m.no.pm <- readRDS("./../saved/pred_dellOrso_m_no_pm")

# Inspect results
tab.dellOrso.m <- t(table(pred.dellOrso.m.no.pm$pruned.labels, denoised.sce.m$sample)[, c("control", "cap50", "cap50_r4h","cap50_r8h","cap50_r16h")])
tab.dellOrso.m
```
**Perform singleR predictions**
```{r singleR-dellOrso-metabolic-genes-no-pm}



# TODO: copy from Puhti script



```


#### 6.2.2.2 Diagnostic plots
```{r dellOrso-annotation-diagnostic-plots-metabolic-genes-no-pm}
# Score heatmap
plotScoreHeatmap(pred.dellOrso.m.no.pm, annotation_col = as.data.frame(colData(denoised.sce.m)[, "sample", drop = FALSE]), width = 10)

# Remove low-quality cells
to.remove <- pruneScores(pred.dellOrso.m.no.pm, min.diff.med = 0.2)
table(Label=pred.dellOrso.m.no.pm$labels, Removed=to.remove)

# Delta distribution
plotScoreDistribution(results = pred.dellOrso.m.no.pm, show = "delta.med")
```

## 7. DE analysis of primary data metabolic genes
### 7.1 Marker genes: Control vs. Cap
```{r marker-genes-metabolic-primary-data}
markers.primary <- findMarkers(denoised.sce.m,
                               groups = denoised.sce.m$sample,
                               lfc = 1)

control.markers <- markers.primary[["control"]]
control.markers.interesting <- control.markers[control.markers$p.value <= 0.05, ]
logFCs <- getMarkerEffects(control.markers.interesting)

pheatmap(logFCs, breaks=seq(-2.5, 2.5, length.out=101), main = "Marker genes: Control sample vs. Cap samples (metabolic genes)", fontsize_row = 12)
```

### 7.2 Inspection of marker genes on trajectory
```{r control-marker-genes-on-deMicheli-trajectory}
# Control marker genes
marker.genes.control <- rownames(control.markers.interesting)

marker.gene.plots <- list()

# Define de Micheli trajectory space
expression.deMicheli <- as.matrix(t(logcounts(denoised.sce.deMicheli)))
set.seed(123)
space.deMicheli <- reduce_dimensionality(expression.deMicheli, dist = "spearman", ndim = 2)

# Plot trajectory for each marker
for(gene in marker.genes.control) {
  p <- draw_trajectory_plot(space = space.deMicheli, 
                            progression_group = expression.deMicheli[, gene],
                            point_size = 1,
                            point_alpha = 0.5,
                            path = traj.deMicheli$path,
                            contour = FALSE) + 
  ggtitle(paste(gene, "gene")) +
  theme(plot.title = element_text(size = 14, face = "bold"), axis.title = element_text(size = 10))
  marker.gene.plots[[gene]] <- p
}

gridExtra::grid.arrange(marker.gene.plots$Cth, marker.gene.plots$Taldo1, marker.gene.plots$Gclm, marker.gene.plots$Psph, marker.gene.plots$Gal, marker.gene.plots$Pgd,
                        marker.gene.plots$Cox6a2, marker.gene.plots$Idi1, marker.gene.plots$Sat1, marker.gene.plots$Pck2, marker.gene.plots$Maoa, marker.gene.plots$Gbe1,
                        marker.gene.plots$Eno3, marker.gene.plots$Atp6v1g1, marker.gene.plots$Hgsnat, marker.gene.plots$Atp6v0b, marker.gene.plots$Hexa, marker.gene.plots$Uap1l1,
                        marker.gene.plots$Aldh2, ncol = 4)
```

```{r control-marker-genes-on-dellOrso-trajectory}
# Control marker genes
marker.genes.control <- rownames(control.markers.interesting)

marker.gene.plots <- list()

# Define de Micheli trajectory space
expression.dellOrso <- as.matrix(t(logcounts(denoised.sce.dellOrso)))
set.seed(123)
space.dellOrso <- reduce_dimensionality(expression.dellOrso, dist = "spearman", ndim = 2)

# Plot trajectory for each marker
for(gene in marker.genes.control) {
  p <- draw_trajectory_plot(space = space.dellOrso, 
                            progression_group = expression.dellOrso[, gene],
                            point_size = 1,
                            point_alpha = 0.5,
                            path = traj.dellOrso$path,
                            contour = FALSE) + 
  ggtitle(paste(gene, "gene")) +
  theme(plot.title = element_text(size = 14, face = "bold"), axis.title = element_text(size = 10))
  marker.gene.plots[[gene]] <- p
}

gridExtra::grid.arrange(marker.gene.plots$Cth, marker.gene.plots$Taldo1, marker.gene.plots$Gclm, marker.gene.plots$Psph, marker.gene.plots$Gal, marker.gene.plots$Pgd,
                        marker.gene.plots$Cox6a2, marker.gene.plots$Idi1, marker.gene.plots$Sat1, marker.gene.plots$Pck2, marker.gene.plots$Maoa, marker.gene.plots$Gbe1,
                        marker.gene.plots$Eno3, marker.gene.plots$Atp6v1g1, marker.gene.plots$Hgsnat, marker.gene.plots$Atp6v0b, marker.gene.plots$Hexa, marker.gene.plots$Uap1l1,
                        marker.gene.plots$Aldh2, ncol = 4)
```

