---
title: "Metabolic_pathways_pt2"
author: "Ville Lehtonen"
date: "11/9/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Set working dir to the source file location
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
```

# Metabolic pathways
Here, we perform the aggregation of gene expression values into KEGG module / pathway expression and perform various types of analyses on this aggregated dataset.

## 1. Load packages and data
### 1.1 Load packages
```{r load-packages, message=FALSE}
library(SCORPIUS)
library(SingleCellExperiment)
library(stats)
library(data.table)
```

### 1.2 Load pre-calculated data
**Single Cell Experiments**
```{r load-sce-objects}
# Load datasets with metabolic genes only
denoised.sce.m <- readRDS("./../saved/denoised_sce_m")
denoised.sce.deMicheli.m <- readRDS("./../saved/denoised_sce_deMicheli_m")
denoised.sce.dellOrso.m <- readRDS("./../saved/denoised_sce_dellOrso_m")
```

**Trajectories**
```{r load-trajectories}
traj.primary <- readRDS("./../saved/traj_primary")
traj.deMicheli <- readRDS("./../saved/traj_deMicheli")
traj.dellOrso <- readRDS("./../saved/traj_dellOrso")
```

**Aggregated sce objects**
```{r load-aggr-sce-objects}
# Primary
sce.primary.module <- readRDS("./../saved/sce_primary_module")
sce.primary.pathway <- readRDS("./../saved/sce_primary_pathway")

# De Micheli
sce.deMicheli.module <- readRDS("./../saved/sce_deMicheli_module")
sce.deMicheli.pathway <- readRDS("./../saved/sce_deMicheli_pathway")

# Dell Orso
sce.dellOrso.module <- readRDS("./../saved/sce_dellOrso_module")
sce.dellOrso.pathway <- readRDS("./../saved/sce_dellOrso_pathway")
```

**Fetch KEGG module and pathway information**
```{r fetch-kegg-entries-for-mouse-pathways-and-modules}
# Get the KEGG entries for modules / pathways
mouse.org.id <- "T01002"
mouse.modules <- keggLink(target = "module", source = mouse.org.id)
mouse.pathways <- keggLink(target = "pathway", source = mouse.org.id)
```



## 2. Perform aggregation of gene expression into pathway expression
### 2.1 Primary dataset
```{r primary-data-KEGG-module-pathway-aggregation, eval=FALSE}
###################################################################
# Get a list of all metabolic pathways/modules present in the data
###################################################################
all_pathways <- c()
all_modules <- c()

for (pathway_set in rowData(denoised.sce.m)$pathway) {
  new_pathways <- strsplit(pathway_set, split = ", ")[[1]] %>% lapply(trimws) %>% unlist()
  all_pathways <- unique(c(all_pathways, new_pathways))
}

for (module_set in rowData(denoised.sce.m)$module) {
  new_modules <- strsplit(module_set, split = ", ")[[1]] %>% lapply(trimws) %>% unlist()
  all_modules <- unique(c(all_modules, new_modules))
}

################################
# Module expression aggregation
################################
# Create a new data frame for storing aggregated expression values for each module
module.expression.profile <- data.frame()

# Aggregate expression for each module
for (module in all_modules) {
  print(paste(module))
  # Get all genes linked to a module and create a subsetted sce object from these
  module.genes <- grepl(module, rowData(denoised.sce.m)$module, fixed = TRUE)
  sce.module.genes <- denoised.sce.m[module.genes, ]
  # Aggregate expression of module genes and transform into a 1-row dataframe
  module.expression <- t(as.data.frame(colSums(as.matrix(logcounts(sce.module.genes)), na.rm = TRUE), drop=FALSE))
  # Divide the expression of the module in each cell by the number of genes contributing to this module (to get the avg. expression)
  module.expression <- module.expression / nrow(sce.module.genes)
  # Add a new aggregated module expression row into the module.expression.profile dataframe
  module.expression.profile <- rbind(module.expression.profile, module.expression)
}
# Add correct rownames to the expression object
rownames(module.expression.profile) <- all_modules

# Create a SingleCellExperiment object to store the data
sce.primary.module <- SingleCellExperiment(assays = list(logcounts = as.matrix(module.expression.profile)),
                                           colData = colData(denoised.sce.m))


################################
# Pathway expression aggregation
################################
# Create a new data frame for storing aggregated expression values for each module
pathway.expression.profile <- data.frame()

# Aggregate expression for each pathway
for (pathway in all_pathways) {
  print(paste(pathway))
  # Get all genes linked to a pathway and create a subsetted sce object from these
  pathway.genes <- grepl(pathway, rowData(denoised.sce.m)$pathway, fixed = TRUE)
  sce.pathway.genes <- denoised.sce.m[pathway.genes, ]
  # Aggregate expression of pathway genes and transform into a 1-row dataframe
  pathway.expression <- t(as.data.frame(colSums(as.matrix(logcounts(sce.pathway.genes)), na.rm = TRUE), drop = FALSE))
  # Divide the expression of the pathway in each cell by the number of genes contributing to this module (to get the avg. expression)
  pathway.expression <- pathway.expression / nrow(sce.pathway.genes)
  # Add a new aggregated pathway expression row into the pathway.expression.profile dataframe
  pathway.expression.profile <- rbind(pathway.expression.profile, pathway.expression)
}

# Add correct rownames to the expression object
rownames(pathway.expression.profile) <- all_pathways

# Create a SingleCellExperiment object to store the data
sce.primary.pathway <- SingleCellExperiment(assays = list(logcounts = as.matrix(pathway.expression.profile)),
                                            colData = colData(denoised.sce.m))


###########################################
# Save aggregated module & pathway objects
###########################################
saveRDS(sce.primary.module, file = "./../saved/sce_primary_module")
saveRDS(sce.primary.pathway, file = "./../saved/sce_primary_pathway")
```

### 2.2 De Micheli dataset
```{r deMicheli-data-KEGG-module-pathway-aggregation, eval=FALSE}
###################################################################
# Get a list of all metabolic pathways/modules present in the data
###################################################################
all_pathways <- c()
all_modules <- c()

for (pathway_set in rowData(denoised.sce.deMicheli.m)$pathway) {
  new_pathways <- strsplit(pathway_set, split = ", ")[[1]] %>% lapply(trimws) %>% unlist()
  all_pathways <- unique(c(all_pathways, new_pathways))
}

for (module_set in rowData(denoised.sce.deMicheli.m)$module) {
  new_modules <- strsplit(module_set, split = ", ")[[1]] %>% lapply(trimws) %>% unlist()
  all_modules <- unique(c(all_modules, new_modules))
}

################################
# Module expression aggregation
################################
# Create a new data frame for storing aggregated expression values for each module
module.expression.profile <- data.frame()

# Aggregate expression for each module
for (module in all_modules) {
  print(paste(module))
  # Get all genes linked to a module and create a subsetted sce object from these
  module.genes <- grepl(module, rowData(denoised.sce.deMicheli.m)$module, fixed = TRUE)
  sce.module.genes <- denoised.sce.deMicheli.m[module.genes, ]
  # Aggregate expression of module genes and transform into a 1-row dataframe
  module.expression <- t(as.data.frame(colSums(as.matrix(logcounts(sce.module.genes)), na.rm = TRUE), drop=FALSE))
  # Divide the expression of the module in each cell by the number of genes contributing to this module (to get the avg. expression)
  module.expression <- module.expression / nrow(sce.module.genes)
  # Add a new aggregated module expression row into the module.expression.profile dataframe
  module.expression.profile <- rbind(module.expression.profile, module.expression)
}
# Add correct rownames to the expression object
rownames(module.expression.profile) <- all_modules

# Create a SingleCellExperiment object to store the data
sce.deMicheli.module <- SingleCellExperiment(assays = list(logcounts = as.matrix(module.expression.profile)),
                                             colData = colData(denoised.sce.deMicheli.m))


################################
# Pathway expression aggregation
################################
# Create a new data frame for storing aggregated expression values for each module
pathway.expression.profile <- data.frame()

# Aggregate expression for each pathway
for (pathway in all_pathways) {
  print(paste(pathway))
  # Get all genes linked to a pathway and create a subsetted sce object from these
  pathway.genes <- grepl(pathway, rowData(denoised.sce.deMicheli.m)$pathway, fixed = TRUE)
  sce.pathway.genes <- denoised.sce.deMicheli.m[pathway.genes, ]
  # Aggregate expression of pathway genes and transform into a 1-row dataframe
  pathway.expression <- t(as.data.frame(colSums(as.matrix(logcounts(sce.pathway.genes)), na.rm = TRUE), drop = FALSE))
  # Divide the expression of the pathway in each cell by the number of genes contributing to this module (to get the avg. expression)
  pathway.expression <- pathway.expression / nrow(sce.pathway.genes)
  # Add a new aggregated pathway expression row into the pathway.expression.profile dataframe
  pathway.expression.profile <- rbind(pathway.expression.profile, pathway.expression)
}

# Add correct rownames to the expression object
rownames(pathway.expression.profile) <- all_pathways

# Create a SingleCellExperiment object to store the data
sce.deMicheli.pathway <- SingleCellExperiment(assays = list(logcounts = as.matrix(pathway.expression.profile)),
                                            colData = colData(denoised.sce.deMicheli.m))


###########################################
# Save aggregated module & pathway objects
###########################################
saveRDS(sce.deMicheli.module, file = "./../saved/sce_deMicheli_module")
saveRDS(sce.deMicheli.pathway, file = "./../saved/sce_deMicheli_pathway")
```

### 2.3 Dell'Orso dataset
```{r dellOrso-data-KEGG-module-pathway-aggregation, eval=FALSE}
###################################################################
# Get a list of all metabolic pathways/modules present in the data
###################################################################
all_pathways <- c()
all_modules <- c()

for (pathway_set in rowData(denoised.sce.dellOrso.m)$pathway) {
  new_pathways <- strsplit(pathway_set, split = ", ")[[1]] %>% lapply(trimws) %>% unlist()
  all_pathways <- unique(c(all_pathways, new_pathways))
}

for (module_set in rowData(denoised.sce.dellOrso.m)$module) {
  new_modules <- strsplit(module_set, split = ", ")[[1]] %>% lapply(trimws) %>% unlist()
  all_modules <- unique(c(all_modules, new_modules))
}

################################
# Module expression aggregation
################################
# Create a new data frame for storing aggregated expression values for each module
module.expression.profile <- data.frame()

# Aggregate expression for each module
for (module in all_modules) {
  print(paste(module))
  # Get all genes linked to a module and create a subsetted sce object from these
  module.genes <- grepl(module, rowData(denoised.sce.dellOrso.m)$module, fixed = TRUE)
  sce.module.genes <- denoised.sce.dellOrso.m[module.genes, ]
  # Aggregate expression of module genes and transform into a 1-row dataframe
  module.expression <- t(as.data.frame(colSums(as.matrix(logcounts(sce.module.genes)), na.rm = TRUE), drop=FALSE))
  # Divide the expression of the module in each cell by the number of genes contributing to this module (to get the avg. expression)
  module.expression <- module.expression / nrow(sce.module.genes)
  # Add a new aggregated module expression row into the module.expression.profile dataframe
  module.expression.profile <- rbind(module.expression.profile, module.expression)
}
# Add correct rownames to the expression object
rownames(module.expression.profile) <- all_modules

# Create a SingleCellExperiment object to store the data
sce.dellOrso.module <- SingleCellExperiment(assays = list(logcounts = as.matrix(module.expression.profile)),
                                             colData = colData(denoised.sce.dellOrso.m))


################################
# Pathway expression aggregation
################################
# Create a new data frame for storing aggregated expression values for each module
pathway.expression.profile <- data.frame()

# Aggregate expression for each pathway
for (pathway in all_pathways) {
  print(paste(pathway))
  # Get all genes linked to a pathway and create a subsetted sce object from these
  pathway.genes <- grepl(pathway, rowData(denoised.sce.dellOrso.m)$pathway, fixed = TRUE)
  sce.pathway.genes <- denoised.sce.dellOrso.m[pathway.genes, ]
  # Aggregate expression of pathway genes and transform into a 1-row dataframe
  pathway.expression <- t(as.data.frame(colSums(as.matrix(logcounts(sce.pathway.genes)), na.rm = TRUE), drop = FALSE))
  # Divide the expression of the pathway in each cell by the number of genes contributing to this module (to get the avg. expression)
  pathway.expression <- pathway.expression / nrow(sce.pathway.genes)
  # Add a new aggregated pathway expression row into the pathway.expression.profile dataframe
  pathway.expression.profile <- rbind(pathway.expression.profile, pathway.expression)
}

# Add correct rownames to the expression object
rownames(pathway.expression.profile) <- all_pathways

# Create a SingleCellExperiment object to store the data
sce.dellOrso.pathway <- SingleCellExperiment(assays = list(logcounts = as.matrix(pathway.expression.profile)),
                                            colData = colData(denoised.sce.dellOrso.m))


###########################################
# Save aggregated module & pathway objects
###########################################
saveRDS(sce.dellOrso.module, file = "./../saved/sce_dellOrso_module")
saveRDS(sce.dellOrso.pathway, file = "./../saved/sce_dellOrso_pathway")
```


## 3. Calculate module / pathway importances
### 3.1 Primary data
#### 3.1.1 - Load pre-saved objects
```{r load-pre-saved-module-and-pathway-importances}
# Module
## Define expression and group objects
expression.primary.module <- as.matrix(t(logcounts(sce.primary.module)))
group_name.primary.module <- as.factor(sce.primary.module$sample)

## Get gimp & module objects
gimp.primary.module <- readRDS("./../saved/gimp_primary_module")
modules.primary.module <- readRDS("./../saved/modules_primary_module")
gimp.primary.module$qvalue <- p.adjust(p = gimp.primary.module$pvalue, method = "BH")
gene_sel.primary.module <- gimp.primary.module$gene[gimp.primary.module$qvalue < .05]
expr_sel.primary.module <- scale_quantile(expression.primary.module[, gene_sel.primary.module])


# Pathway
## Define expression and group objects
expression.primary.pathway <- as.matrix(t(logcounts(sce.primary.pathway)))
group_name.primary.pathway <- as.factor(sce.primary.pathway$sample)

## Get gimp & module objects
gimp.primary.pathway <- readRDS("./../saved/gimp_primary_pathway")
modules.primary.pathway <- readRDS("./../saved/modules_primary_pathway")
gimp.primary.pathway$qvalue <- p.adjust(p = gimp.primary.pathway$pvalue, method = "BH")
gene_sel.primary.pathway <- gimp.primary.pathway$gene[gimp.primary.pathway$qvalue < .05]
expr_sel.primary.pathway <- scale_quantile(expression.primary.pathway[, gene_sel.primary.pathway])
```



#### 3.1.2 Compute Module importances (optional)
```{r compute-primary-module-importances, eval=FASLE}
# Load required objects (if required)
if(!exists("sce.primary.module")) {sce.primary.module <- readRDS("./../saved/sce_primary_module")}

# Create expression and sample objects
expression.primary.module <- as.matrix(t(logcounts(sce.primary.module)))
group_name.primary.module <- as.factor(sce.primary.module$sample)

# Compute gene importances
gimp.primary.module <- gene_importances(x = expression.primary.module,
                                        time = traj.primary$time,
                                        num_permutations = 10,
                                        num_threads = 8)

gimp.primary.module$qvalue <- p.adjust(p = gimp.primary.module$pvalue, method = "BH")
gene_sel.primary.module <- gimp.primary.module$gene[gimp.primary.module$qvalue < .05]
expr_sel.primary.module <- scale_quantile(expression.primary.module[, gene_sel.primary.module])

saveRDS(gimp.primary.module, file = "./../saved/gimp_primary_module")

# Extract modules
modules.primary.module <- extract_modules(expr_sel.primary.module,
                                          time = traj.primary$time,
                                          verbose = FALSE)

saveRDS(modules.primary.module, file = "./../saved/modules_primary_module")
```

#### 3.1.3 Compute pathway importances (optional)
```{r compute-primary-pathway-importances, eval=FASLE}

#TODO

```

#### 3.1.4 get KEGG module / pathway information
```{r}
# Add 'md:' and 'path:' identifiers to the list of important modules/pathways
important.modules.primary <- paste0("md:", colnames(expr_sel.primary.module))
important.pathways.primary <- paste0("path:", colnames(expr_sel.primary.pathway))

# Filter only unique modules / pathways with their associated KEGG entries
mouse.modules.f <- mouse.modules[mouse.modules %in% important.modules.primary]
mouse.modules.f <- mouse.modules.f[!duplicated(mouse.modules.f)]
mouse.pathways.f <- mouse.pathways[mouse.pathways %in% important.pathways.primary]
mouse.pathways.f <- mouse.pathways.f[!duplicated(mouse.pathways.f)]

# Query module information with KEGG entries
df.module.info <- data.frame(module = NA, module.info = NA)

for (i in 1:length(names(mouse.modules.f))) {
  entry <- names(mouse.modules.f)[i]
  # Get module and its corresponding name
  query <- keggGet(entry)
  if("mmu01100" %in% names(query[[1]]$PATHWAY)) {
    module.id <- mouse.modules.f[[i]] %>% substring(4)
    module.info <- query[[1]]$MODULE[names(query[[1]]$MODULE) == module.id]
    print(paste("i", i, " entry:", entry, " module id:", module.id, " module info:", module.info))
    
    # Updata values to dataframe
    df.module.info[i, ]$module <- module.id
    df.module.info[i, ]$module.info <- module.info
  }
  }

# Query pathway information with KEGG entries
df.pathway.info <- data.frame(pathway = NA, pathway.info = NA)

for (i in 1:length(names(mouse.pathways.f))) {
  entry <- names(mouse.pathways.f)[i]
  # Get pathway and its corresponding name
  query <- keggGet(entry)
  if("mmu01100" %in% names(query[[1]]$PATHWAY)) {
    pathway.id <- mouse.pathways.f[[i]] %>% substring(6)
    pathway.info <- query[[1]]$PATHWAY[names(query[[1]]$PATHWAY) == pathway.id]
    print(paste("i", i, " entry:", entry, " pathway id:", pathway.id, " pathway info:", pathway.info))
    
    # Updata values to dataframe
    df.pathway.info[i, ]$pathway <- pathway.id
    df.pathway.info[i, ]$pathway.info <- pathway.info
  }
}

# save objects
saveRDS(df.module.info, file = "./../saved/df_module_info_primary")
saveRDS(df.pathway.info, file = "./../saved/df_pathway_info_primary")
write.csv(df.module.info, file = "./../saved/df_module_info_primary.csv")
write.csv(df.pathway.info, file = "./../saved/df_pathway_info_primary.csv")
```

#### 3.1.5 Visualize module and pathway importances
```{r primary-data-visualize-module-and-pathway-importances, eval=FALSE}
# Modules
pdf("./../saved/modules_heatmap_primary_KEGG_modules.pdf")
draw_trajectory_heatmap(expr_sel.primary.module, 
                        time = traj.primary$time,
                        progression_group = group_name.primary.module,
                        modules = modules.primary.module,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Pathways
pdf("./../saved/modules_heatmap_primary_KEGG_pathways.pdf")
draw_trajectory_heatmap(expr_sel.primary.pathway, 
                        time = traj.primary$time,
                        progression_group = group_name.primary.pathway,
                        modules = modules.primary.pathway,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

```




### 3.2 De Micheli data
#### 3.2.1 - Load pre-saved objects
**All cell types**
```{r load-deMicheli-module-and-pathway-importances}
# Module
## Define expression and group objects
expression.deMicheli.module <- as.matrix(t(logcounts(sce.deMicheli.module)))
group_name.deMicheli.module <- as.factor(sce.deMicheli.module$cell_type)

## Get gimp & module objects
gimp.deMicheli.module <- readRDS("./../saved/gimp_deMicheli_module")
modules.deMicheli.module <- readRDS("./../saved/modules_deMicheli_module")
gimp.deMicheli.module$qvalue <- p.adjust(p = gimp.deMicheli.module$pvalue, method = "BH")
gene_sel.deMicheli.module <- gimp.deMicheli.module$gene[gimp.deMicheli.module$qvalue < .05]
expr_sel.deMicheli.module <- scale_quantile(expression.deMicheli.module[, gene_sel.deMicheli.module])


# Pathway
## Define expression and group objects
expression.deMicheli.pathway <- as.matrix(t(logcounts(sce.deMicheli.pathway)))
group_name.deMicheli.pathway <- as.factor(sce.deMicheli.pathway$cell_type)

## Get gimp & module objects
gimp.deMicheli.pathway <- readRDS("./../saved/gimp_deMicheli_pathway")
modules.deMicheli.pathway <- readRDS("./../saved/modules_deMicheli_pathway")
gimp.deMicheli.pathway$qvalue <- p.adjust(p = gimp.deMicheli.pathway$pvalue, method = "BH")
gene_sel.deMicheli.pathway <- gimp.deMicheli.pathway$gene[gimp.deMicheli.pathway$qvalue < .05]
expr_sel.deMicheli.pathway <- scale_quantile(expression.deMicheli.pathway[, gene_sel.deMicheli.pathway])
```

**Without mature skeletal muscle**
```{r load-deMicheli-module-and-pathway-importances-no-muscle}

```


#### 3.2.2 Compute module importances (optional)
```{r compute-deMicheli-module-importances, eval=FASLE}
#TODO
```

#### 3.2.3 Compute pathway importances (optional)
```{r compute-deMicheli-pathway-importances, eval=FASLE}
#TODO
```

#### 3.2.4 Get KEGG module / pathway information
```{r deMicheli-fetch-KEGG-module-and-pathway-information}
# Get the important modules / pathways and their associated KEGG entries
# Add 'md:' and 'path:' identifiers to the list of important modules/pathways
important.modules.deMicheli <- paste0("md:", colnames(expr_sel.deMicheli.module))
important.pathways.deMicheli <- paste0("path:", colnames(expr_sel.deMicheli.pathway))

# Filter only unique modules / pathways with their associated KEGG entries
mouse.modules.f <- mouse.modules[mouse.modules %in% important.modules.deMicheli]
mouse.modules.f <- mouse.modules.f[!duplicated(mouse.modules.f)]
mouse.pathways.f <- mouse.pathways[mouse.pathways %in% important.pathways.deMicheli]
mouse.pathways.f <- mouse.pathways.f[!duplicated(mouse.pathways.f)]

# Query module information with KEGG entries
df.module.info <- data.frame(module = NA, module.info = NA)

for (i in 1:length(names(mouse.modules.f))) {
  entry <- names(mouse.modules.f)[i]
  # Get module and its corresponding name
  query <- keggGet(entry)
  module.id <- mouse.modules.f[[i]] %>% substring(4)
  module.info <- query[[1]]$MODULE[names(query[[1]]$MODULE) == module.id]
  print(paste("i", i, " entry:", entry, " module id:", module.id, " module info:", module.info))
  
  # Updata values to dataframe
  df.module.info[i, ]$module <- module.id
  df.module.info[i, ]$module.info <- module.info
}

# Query pathway information with KEGG entries
df.pathway.info <- data.frame(pathway = NA, pathway.info = NA)

for (i in 1:length(names(mouse.pathways.f))) {
  entry <- names(mouse.pathways.f)[i]
  # Get pathway and its corresponding name
  query <- keggGet(entry)
  pathway.id <- mouse.pathways.f[[i]] %>% substring(6)
  pathway.info <- query[[1]]$PATHWAY[names(query[[1]]$PATHWAY) == pathway.id]
  print(paste("i", i, " entry:", entry, " pathway id:", pathway.id, " pathway info:", pathway.info))
  
  # Updata values to dataframe
  df.pathway.info[i, ]$pathway <- pathway.id
  df.pathway.info[i, ]$pathway.info <- pathway.info
}

# save objects
saveRDS(df.module.info, file = "./../saved/df_module_info_deMicheli")
saveRDS(df.pathway.info, file = "./../saved/df_pathway_info_deMicheli")
write.csv(df.module.info, file = "./../saved/df_module_info_deMicheli.csv")
write.csv(df.pathway.info, file = "./../saved/df_pathway_info_deMicheli.csv")
```


#### 3.2.5 Visualize module and pathway importances
**Module importances**
```{r deMicheli-module-importance-heatmap, eval=FALSE}
# Create annotation row for modules
ann.row.deMicheli.modules <- data.frame(KEGG_module = factor(df.module.info$module.info))
rownames(ann.row.deMicheli.modules) <- df.module.info$module

# Create annotation row for pathways
ann.row.deMicheli.pathways <- data.frame(KEGG_pathway = factor(df.pathway.info$pathway.info))
rownames(ann.row.deMicheli.pathways) <- df.pathway.info$pathway

# Modules
pdf("./../saved/modules_heatmap_deMicheli_KEGG_modules.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli.module, 
                        time = traj.deMicheli$time,
                        progression_group = group_name.deMicheli.module,
                        annotation_row = ann.row.deMicheli.modules,
                        modules = modules.deMicheli.module,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()

# Pathways
pdf("./../saved/modules_heatmap_deMicheli_KEGG_pathways.pdf")
draw_trajectory_heatmap(expr_sel.deMicheli.pathway, 
                        time = traj.deMicheli$time,
                        progression_group = group_name.deMicheli.pathway,
                        annotation_row = ann.row.deMicheli.pathways,
                        modules = modules.deMicheli.pathway,
                        show_labels_row = TRUE,
                        fontsize_row = 5,
                        fontsize = 4)
dev.off()
```


**Pathway importances**
```{r deMicheli-pathway-importance-heatmap, eval=FALSE}

```




### 3.3 Dell'Orso data
#### 3.3.1 Compute module importances (optional)
```{r compute-dellOrso-module-importances, eval=FASLE}
#TODO
```


#### 3.3.2 Compute pathway importances (optional)
```{r compute-dellOrso-pathway-importances, eval=FASLE}
#TODO
```


